<!doctype html><!-- This site was created with Wowchemy. https://www.wowchemy.com --><!-- Last Published: July 28, 2023 --><html lang=en-us><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta http-equiv=x-ua-compatible content="IE=edge"><meta name=generator content="Wowchemy 5.7.0 for Hugo"><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=preload as=style href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap"><link rel=stylesheet href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media=print onload='this.media="all"'><link rel=stylesheet href=/css/vendor-bundle.min.16f785cdb553c8c4431db6775122af35.css media=print onload='this.media="all"'><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/academicons@1.9.2/css/academicons.min.css integrity="sha512-KlJCpRsLf+KKu2VQa5vmRuClRFjxc5lXO03ixZt82HZUk41+1I0bD8KBSA0fY290ayMfWYI9udIqeOWSu1/uZg==" crossorigin=anonymous media=print onload='this.media="all"'><link rel=stylesheet href=/css/wowchemy.e5d7adca760216d3b7e28ea434e81f6f.css><link rel=stylesheet href=/css/libs/chroma/github-light.min.css title=hl-light media=print onload='this.media="all"'><link rel=stylesheet href=/css/libs/chroma/dracula.min.css title=hl-dark media=print onload='this.media="all"' disabled><script async src="https://www.googletagmanager.com/gtag/js?id=UA-162525500-1"></script>
<script>window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments)}function trackOutboundLink(e,t){gtag("event","click",{event_category:"outbound",event_label:e,transport_type:"beacon",event_callback:function(){t!=="_blank"&&(document.location=e)}}),console.debug("Outbound link clicked: "+e)}function onClickCallback(e){if(e.target.tagName!=="A"||e.target.host===window.location.host)return;trackOutboundLink(e.target,e.target.getAttribute("target"))}gtag("js",new Date),gtag("config","UA-162525500-1",{}),gtag("set",{cookie_flags:"SameSite=None;Secure"}),document.addEventListener("click",onClickCallback,!1)</script><script>var _hmt=_hmt||[];(function(){var e,t=document.createElement("script");t.src="https://hm.baidu.com/hm.js?67583aee7371a754cff04fa0a471bca3",e=document.getElementsByTagName("script")[0],e.parentNode.insertBefore(t,e)})()</script><meta name=author content="Hua Zou"><meta name=description content="差异分析是转录组数据分析的必备技能之一，但众多的转录组分析R包*DESeq2*，*limma*和*edgeR*让分析人员有选择问题，那么它们之间的优劣到底是如何的呢?"><link rel=alternate hreflang=en-us href=https://zouhua.top/post/math_statistics/2022-10-27-da-methods-comparsion/><link rel=canonical href=https://zouhua.top/post/math_statistics/2022-10-27-da-methods-comparsion/><link rel=manifest href=/manifest.webmanifest><link rel=icon type=image/png href=/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_32x32_fill_lanczos_center_3.png><link rel=apple-touch-icon type=image/png href=/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_180x180_fill_lanczos_center_3.png><meta name=theme-color content="#1565c0"><meta property="twitter:card" content="summary"><meta property="twitter:site" content="@Hua Zou"><meta property="twitter:creator" content="@Hua Zou"><meta property="twitter:image" content="https://zouhua.top/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_512x512_fill_lanczos_center_3.png"><meta property="og:site_name" content="Hua's Cabin"><meta property="og:url" content="https://zouhua.top/post/math_statistics/2022-10-27-da-methods-comparsion/"><meta property="og:title" content="转录组差异分析（DESeq2+limma+edgeR+t-test/wilcox-test）总结 | Hua's Cabin"><meta property="og:description" content="差异分析是转录组数据分析的必备技能之一，但众多的转录组分析R包*DESeq2*，*limma*和*edgeR*让分析人员有选择问题，那么它们之间的优劣到底是如何的呢?"><meta property="og:image" content="https://zouhua.top/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_512x512_fill_lanczos_center_3.png"><meta property="og:locale" content="en-us"><meta property="article:published_time" content="2022-10-20T20:13:14+00:00"><meta property="article:modified_time" content="2022-10-20T22:13:14+00:00"><script type=application/ld+json>{"@context":"https://schema.org","@type":"BlogPosting","mainEntityOfPage":{"@type":"WebPage","@id":"https://zouhua.top/post/math_statistics/2022-10-27-da-methods-comparsion/"},"headline":"转录组差异分析（DESeq2+limma+edgeR+t-test/wilcox-test）总结","datePublished":"2022-10-20T20:13:14Z","dateModified":"2022-10-20T22:13:14Z","author":{"@type":"Person","name":"Hua Zou"},"publisher":{"@type":"Organization","name":"Hua's Cabin","logo":{"@type":"ImageObject","url":"https://zouhua.top/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_192x192_fill_lanczos_center_3.png"}},"description":"差异分析是转录组数据分析的必备技能之一，但众多的转录组分析R包*DESeq2*，*limma*和*edgeR*让分析人员有选择问题，那么它们之间的优劣到底是如何的呢?"}</script><title>转录组差异分析（DESeq2+limma+edgeR+t-test/wilcox-test）总结 | Hua's Cabin</title></head><body id=top data-spy=scroll data-offset=70 data-target=#TableOfContents class=page-wrapper data-wc-page-id=502dca6c954348598c688fab47f1c873><script src=/js/wowchemy-init.min.ec9d49ca50e4b80bdb08f0417a28ed84.js></script><div class="page-header header--fixed"><header><nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id=navbar-main><div class=container-xl><div class="d-none d-lg-inline-flex"><a class=navbar-brand href=/>Hua's Cabin</a></div><button type=button class=navbar-toggler data-toggle=collapse data-target=#navbar-content aria-controls=navbar-content aria-expanded=false aria-label="Toggle navigation">
<span><i class="fas fa-bars"></i></span></button><div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none"><a class=navbar-brand href=/>Hua's Cabin</a></div><div class="navbar-collapse main-menu-item collapse justify-content-start" id=navbar-content><ul class="navbar-nav d-md-inline-flex"><li class=nav-item><a class=nav-link href=/#about><span>Home</span></a></li><li class=nav-item><a class=nav-link href=/#posts><span>Posts</span></a></li><li class=nav-item><a class=nav-link href=/#experience><span>Experience</span></a></li><li class=nav-item><a class=nav-link href=/#publications><span>Publications</span></a></li><li class=nav-item><a class=nav-link href=/#contact><span>Contact</span></a></li><li class=nav-item><a class=nav-link href=/#bookdown><span>Bookdown</span></a></li></ul></div><ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2"><li class=nav-item><a class="nav-link js-search" href=# aria-label=Search><i class="fas fa-search" aria-hidden=true></i></a></li><li class="nav-item dropdown theme-dropdown"><a href=# class=nav-link data-toggle=dropdown aria-haspopup=true aria-label="Display preferences"><i class="fas fa-moon" aria-hidden=true></i></a><div class=dropdown-menu><a href=# class="dropdown-item js-set-theme-light"><span>Light</span></a>
<a href=# class="dropdown-item js-set-theme-dark"><span>Dark</span></a>
<a href=# class="dropdown-item js-set-theme-auto"><span>Automatic</span></a></div></li></ul></div></nav></header></div><div class=page-body><article class=article><div class="article-container pt-3"><h1>转录组差异分析（DESeq2+limma+edgeR+t-test/wilcox-test）总结</h1><p class=page-subtitle>Comparison of Differential Analysis Methods on RNA-seq data</p><div class=article-metadata><div><span>Hua Zou</span></div><span class=article-date>Last updated on
Oct 20, 2022</span>
<span class=middot-divider></span>
<span class=article-reading-time>18 min read</span>
<span class=middot-divider></span>
<span class=article-categories><i class="fas fa-folder mr-1"></i><a href=/category/statistics/>Statistics</a></span></div></div><div class=article-container><div class=article-style><div><h2>Table Of Contents</h2><nav id=TableOfContents><ul><li><a href=#前言>前言</a></li><li><a href=#大纲>大纲</a></li><li><a href=#导入r包>导入R包</a></li><li><a href=#转录组数据>转录组数据</a></li><li><a href=#标准化>标准化</a></li><li><a href=#整体水平比较>整体水平比较</a><ul><li></li><li><a href=#降维分析>降维分析</a></li><li><a href=#热图>热图</a></li></ul></li><li><a href=#deseq2>DESeq2</a><ul><li></li><li><a href=#提取结果>提取结果</a></li><li><a href=#差异分组>差异分组</a></li><li><a href=#volcano-plot>Volcano plot</a></li><li><a href=#boxplot>boxplot</a></li></ul></li><li><a href=#limma>limma</a><ul><li><a href=#voom功能>Voom功能</a></li><li><a href=#计算过程>计算过程</a></li><li><a href=#差异分组-1>差异分组</a></li></ul></li><li><a href=#edger>edgeR</a><ul><li><a href=#计算过程-1>计算过程</a></li><li><a href=#差异分组-2>差异分组</a></li></ul></li><li><a href=#wilcox-rank-sum-test-or-t-test>Wilcox-rank-sum test or t-test</a><ul><li><a href=#计算差异结果-1>计算差异结果</a></li><li><a href=#差异分组-3>差异分组</a></li></ul></li><li><a href=#不同方法的结果比较>不同方法的结果比较</a><ul><li><a href=#保存结果>保存结果</a></li><li><a href=#差异基因数目分布>差异基因数目分布</a></li><li><a href=#差异基因交集venn>差异基因交集Venn</a></li><li><a href=#差异基因heatmapvolcano>差异基因heatmap+volcano</a></li><li><a href=#共有差异基因venn>共有差异基因venn</a></li><li><a href=#差异基因log2foldchange相关性分析>差异基因Log2Foldchange相关性分析</a></li><li><a href=#共有差异基因总结>共有差异基因总结</a></li></ul></li><li><a href=#总结>总结</a></li><li><a href=#systemic-information>systemic information</a></li><li><a href=#reference>Reference</a></li></ul></nav></div><h2 id=前言>前言</h2><p>差异分析是转录组数据分析的必需技能之一，但众多的转录组分析R包如<em>DESeq2</em>，<em>limma</em>和<em>edgeR</em>等让分析人员不知如何选择，还有它们之间的优劣如何？我将在本文详细探讨常用差异分析R包以及结合t-test/wilcox-rank-sum test的差异分析结果的异同点。</p><h2 id=大纲>大纲</h2><p>本文要点由以下几点构成：</p><ol><li><p>下载以及导入测试数据（批量安装R包）；</p></li><li><p>基因表达count矩阵的标准化方法（F(R)PKM/TPM）；</p></li><li><p>基因整体水平分布（PCA/tSNE/UMAP；heatmap）；</p></li><li><p><em>DESeq2</em>差异分析实现以及结果解析；</p></li><li><p><em>limma</em>差异分析实现以及结果解析；</p></li><li><p><em>edgeR</em>差异分析实现以及结果解析；</p></li><li><p>结合<em>t-test</em>或<em>wilcox-rank-sum-test</em>方法的差异分析实现以及结果解析（是否符合正态分布选择检验方法）；</p></li><li><p>不同方法的结果比较（volcano plot+heatmap+venn）；</p></li><li><p>总结。</p></li></ol><h2 id=导入r包>导入R包</h2><p>本次分析需要在R中批量安装包，具体安装方法可参考<a href=https://zouhua.top/archives/fa0b9789.html target=_blank rel=noopener>如何安装R包</a>。先导入基础R包，在后面每个差异分析模块再导入所需要的差异分析R包。</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>dplyr</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>tibble</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>data.table</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>ggplot2</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>patchwork</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>cowplot</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># rm(list = ls())</span>
</span></span><span class=line><span class=cl><span class=nf>options</span><span class=p>(</span><span class=n>stringsAsFactors</span> <span class=o>=</span> <span class=bp>F</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>options</span><span class=p>(</span><span class=n>future.globals.maxSize</span> <span class=o>=</span> <span class=m>1000</span> <span class=o>*</span> <span class=m>1024</span><span class=n>^2</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>grp</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Normal&#34;</span><span class=p>,</span> <span class=s>&#34;Tumor&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>grp.col</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;#568875&#34;</span><span class=p>,</span> <span class=s>&#34;#73FAFC&#34;</span><span class=p>)</span>
</span></span></code></pre></div><h2 id=转录组数据>转录组数据</h2><p>本文下载的TCGA-HNSC转录组数据是通过本人先前撰写的R脚本实现的，大家可参考<a href=https://zouhua.top/archives/be2153a7.html target=_blank rel=noopener>Downloading and preprocessing TCGA Data through R program language</a>文章自行下载，也可以邮件询问我百度网盘密码。</p><p><strong>百度网盘链接：https://pan.baidu.com/s/1Daz5UsOd39T8r8K6zxPASQ</strong></p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>phenotype</span> <span class=o>&lt;-</span> <span class=nf>fread</span><span class=p>(</span><span class=s>&#34;TCGA-HNSC-post_mRNA_clinical.csv&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>count</span> <span class=o>&lt;-</span> <span class=nf>fread</span><span class=p>(</span><span class=s>&#34;TCGA-HNSC-post_mRNA_profile.tsv&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>table</span><span class=p>(</span><span class=n>phenotype</span><span class=o>$</span><span class=n>Group</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126151602.png alt loading=lazy data-zoomable></div></div></figure></p><h2 id=标准化>标准化</h2><p>标准化的目的是为了降低测序深度以及基因长度对基因表达谱下游分析的影响。测序深度越高则map到基因的reads也越多，同理基因长度越长则map到的reads也越多，最后对应的counts数目也越多。</p><ul><li>RPM/CPM: Reads/Counts of exon model per Million mapped reads (每百万映射读取的reads)</li></ul><p>$$RPM = \frac{ExonMappedReads * 10^{6}}{TotalMappedReads}$$</p><ul><li>RPKM: Reads Per Kilobase of exon model per Million mapped reads (每千个碱基的转录每百万映射读取的reads)</li></ul><p>$$RPKM = \frac{ExonMappedReads * 10^{9}}{TotalMappedReads * ExonLength}$$</p><ul><li>FPKM: Fragments Per Kilobase of exon model per Million mapped fragments(每千个碱基的转录每百万映射读取的fragments)， 适用于PE测序。</li></ul><p>$$ FPKM = \frac{ExonMappedFragments * 10^{9}}{TotalMappedFragments * ExonLength}$$</p><ul><li>TPM：Transcripts Per Kilobase of exon model per Million mapped reads (每千个碱基的转录每百万映射读取的Transcripts)</li></ul><p>$$TPM= \frac{N_i/L_i * 10^{6}}{sum(N_1/L_1+N_2/L_2+&mldr;+N_j/L_j+&mldr;+N_n/L_n)}$$
$N_i$为比对到第i个exon的reads数； $L_i$为第i个exon的长度；sum()为所有 (n个)exon按长度进行标准化之后数值的和</p><ul><li>获取gene length表：对<em>Homo_sapiens.GRCh38.101</em>版本数据处理获取gene length数据；<em>human_gene_all.tsv</em>是使用biomart包获取gene symbol和ensembleID的对应关系表。</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>geneLength</span> <span class=o>&lt;-</span> <span class=nf>fread</span><span class=p>(</span><span class=s>&#34;Homo_sapiens.GRCh38.101.genelength.tsv&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>geneIdAll</span> <span class=o>&lt;-</span> <span class=nf>fread</span><span class=p>(</span><span class=s>&#34;human_gene_all.tsv&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>geneIdLength</span> <span class=o>&lt;-</span> <span class=n>geneIdAll</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>transcript_biotype</span> <span class=o>==</span> <span class=s>&#34;protein_coding&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>ensembl_gene_id</span><span class=p>,</span> <span class=n>external_gene_name</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>inner_join</span><span class=p>(</span><span class=n>geneLength</span><span class=p>,</span> <span class=n>by</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;ensembl_gene_id&#34;</span><span class=o>=</span><span class=s>&#34;V1&#34;</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=o>-</span><span class=n>ensembl_gene_id</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>distinct</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>Length</span><span class=o>=</span><span class=n>V2</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>geneIdLengthUniq</span> <span class=o>&lt;-</span> <span class=n>geneIdLength</span><span class=nf>[pmatch</span><span class=p>(</span><span class=n>count</span><span class=o>$</span><span class=n>Feature</span><span class=p>,</span> <span class=n>geneIdLength</span><span class=o>$</span><span class=n>external_gene_name</span><span class=p>),</span> <span class=n>]</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>filter</span><span class=p>(</span><span class=o>!</span><span class=nf>is.na</span><span class=p>(</span><span class=n>Length</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>arrange</span><span class=p>(</span><span class=n>external_gene_name</span><span class=p>)</span> 
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>count_cln</span> <span class=o>&lt;-</span> <span class=n>count</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Feature</span><span class=o>%in%</span><span class=n>geneIdLengthUniq</span><span class=o>$</span><span class=n>external_gene_name</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=nf>arrange</span><span class=p>(</span><span class=n>Feature</span><span class=p>)</span> <span class=o>%&gt;%</span> <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;Feature&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>if</span><span class=p>(</span><span class=o>!</span><span class=nf>any</span><span class=p>(</span><span class=n>geneIdLengthUniq</span><span class=o>$</span><span class=n>external_gene_name</span> <span class=o>==</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>count_cln</span><span class=p>))){</span>
</span></span><span class=line><span class=cl>  <span class=nf>message</span><span class=p>(</span><span class=s>&#34;Order of GeneName is wrong&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>gene_lengths</span> <span class=o>&lt;-</span><span class=n>geneIdLengthUniq</span><span class=o>$</span><span class=n>Length</span>
</span></span><span class=line><span class=cl><span class=nf>head</span><span class=p>(</span><span class=n>geneIdLengthUniq</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126151702.png alt loading=lazy data-zoomable></div></div></figure></p><ul><li>RPKM/FPKM/TPM 转换</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-fallback data-lang=fallback><span class=line><span class=cl>countToFpkm &lt;- function(counts, lengths){
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  pm &lt;- sum(counts) /1e6
</span></span><span class=line><span class=cl>  rpm &lt;- counts/pm
</span></span><span class=line><span class=cl>  rpm/(lengths/1000)
</span></span><span class=line><span class=cl>}
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>countToTpm &lt;- function(counts, lengths) {
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  rpk &lt;- counts/(lengths/1000)
</span></span><span class=line><span class=cl>  coef &lt;- sum(rpk) / 1e6
</span></span><span class=line><span class=cl>  rpk/coef
</span></span><span class=line><span class=cl>}
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl># FPKM
</span></span><span class=line><span class=cl>count_FPKM &lt;- apply(count_cln, 2, function(x){countToFpkm(x, gene_lengths)}) %&gt;% 
</span></span><span class=line><span class=cl>  data.frame()
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl># TPM
</span></span><span class=line><span class=cl>count_TPM &lt;- apply(count_cln, 2, function(x){countToTpm(x, gene_lengths)}) %&gt;% 
</span></span><span class=line><span class=cl>  data.frame() 
</span></span><span class=line><span class=cl>head(count_cln)
</span></span><span class=line><span class=cl>head(count_FPKM)
</span></span><span class=line><span class=cl>head(count_TPM)
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126151741.png alt loading=lazy data-zoomable></div></div></figure></p><h2 id=整体水平比较>整体水平比较</h2><p>在做差异分析前，一般可以对数据做一个降维处理，然后看不同分组是否能在二维展开平面区分开。</p><h4 id=expressionset>ExpressionSet</h4><p>先将数据存成ExpressionSet格式，可参考<a href=https://zouhua.top/archives/619fe135.html target=_blank rel=noopener>RNA-seq数据的批次校正方法</a>文章。ExpressionSet对象数据包含表达谱和metadata等数据，这方便后期分析。</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>getExprSet</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>metadata</span><span class=o>=</span><span class=n>phenotype</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>profile</span><span class=o>=</span><span class=n>count_cln</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>occurrence</span><span class=o>=</span><span class=m>0.2</span><span class=p>){</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=c1># metadata=phenotype</span>
</span></span><span class=line><span class=cl>  <span class=c1># profile=count_cln</span>
</span></span><span class=line><span class=cl>  <span class=c1># occurrence=0.2</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>sid</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>metadata</span><span class=o>$</span><span class=n>SampleID</span><span class=p>,</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>profile</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=c1># phenotype</span>
</span></span><span class=line><span class=cl>  <span class=n>phe</span> <span class=o>&lt;-</span> <span class=n>metadata</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>SampleID</span><span class=o>%in%</span><span class=n>sid</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=c1># profile by occurrence</span>
</span></span><span class=line><span class=cl>  <span class=n>prf</span> <span class=o>&lt;-</span> <span class=n>profile</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;tmp&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>filter</span><span class=p>(</span><span class=nf>apply</span><span class=p>(</span><span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>.,</span> <span class=o>-</span><span class=nf>one_of</span><span class=p>(</span><span class=s>&#34;tmp&#34;</span><span class=p>)),</span> <span class=m>1</span><span class=p>,</span> <span class=nf>function</span><span class=p>(</span><span class=n>x</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>            <span class=nf>sum</span><span class=p>(</span><span class=n>x</span> <span class=o>!=</span> <span class=m>0</span><span class=p>)</span><span class=o>/</span><span class=nf>length</span><span class=p>(</span><span class=n>x</span><span class=p>)})</span> <span class=o>&gt;</span> <span class=n>occurrence</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=n>tmp</span><span class=p>,</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>phe</span><span class=p>)))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;tmp&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># determine the right order between profile and phenotype</span>
</span></span><span class=line><span class=cl>  <span class=nf>for</span><span class=p>(</span><span class=n>i</span> <span class=n>in</span> <span class=m>1</span><span class=o>:</span><span class=nf>ncol</span><span class=p>(</span><span class=n>prf</span><span class=p>)){</span>
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>prf</span><span class=p>)</span><span class=n>[i]</span> <span class=o>==</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>phe</span><span class=p>)</span><span class=n>[i]</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=nf>stop</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=n>i</span><span class=p>,</span> <span class=s>&#34; Wrong&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>convert</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>exprs</span> <span class=o>&lt;-</span> <span class=nf>as.matrix</span><span class=p>(</span><span class=n>prf</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>adf</span> <span class=o>&lt;-</span>  <span class=nf>new</span><span class=p>(</span><span class=s>&#34;AnnotatedDataFrame&#34;</span><span class=p>,</span> <span class=n>data</span><span class=o>=</span><span class=n>phe</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>experimentData</span> <span class=o>&lt;-</span> <span class=nf>new</span><span class=p>(</span><span class=s>&#34;MIAME&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>        <span class=n>name</span><span class=o>=</span><span class=s>&#34;Hua Zou&#34;</span><span class=p>,</span> <span class=n>lab</span><span class=o>=</span><span class=s>&#34;UCAS&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>        <span class=n>contact</span><span class=o>=</span><span class=s>&#34;zouhua@outlook.com&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>        <span class=n>title</span><span class=o>=</span><span class=s>&#34;TCGA-HNSC&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>        <span class=n>abstract</span><span class=o>=</span><span class=s>&#34;The gene ExpressionSet&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>        <span class=n>url</span><span class=o>=</span><span class=s>&#34;www.zouhua.top&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>        <span class=n>other</span><span class=o>=</span><span class=nf>list</span><span class=p>(</span><span class=n>notes</span><span class=o>=</span><span class=s>&#34;Created from text files&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>expressionSet</span> <span class=o>&lt;-</span> <span class=nf>new</span><span class=p>(</span><span class=s>&#34;ExpressionSet&#34;</span><span class=p>,</span> <span class=n>exprs</span><span class=o>=</span><span class=n>exprs</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>phenoData</span><span class=o>=</span><span class=n>adf</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>experimentData</span><span class=o>=</span><span class=n>experimentData</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>expressionSet</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>ExprSet_count</span> <span class=o>&lt;-</span> <span class=nf>getExprSet</span><span class=p>(</span><span class=n>profile</span><span class=o>=</span><span class=n>count_cln</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>ExprSet_FPKM</span> <span class=o>&lt;-</span> <span class=nf>getExprSet</span><span class=p>(</span><span class=n>profile</span><span class=o>=</span><span class=n>count_FPKM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>ExprSet_TPM</span> <span class=o>&lt;-</span> <span class=nf>getExprSet</span><span class=p>(</span><span class=n>profile</span><span class=o>=</span><span class=n>count_TPM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>ExprSet_count</span>
</span></span><span class=line><span class=cl><span class=n>ExprSet_FPKM</span>
</span></span><span class=line><span class=cl><span class=n>ExprSet_TPM</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126151828.png alt loading=lazy data-zoomable></div></div></figure></p><h3 id=降维分析>降维分析</h3><p>高纬度数据降维方法很多，我们这里选择了PCA+tSNE+UMAP分别展示降维后的结果，可参考文章<a href=https://zouhua.top/archives/f4cccace.html target=_blank rel=noopener>高纬度数据降维方法的R实现</a>。另外合并多个图的方法，请参考文章<a href=https://zouhua.top/archives/b9aef638.html target=_blank rel=noopener>patchwork:合并多个R图的包</a>。</p><ul><li>PCA</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>PCAFun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_count</span><span class=p>){</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># dataset = ExprSet_count </span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>convert</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>dataset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>profile</span> <span class=o>&lt;-</span> <span class=nf>exprs</span><span class=p>(</span><span class=n>dataset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># pca </span>
</span></span><span class=line><span class=cl>  <span class=n>pca</span> <span class=o>&lt;-</span> <span class=nf>prcomp</span><span class=p>(</span><span class=nf>scale</span><span class=p>(</span><span class=nf>t</span><span class=p>(</span><span class=n>profile</span><span class=p>),</span> <span class=n>center</span> <span class=o>=</span> <span class=bp>T</span><span class=p>,</span> <span class=n>scale</span> <span class=o>=</span> <span class=bp>T</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>factoextra</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>eig</span> <span class=o>&lt;-</span> <span class=nf>get_eig</span><span class=p>(</span><span class=n>pca</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1># explains variable </span>
</span></span><span class=line><span class=cl>  <span class=n>explains</span> <span class=o>&lt;-</span> <span class=nf>paste0</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;PC&#34;</span><span class=p>,</span> <span class=nf>seq</span><span class=p>(</span><span class=m>2</span><span class=p>)),</span> <span class=s>&#34;(&#34;</span><span class=p>,</span> <span class=nf>paste0</span><span class=p>(</span><span class=nf>round</span><span class=p>(</span><span class=n>eig[1</span><span class=o>:</span><span class=m>2</span><span class=p>,</span> <span class=m>2</span><span class=n>]</span><span class=p>,</span> <span class=m>2</span><span class=p>),</span> <span class=s>&#34;%&#34;</span><span class=p>),</span> <span class=s>&#34;)&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1># principal component score of each sample</span>
</span></span><span class=line><span class=cl>  <span class=n>score</span> <span class=o>&lt;-</span> <span class=nf>inner_join</span><span class=p>(</span><span class=n>pca</span><span class=o>$</span><span class=n>x</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>                        <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=m>1</span><span class=o>:</span><span class=m>2</span><span class=p>))</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>                        <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>),</span> 
</span></span><span class=line><span class=cl>                      <span class=n>metadata</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                      <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>grp</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># PERMANOVA</span>
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>vegan</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>set.seed</span><span class=p>(</span><span class=m>123</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>if</span><span class=p>(</span><span class=nf>any</span><span class=p>(</span><span class=n>profile</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>)){</span>
</span></span><span class=line><span class=cl>    <span class=n>res_adonis</span> <span class=o>&lt;-</span> <span class=nf>adonis</span><span class=p>(</span><span class=nf>vegdist</span><span class=p>(</span><span class=nf>t</span><span class=p>(</span><span class=n>profile</span><span class=p>),</span> <span class=n>method</span> <span class=o>=</span> <span class=s>&#34;manhattan&#34;</span><span class=p>)</span> <span class=o>~</span> <span class=n>metadata</span><span class=o>$</span><span class=n>Group</span><span class=p>,</span> <span class=n>permutations</span> <span class=o>=</span> <span class=m>999</span><span class=p>)</span> 
</span></span><span class=line><span class=cl>  <span class=p>}</span><span class=n>else</span><span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>res_adonis</span> <span class=o>&lt;-</span> <span class=nf>adonis</span><span class=p>(</span><span class=nf>vegdist</span><span class=p>(</span><span class=nf>t</span><span class=p>(</span><span class=n>profile</span><span class=p>),</span> <span class=n>method</span> <span class=o>=</span> <span class=s>&#34;bray&#34;</span><span class=p>)</span> <span class=o>~</span> <span class=n>metadata</span><span class=o>$</span><span class=n>Group</span><span class=p>,</span> <span class=n>permutations</span> <span class=o>=</span> <span class=m>999</span><span class=p>)</span>    
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_pvalue</span> <span class=o>&lt;-</span> <span class=n>res_adonis[[1]][[</span><span class=s>&#34;Pr(&gt;F)&#34;</span><span class=n>]][1]</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_rsquared</span> <span class=o>&lt;-</span> <span class=nf>round</span><span class=p>(</span><span class=n>res_adonis[[1]][[</span><span class=s>&#34;R2&#34;</span><span class=n>]][1]</span><span class=p>,</span><span class=m>3</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1>#use the bquote function to format adonis results to be annotated on the ordination plot.</span>
</span></span><span class=line><span class=cl>  <span class=n>signi_label</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>cut</span><span class=p>(</span><span class=n>adn_pvalue</span><span class=p>,</span><span class=n>breaks</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=o>-</span><span class=kc>Inf</span><span class=p>,</span> <span class=m>0.001</span><span class=p>,</span> <span class=m>0.01</span><span class=p>,</span> <span class=m>0.05</span><span class=p>,</span> <span class=kc>Inf</span><span class=p>),</span> <span class=n>label</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;***&#34;</span><span class=p>,</span> <span class=s>&#34;**&#34;</span><span class=p>,</span> <span class=s>&#34;*&#34;</span><span class=p>,</span> <span class=s>&#34;.&#34;</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_res_format</span> <span class=o>&lt;-</span> <span class=nf>bquote</span><span class=p>(</span><span class=nf>atop</span><span class=p>(</span><span class=nf>atop</span><span class=p>(</span><span class=s>&#34;PERMANOVA&#34;</span><span class=p>,</span><span class=n>R^2</span><span class=o>==~</span><span class=n>.(adn_rsquared</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                                  <span class=nf>atop</span><span class=p>(</span><span class=s>&#34;p-value=&#34;</span><span class=o>~</span><span class=n>.(adn_pvalue</span><span class=p>)</span><span class=o>~</span><span class=n>.(signi_label</span><span class=p>),</span> <span class=nf>phantom</span><span class=p>())))</span> 
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>score</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>PC1</span><span class=p>,</span> <span class=n>y</span><span class=o>=</span><span class=n>PC2</span><span class=p>))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>fill</span><span class=o>=</span><span class=n>Group</span><span class=p>),</span> <span class=n>size</span><span class=o>=</span><span class=m>2</span><span class=p>,</span> <span class=n>shape</span><span class=o>=</span><span class=m>21</span><span class=p>,</span> <span class=n>stroke</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>stat_ellipse</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=n>Group</span><span class=p>),</span> <span class=n>level</span> <span class=o>=</span> <span class=m>0.95</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>1.5</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>labs</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>explains[1]</span><span class=p>,</span> <span class=n>y</span><span class=o>=</span><span class=n>explains[2]</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>scale_color_manual</span><span class=p>(</span><span class=n>values</span> <span class=o>=</span> <span class=n>grp.col</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>scale_fill_manual</span><span class=p>(</span><span class=n>name</span> <span class=o>=</span> <span class=s>&#34;Condition&#34;</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                                <span class=n>values</span> <span class=o>=</span> <span class=n>grp.col</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span> <span class=n>x</span> <span class=o>=</span> <span class=nf>max</span><span class=p>(</span><span class=n>score</span><span class=o>$</span><span class=n>PC1</span><span class=p>)</span> <span class=o>-</span> <span class=m>8</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>y</span> <span class=o>=</span> <span class=nf>min</span><span class=p>(</span><span class=n>score</span><span class=o>$</span><span class=n>PC1</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                       <span class=n>label</span> <span class=o>=</span> <span class=n>adn_res_format</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>size</span> <span class=o>=</span> <span class=m>6</span><span class=p>)</span><span class=o>+</span> 
</span></span><span class=line><span class=cl>              <span class=nf>guides</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=s>&#34;none&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>theme_classic</span><span class=p>()</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>theme</span><span class=p>(</span><span class=n>axis.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>axis.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>strip.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span> 
</span></span><span class=line><span class=cl>                    <span class=n>panel.grid</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>11</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.position</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.justification</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.background</span> <span class=o>=</span> <span class=nf>element_rect</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>fill</span> <span class=o>=</span> <span class=s>&#34;white&#34;</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.5</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>PCA_count</span> <span class=o>&lt;-</span> <span class=nf>PCAFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_count</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>PCA_FPKM</span> <span class=o>&lt;-</span> <span class=nf>PCAFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_FPKM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>PCA_TPM</span> <span class=o>&lt;-</span> <span class=nf>PCAFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_TPM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>plot_grid</span><span class=p>(</span><span class=n>PCA_count</span><span class=p>,</span> <span class=n>PCA_FPKM</span><span class=p>,</span> <span class=n>PCA_TPM</span><span class=p>,</span> <span class=n>nrow</span><span class=o>=</span><span class=m>1</span><span class=p>,</span> <span class=n>align</span><span class=o>=</span><span class=s>&#34;hv&#34;</span><span class=p>,</span> <span class=n>labels</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;Counts&#34;</span><span class=p>,</span> <span class=s>&#34;FPKM&#34;</span><span class=p>,</span> <span class=s>&#34;TPM&#34;</span><span class=p>))</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126151922.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Notes:</strong> PCA的结果在三类数据中表现都不是很好，也就是组间区分不是明显，但是PERMANOVA的检验结果显示p值是显著差异的，可是$R^2$却偏低，也就是解释度很低。</p><ul><li>tSNE</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>RtsneFun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_count</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                     <span class=n>perpl</span> <span class=o>=</span> <span class=m>30</span><span class=p>){</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># dataset = ExprSet_count</span>
</span></span><span class=line><span class=cl>  <span class=c1># perpl = 10</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>convert</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>dataset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>profile</span> <span class=o>&lt;-</span> <span class=nf>t</span><span class=p>(</span><span class=nf>exprs</span><span class=p>(</span><span class=n>dataset</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># Rtsne</span>
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>Rtsne</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1>#set.seed(123)</span>
</span></span><span class=line><span class=cl>  <span class=n>Rtsne</span> <span class=o>&lt;-</span> <span class=nf>Rtsne</span><span class=p>(</span><span class=n>profile</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                 <span class=n>dims</span><span class=o>=</span><span class=m>2</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                 <span class=n>perplexity</span><span class=o>=</span><span class=n>perpl</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                 <span class=n>verbose</span><span class=o>=</span><span class=kc>TRUE</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                 <span class=n>max_iter</span><span class=o>=</span><span class=m>500</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                 <span class=n>eta</span><span class=o>=</span><span class=m>200</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>point</span> <span class=o>&lt;-</span> <span class=n>Rtsne</span><span class=o>$</span><span class=n>Y</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=m>1</span><span class=o>:</span><span class=m>2</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>setNames</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;tSNE1&#34;</span><span class=p>,</span> <span class=s>&#34;tSNE2&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>rownames</span><span class=p>(</span><span class=n>point</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>profile</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>score</span> <span class=o>&lt;-</span> <span class=nf>inner_join</span><span class=p>(</span><span class=n>point</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>),</span> 
</span></span><span class=line><span class=cl>                      <span class=n>metadata</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                      <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>grp</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># PERMANOVA</span>
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>vegan</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>set.seed</span><span class=p>(</span><span class=m>123</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>if</span><span class=p>(</span><span class=nf>any</span><span class=p>(</span><span class=n>profile</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>)){</span>
</span></span><span class=line><span class=cl>    <span class=n>res_adonis</span> <span class=o>&lt;-</span> <span class=nf>adonis</span><span class=p>(</span><span class=nf>vegdist</span><span class=p>(</span><span class=n>profile</span><span class=p>,</span> <span class=n>method</span> <span class=o>=</span> <span class=s>&#34;manhattan&#34;</span><span class=p>)</span> <span class=o>~</span> <span class=n>metadata</span><span class=o>$</span><span class=n>Group</span><span class=p>,</span> <span class=n>permutations</span> <span class=o>=</span> <span class=m>999</span><span class=p>)</span> 
</span></span><span class=line><span class=cl>  <span class=p>}</span><span class=n>else</span><span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>res_adonis</span> <span class=o>&lt;-</span> <span class=nf>adonis</span><span class=p>(</span><span class=nf>vegdist</span><span class=p>(</span><span class=n>profile</span><span class=p>,</span> <span class=n>method</span> <span class=o>=</span> <span class=s>&#34;bray&#34;</span><span class=p>)</span> <span class=o>~</span> <span class=n>metadata</span><span class=o>$</span><span class=n>Group</span><span class=p>,</span> <span class=n>permutations</span> <span class=o>=</span> <span class=m>999</span><span class=p>)</span>    
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_pvalue</span> <span class=o>&lt;-</span> <span class=n>res_adonis[[1]][[</span><span class=s>&#34;Pr(&gt;F)&#34;</span><span class=n>]][1]</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_rsquared</span> <span class=o>&lt;-</span> <span class=nf>round</span><span class=p>(</span><span class=n>res_adonis[[1]][[</span><span class=s>&#34;R2&#34;</span><span class=n>]][1]</span><span class=p>,</span><span class=m>3</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1>#use the bquote function to format adonis results to be annotated on the ordination plot.</span>
</span></span><span class=line><span class=cl>  <span class=n>signi_label</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>cut</span><span class=p>(</span><span class=n>adn_pvalue</span><span class=p>,</span><span class=n>breaks</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=o>-</span><span class=kc>Inf</span><span class=p>,</span> <span class=m>0.001</span><span class=p>,</span> <span class=m>0.01</span><span class=p>,</span> <span class=m>0.05</span><span class=p>,</span> <span class=kc>Inf</span><span class=p>),</span> <span class=n>label</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;***&#34;</span><span class=p>,</span> <span class=s>&#34;**&#34;</span><span class=p>,</span> <span class=s>&#34;*&#34;</span><span class=p>,</span> <span class=s>&#34;.&#34;</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_res_format</span> <span class=o>&lt;-</span> <span class=nf>bquote</span><span class=p>(</span><span class=nf>atop</span><span class=p>(</span><span class=nf>atop</span><span class=p>(</span><span class=s>&#34;PERMANOVA&#34;</span><span class=p>,</span><span class=n>R^2</span><span class=o>==~</span><span class=n>.(adn_rsquared</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                                  <span class=nf>atop</span><span class=p>(</span><span class=s>&#34;p-value=&#34;</span><span class=o>~</span><span class=n>.(adn_pvalue</span><span class=p>)</span><span class=o>~</span><span class=n>.(signi_label</span><span class=p>),</span> <span class=nf>phantom</span><span class=p>())))</span> 
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>score</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>tSNE1</span><span class=p>,</span> <span class=n>y</span><span class=o>=</span><span class=n>tSNE2</span><span class=p>))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>fill</span><span class=o>=</span><span class=n>Group</span><span class=p>),</span> <span class=n>size</span><span class=o>=</span><span class=m>3.5</span><span class=p>,</span> <span class=n>shape</span><span class=o>=</span><span class=m>21</span><span class=p>,</span> <span class=n>stroke</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>stat_ellipse</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=n>Group</span><span class=p>),</span> <span class=n>level</span> <span class=o>=</span> <span class=m>0.95</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>1.5</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>scale_color_manual</span><span class=p>(</span><span class=n>values</span> <span class=o>=</span> <span class=n>grp.col</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>scale_fill_manual</span><span class=p>(</span><span class=n>name</span> <span class=o>=</span> <span class=s>&#34;Condition&#34;</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                                <span class=n>values</span> <span class=o>=</span> <span class=n>grp.col</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span> <span class=n>x</span> <span class=o>=</span> <span class=nf>max</span><span class=p>(</span><span class=n>score</span><span class=o>$</span><span class=n>tSNE1</span><span class=p>)</span> <span class=o>-</span> <span class=m>8</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>y</span> <span class=o>=</span> <span class=nf>max</span><span class=p>(</span><span class=n>score</span><span class=o>$</span><span class=n>tSNE2</span><span class=p>)</span><span class=m>-5</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>label</span> <span class=o>=</span> <span class=n>adn_res_format</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>size</span> <span class=o>=</span> <span class=m>6</span><span class=p>)</span><span class=o>+</span> 
</span></span><span class=line><span class=cl>              <span class=nf>guides</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=s>&#34;none&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>theme_classic</span><span class=p>()</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>theme</span><span class=p>(</span><span class=n>axis.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>axis.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>strip.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span> 
</span></span><span class=line><span class=cl>                    <span class=n>panel.grid</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>11</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.position</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.justification</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.background</span> <span class=o>=</span> <span class=nf>element_rect</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>fill</span> <span class=o>=</span> <span class=s>&#34;white&#34;</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.5</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>Rtsne_count</span> <span class=o>&lt;-</span> <span class=nf>RtsneFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_count</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Rtsne_FPKM</span> <span class=o>&lt;-</span> <span class=nf>RtsneFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_FPKM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Rtsne_TPM</span> <span class=o>&lt;-</span> <span class=nf>RtsneFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_TPM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>plot_grid</span><span class=p>(</span><span class=n>Rtsne_count</span><span class=p>,</span> <span class=n>Rtsne_FPKM</span><span class=p>,</span> <span class=n>Rtsne_TPM</span><span class=p>,</span> <span class=n>nrow</span><span class=o>=</span><span class=m>1</span><span class=p>,</span> <span class=n>align</span><span class=o>=</span><span class=s>&#34;hv&#34;</span><span class=p>,</span> <span class=n>labels</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;Counts&#34;</span><span class=p>,</span> <span class=s>&#34;FPKM&#34;</span><span class=p>,</span> <span class=s>&#34;TPM&#34;</span><span class=p>))</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126151957.png alt loading=lazy data-zoomable></div></div></figure></p><ul><li>UMAP</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>UMAPFun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_count</span><span class=p>){</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># dataset = ExprSet_count</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>dataset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>profile</span> <span class=o>&lt;-</span> <span class=nf>t</span><span class=p>(</span><span class=nf>exprs</span><span class=p>(</span><span class=n>dataset</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># umap </span>
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>umap</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>umap</span> <span class=o>&lt;-</span> <span class=n>umap</span><span class=o>::</span><span class=nf>umap</span><span class=p>(</span><span class=n>profile</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>point</span> <span class=o>&lt;-</span> <span class=n>umap</span><span class=o>$</span><span class=n>layout</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>setNames</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;UMAP1&#34;</span><span class=p>,</span> <span class=s>&#34;UMAP2&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>rownames</span><span class=p>(</span><span class=n>point</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>profile</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>score</span> <span class=o>&lt;-</span> <span class=nf>inner_join</span><span class=p>(</span><span class=n>point</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>),</span> 
</span></span><span class=line><span class=cl>                      <span class=n>metadata</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                      <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>grp</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># PERMANOVA</span>
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>vegan</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>set.seed</span><span class=p>(</span><span class=m>123</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>if</span><span class=p>(</span><span class=nf>any</span><span class=p>(</span><span class=n>profile</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>)){</span>
</span></span><span class=line><span class=cl>    <span class=n>res_adonis</span> <span class=o>&lt;-</span> <span class=nf>adonis</span><span class=p>(</span><span class=nf>vegdist</span><span class=p>(</span><span class=n>profile</span><span class=p>,</span> <span class=n>method</span> <span class=o>=</span> <span class=s>&#34;manhattan&#34;</span><span class=p>)</span> <span class=o>~</span> <span class=n>metadata</span><span class=o>$</span><span class=n>Group</span><span class=p>,</span> <span class=n>permutations</span> <span class=o>=</span> <span class=m>999</span><span class=p>)</span> 
</span></span><span class=line><span class=cl>  <span class=p>}</span><span class=n>else</span><span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>res_adonis</span> <span class=o>&lt;-</span> <span class=nf>adonis</span><span class=p>(</span><span class=nf>vegdist</span><span class=p>(</span><span class=n>profile</span><span class=p>,</span> <span class=n>method</span> <span class=o>=</span> <span class=s>&#34;bray&#34;</span><span class=p>)</span> <span class=o>~</span> <span class=n>metadata</span><span class=o>$</span><span class=n>Group</span><span class=p>,</span> <span class=n>permutations</span> <span class=o>=</span> <span class=m>999</span><span class=p>)</span>    
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_pvalue</span> <span class=o>&lt;-</span> <span class=n>res_adonis[[1]][[</span><span class=s>&#34;Pr(&gt;F)&#34;</span><span class=n>]][1]</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_rsquared</span> <span class=o>&lt;-</span> <span class=nf>round</span><span class=p>(</span><span class=n>res_adonis[[1]][[</span><span class=s>&#34;R2&#34;</span><span class=n>]][1]</span><span class=p>,</span><span class=m>3</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1>#use the bquote function to format adonis results to be annotated on the ordination plot.</span>
</span></span><span class=line><span class=cl>  <span class=n>signi_label</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>cut</span><span class=p>(</span><span class=n>adn_pvalue</span><span class=p>,</span><span class=n>breaks</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=o>-</span><span class=kc>Inf</span><span class=p>,</span> <span class=m>0.001</span><span class=p>,</span> <span class=m>0.01</span><span class=p>,</span> <span class=m>0.05</span><span class=p>,</span> <span class=kc>Inf</span><span class=p>),</span> <span class=n>label</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;***&#34;</span><span class=p>,</span> <span class=s>&#34;**&#34;</span><span class=p>,</span> <span class=s>&#34;*&#34;</span><span class=p>,</span> <span class=s>&#34;.&#34;</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>  <span class=n>adn_res_format</span> <span class=o>&lt;-</span> <span class=nf>bquote</span><span class=p>(</span><span class=nf>atop</span><span class=p>(</span><span class=nf>atop</span><span class=p>(</span><span class=s>&#34;PERMANOVA&#34;</span><span class=p>,</span><span class=n>R^2</span><span class=o>==~</span><span class=n>.(adn_rsquared</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                                  <span class=nf>atop</span><span class=p>(</span><span class=s>&#34;p-value=&#34;</span><span class=o>~</span><span class=n>.(adn_pvalue</span><span class=p>)</span><span class=o>~</span><span class=n>.(signi_label</span><span class=p>),</span> <span class=nf>phantom</span><span class=p>())))</span>   
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>score</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>UMAP1</span><span class=p>,</span> <span class=n>y</span><span class=o>=</span><span class=n>UMAP2</span><span class=p>))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>geom_point</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>fill</span><span class=o>=</span><span class=n>Group</span><span class=p>),</span> <span class=n>size</span><span class=o>=</span><span class=m>2</span><span class=p>,</span> <span class=n>shape</span><span class=o>=</span><span class=m>21</span><span class=p>,</span> <span class=n>stroke</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>stat_ellipse</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=n>Group</span><span class=p>),</span> <span class=n>level</span> <span class=o>=</span> <span class=m>0.95</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>1.5</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>scale_color_manual</span><span class=p>(</span><span class=n>values</span> <span class=o>=</span> <span class=n>grp.col</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>scale_fill_manual</span><span class=p>(</span><span class=n>name</span> <span class=o>=</span> <span class=s>&#34;Condition&#34;</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                                <span class=n>values</span> <span class=o>=</span> <span class=n>grp.col</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span> <span class=n>x</span> <span class=o>=</span> <span class=nf>max</span><span class=p>(</span><span class=n>score</span><span class=o>$</span><span class=n>UMAP1</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                       <span class=n>y</span> <span class=o>=</span> <span class=nf>min</span><span class=p>(</span><span class=n>score</span><span class=o>$</span><span class=n>UMAP2</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                       <span class=n>label</span> <span class=o>=</span> <span class=n>adn_res_format</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>size</span> <span class=o>=</span> <span class=m>6</span><span class=p>)</span><span class=o>+</span> 
</span></span><span class=line><span class=cl>              <span class=nf>guides</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=s>&#34;none&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>theme_classic</span><span class=p>()</span><span class=o>+</span>
</span></span><span class=line><span class=cl>              <span class=nf>theme</span><span class=p>(</span><span class=n>axis.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>axis.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>strip.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span> 
</span></span><span class=line><span class=cl>                    <span class=n>panel.grid</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>11</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span> <span class=o>=</span> <span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.position</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.justification</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=n>legend.background</span> <span class=o>=</span> <span class=nf>element_rect</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>fill</span> <span class=o>=</span> <span class=s>&#34;white&#34;</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.5</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>UMAP_count</span> <span class=o>&lt;-</span> <span class=nf>UMAPFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_count</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>UMAP_FPKM</span> <span class=o>&lt;-</span> <span class=nf>UMAPFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_FPKM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>UMAP_TPM</span> <span class=o>&lt;-</span> <span class=nf>UMAPFun</span><span class=p>(</span><span class=n>dataset</span> <span class=o>=</span> <span class=n>ExprSet_TPM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>(</span><span class=n>UMAP_count</span> <span class=o>+</span> <span class=n>UMAP_FPKM</span> <span class=o>+</span> <span class=n>UMAP_TPM</span><span class=p>)</span> <span class=o>+</span> <span class=nf>plot_layout</span><span class=p>(</span><span class=n>nrow</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>guides</span> <span class=o>=</span> <span class=s>&#34;collect&#34;</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152018.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>总结</strong>：综合上述三种聚类分析方法，后两种基于非线性模型的方法具有较好的区分样本效果，而且TPM的区分效果貌似比Count和FPKM标准化方法更好。我后面应该度降维方法做一个原理上的总结。</p><h3 id=热图>热图</h3><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>heatFun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>){</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># datset=ExprSet_count</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pheno</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>grp</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>arrange</span><span class=p>(</span><span class=n>Group</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>edata</span> <span class=o>&lt;-</span> <span class=nf>exprs</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># scale data: z-score</span>
</span></span><span class=line><span class=cl>  <span class=n>scale_rows</span> <span class=o>&lt;-</span> <span class=nf>function </span><span class=p>(</span><span class=n>x</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>m</span> <span class=o>=</span> <span class=nf>apply</span><span class=p>(</span><span class=n>x</span><span class=p>,</span> <span class=m>1</span><span class=p>,</span> <span class=n>mean</span><span class=p>,</span> <span class=n>na.rm</span> <span class=o>=</span> <span class=bp>T</span><span class=p>)</span>
</span></span><span class=line><span class=cl>      <span class=n>s</span> <span class=o>=</span> <span class=nf>apply</span><span class=p>(</span><span class=n>x</span><span class=p>,</span> <span class=m>1</span><span class=p>,</span> <span class=n>sd</span><span class=p>,</span> <span class=n>na.rm</span> <span class=o>=</span> <span class=bp>T</span><span class=p>)</span>
</span></span><span class=line><span class=cl>      <span class=nf>return</span><span class=p>((</span><span class=n>x</span> <span class=o>-</span> <span class=n>m</span><span class=p>)</span><span class=o>/</span><span class=n>s</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>  
</span></span><span class=line><span class=cl>  <span class=n>edata_scaled</span> <span class=o>&lt;-</span> <span class=nf>t</span><span class=p>(</span><span class=nf>scale_rows</span><span class=p>(</span><span class=n>edata</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>circlize</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>col_fun</span> <span class=o>&lt;-</span> <span class=nf>colorRamp2</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=nf>round</span><span class=p>(</span><span class=nf>range</span><span class=p>(</span><span class=n>edata_scaled</span><span class=p>)</span><span class=n>[1]</span><span class=p>),</span> <span class=m>0</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                          <span class=nf>round</span><span class=p>(</span><span class=nf>range</span><span class=p>(</span><span class=n>edata_scaled</span><span class=p>)</span><span class=n>[2]</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                        <span class=nf>c</span><span class=p>(</span><span class=s>&#34;blue&#34;</span><span class=p>,</span> <span class=s>&#34;white&#34;</span><span class=p>,</span> <span class=s>&#34;red&#34;</span><span class=p>))</span> 
</span></span><span class=line><span class=cl>  <span class=c1># row split </span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status</span> <span class=o>&lt;-</span> <span class=nf>table</span><span class=p>(</span><span class=n>pheno</span><span class=o>$</span><span class=n>Group</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_number</span> <span class=o>&lt;-</span> <span class=nf>as.numeric</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_name</span> <span class=o>&lt;-</span> <span class=nf>names</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>row_split</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  <span class=nf>for </span><span class=p>(</span><span class=n>i</span> <span class=n>in</span> <span class=m>1</span><span class=o>:</span><span class=nf>length</span><span class=p>(</span><span class=n>dat_status_number</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>row_split</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>row_split</span><span class=p>,</span> <span class=nf>rep</span><span class=p>(</span><span class=n>i</span><span class=p>,</span> <span class=n>dat_status_number[i]</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  <span class=nf>require</span><span class=p>(</span><span class=n>ComplexHeatmap</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>Heatmap</span><span class=p>(</span>
</span></span><span class=line><span class=cl>          <span class=n>edata_scaled</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>          <span class=c1>#col = col_fun,</span>
</span></span><span class=line><span class=cl>          <span class=n>cluster_rows</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>          <span class=n>row_order</span> <span class=o>=</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>pheno</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>show_column_names</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>          <span class=n>show_row_names</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>          <span class=n>row_names_gp</span> <span class=o>=</span> <span class=nf>gpar</span><span class=p>(</span><span class=n>fontsize</span> <span class=o>=</span> <span class=m>12</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>row_names_side</span> <span class=o>=</span> <span class=s>&#34;right&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>          <span class=n>row_dend_side</span> <span class=o>=</span> <span class=s>&#34;left&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>          <span class=n>column_title</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>          <span class=n>heatmap_legend_param</span> <span class=o>=</span> <span class=nf>list</span><span class=p>(</span>
</span></span><span class=line><span class=cl>            <span class=n>title</span> <span class=o>=</span> <span class=s>&#34;Abundance&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>title_position</span> <span class=o>=</span> <span class=s>&#34;topcenter&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>border</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>legend_height</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>10</span><span class=p>,</span> <span class=s>&#34;cm&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>direction</span> <span class=o>=</span> <span class=s>&#34;horizontal&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>         <span class=n>row_split</span> <span class=o>=</span> <span class=n>row_split</span><span class=p>,</span>
</span></span><span class=line><span class=cl>        <span class=n>left_annotation</span> <span class=o>=</span> <span class=nf>rowAnnotation</span><span class=p>(</span><span class=n>foo</span> <span class=o>=</span> <span class=nf>anno_block</span><span class=p>(</span><span class=n>gp</span> <span class=o>=</span> <span class=nf>gpar</span><span class=p>(</span><span class=n>fill</span> <span class=o>=</span> <span class=m>2</span><span class=o>:</span><span class=m>4</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>labels</span> <span class=o>=</span> <span class=n>grp</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>            <span class=n>labels_gp</span> <span class=o>=</span> <span class=nf>gpar</span><span class=p>(</span><span class=n>col</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>fontsize</span> <span class=o>=</span> <span class=m>12</span><span class=p>))),</span>         
</span></span><span class=line><span class=cl>         <span class=n>column_km</span> <span class=o>=</span> <span class=m>2</span>
</span></span><span class=line><span class=cl>    <span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>Heat_count</span> <span class=o>&lt;-</span> <span class=nf>heatFun</span><span class=p>(</span><span class=n>datset</span> <span class=o>=</span> <span class=n>ExprSet_count</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Heat_FPKM</span> <span class=o>&lt;-</span> <span class=nf>heatFun</span><span class=p>(</span><span class=n>datset</span> <span class=o>=</span> <span class=n>ExprSet_FPKM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Heat_TPM</span> <span class=o>&lt;-</span> <span class=nf>heatFun</span><span class=p>(</span><span class=n>datset</span> <span class=o>=</span> <span class=n>ExprSet_TPM</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Heat_count</span> 
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152105.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Note</strong>: 19000+基因呈现出来的heatmap太大了，暂时展示counts矩阵的结果。从这图能看出还是有部分区块基因是存在富聚集的。</p><h2 id=deseq2>DESeq2</h2><p>DESeq2包输入的数据需要是counts矩阵，它使用负二项分布广义线性模型处理测序深度影响。该包的原理请参考文章<a href=https://zouhua.top/archives/b6151abf.html target=_blank rel=noopener>差异表达分析之Deseq2</a>。</p><h4 id=计算差异结果>计算差异结果</h4><ol><li><p><em>DESeqDataSetFromMatrix</em>构建<em>DESeq</em>函数所需要的包含count矩阵的数据对象；</p></li><li><p><em>DESeq</em>函数进行差异分析。</p></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>DESeq2Fun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>){</span>
</span></span><span class=line><span class=cl> 
</span></span><span class=line><span class=cl>  <span class=c1># datset=ExprSet_count</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>profile</span> <span class=o>&lt;-</span> <span class=nf>exprs</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>sid</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=nf>rownames</span><span class=p>(</span><span class=n>metadata</span><span class=p>),</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>profile</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>colData</span> <span class=o>&lt;-</span> <span class=n>metadata</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>SampleID</span><span class=p>,</span> <span class=n>Group</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>filter</span><span class=p>(</span><span class=n>SampleID</span><span class=o>%in%</span><span class=n>sid</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>grp</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>countData</span> <span class=o>&lt;-</span> <span class=n>profile</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=nf>all_of</span><span class=p>(</span><span class=n>sid</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>as.matrix</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if</span><span class=p>(</span><span class=o>!</span><span class=nf>all</span><span class=p>(</span><span class=nf>rownames</span><span class=p>(</span><span class=n>colData</span><span class=p>)</span> <span class=o>==</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>countData</span><span class=p>))){</span>
</span></span><span class=line><span class=cl>    <span class=nf>stop</span><span class=p>(</span><span class=s>&#34;Order is wrong please check your data&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>ddsm</span> <span class=o>&lt;-</span> <span class=nf>DESeqDataSetFromMatrix</span><span class=p>(</span><span class=n>countData</span> <span class=o>=</span> <span class=n>countData</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                                 <span class=n>colData</span> <span class=o>=</span> <span class=n>colData</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                                 <span class=n>design</span> <span class=o>=</span> <span class=o>~</span> <span class=n>Group</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>dds</span> <span class=o>&lt;-</span> <span class=nf>DESeq</span><span class=p>(</span><span class=n>ddsm</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>dds</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>DESeq_dds</span> <span class=o>&lt;-</span> <span class=nf>DESeq2Fun</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>DESeq_dds</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152139.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Notes</strong>: DESeq2_1.28.1更新了构建<em>DESeqDataSetFromMatrix</em>函数，原来需要将feature单独成一列，现在则不需要了，但需要保障colData的行名和countData的列名保持一致并且都是样本名字。</p><h3 id=提取结果>提取结果</h3><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>DESeq_result</span> <span class=o>&lt;-</span> <span class=nf>results</span><span class=p>(</span><span class=n>DESeq_dds</span><span class=p>,</span> <span class=n>contrast</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Group&#34;</span><span class=p>,</span> <span class=nf>rev</span><span class=p>(</span><span class=n>grp</span><span class=p>)))</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152210.png alt loading=lazy data-zoomable></div></div></figure></p><p>按照教程运行报错了，后来才发现是DESeq2_1.28.1版本的问题，更新到更高级的版本可解决该问题。这里因为是在Linux Rstudio server分析的，为了快捷安装我采用了conda模式。</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-bash data-lang=bash><span class=line><span class=cl>conda search DEseq2
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>conda install -c bioconda bioconductor-deseq2<span class=o>=</span>1.30.0 -y
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152234.png alt loading=lazy data-zoomable></div></div></figure></p><p>也可以通过BiocManager::install安装</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>remove.packages</span><span class=p>(</span><span class=s>&#34;DESeq2&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>BiocManager</span><span class=o>::</span><span class=nf>install</span><span class=p>(</span><span class=s>&#34;DESeq2&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>DESeq2</span><span class=p>)</span>
</span></span></code></pre></div><p>在安装完成后，我们重新运行上述R代码，提取出DESeq2的结果文件。</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>DESeq_dds</span> <span class=o>&lt;-</span> <span class=nf>DESeq2Fun</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152311.png alt loading=lazy data-zoomable></div></div></figure></p><ul><li>提取结果</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>DESeq_result</span> <span class=o>&lt;-</span> <span class=nf>results</span><span class=p>(</span><span class=n>DESeq_dds</span><span class=p>,</span> <span class=n>contrast</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Group&#34;</span><span class=p>,</span> <span class=nf>rev</span><span class=p>(</span><span class=n>grp</span><span class=p>)))</span>
</span></span><span class=line><span class=cl><span class=nf>head</span><span class=p>(</span><span class=n>DESeq_result</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152339.png alt loading=lazy data-zoomable></div></div></figure></p><h3 id=差异分组>差异分组</h3><p>设置差异基因过滤阈值（Foldchange+adjPval），上图设置的比较组顺序是Tumor vs Normal，所以<strong>log2foldchange > 0</strong>是Tumor组，反之则是Normal组。</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>DESeq_result_df</span> <span class=o>&lt;-</span> <span class=n>DESeq_result</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> <span class=nf>arrange</span><span class=p>(</span><span class=n>log2FoldChange</span><span class=p>,</span> <span class=n>padj</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>log2fc_threshold</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>DESeq_result_df</span><span class=p>,</span> <span class=nf>mean</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>log2FoldChange</span><span class=p>))</span> <span class=o>+</span> <span class=m>1.5</span> <span class=o>*</span> <span class=nf>sd</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>log2FoldChange</span><span class=p>)))</span>
</span></span><span class=line><span class=cl><span class=nf>message</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=s>&#34;Threshold of log2Foldchange [Mean+1.5(SD)] is&#34;</span><span class=p>,</span> <span class=n>log2fc_threshold</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=n>pval</span> <span class=o>&lt;-</span> <span class=m>0.05</span>
</span></span><span class=line><span class=cl><span class=n>DESeq_result_df</span><span class=nf>[which</span><span class=p>(</span><span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>log2FoldChange</span> <span class=o>&gt;=</span> <span class=n>log2fc_threshold</span> <span class=o>&amp;</span> <span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>padj</span> <span class=o>&lt;</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>grp[2]</span>
</span></span><span class=line><span class=cl><span class=n>DESeq_result_df</span><span class=nf>[which</span><span class=p>(</span><span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>log2FoldChange</span> <span class=o>&lt;=</span> <span class=o>-</span><span class=n>log2fc_threshold</span> <span class=o>&amp;</span> <span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>padj</span> <span class=o>&lt;</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>grp[1]</span>
</span></span><span class=line><span class=cl><span class=n>DESeq_result_df</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>log2FoldChange</span><span class=p>)</span> <span class=o>&lt;</span> <span class=n>log2fc_threshold</span> <span class=o>|</span> <span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>padj</span> <span class=o>&gt;=</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignf&#34;</span>
</span></span><span class=line><span class=cl><span class=nf>table</span><span class=p>(</span><span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>Enrichment</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152410.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>结果：差异显著的基因数目在Tumor和Normal组分别是552和850，可通过火山图展示差异基因。</strong></p><h3 id=volcano-plot>Volcano plot</h3><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>ggrepel</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>VolcanoFun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>DESeq_result_df</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>genelist</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;SMR3B&#34;</span><span class=p>,</span> <span class=s>&#34;BPIFA2&#34;</span><span class=p>,</span> <span class=s>&#34;HTN1&#34;</span><span class=p>,</span> <span class=s>&#34;NOBOX&#34;</span><span class=p>,</span> <span class=s>&#34;MAGEA9B&#34;</span><span class=p>,</span> <span class=s>&#34;MAGEA10&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                       <span class=n>group_name</span><span class=o>=</span><span class=n>grp</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>group_col</span><span class=o>=</span><span class=n>grp.col</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>Pval</span><span class=o>=</span><span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>LogFC</span><span class=o>=</span><span class=nf>round</span><span class=p>(</span><span class=n>log2fc_threshold</span><span class=p>,</span> <span class=m>2</span><span class=p>)){</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=c1># datset=DESeq_result_df</span>
</span></span><span class=line><span class=cl>  <span class=c1># genelist=c(&#34;SMR3B&#34;, &#34;BPIFA2&#34;, &#34;HTN1&#34;, &#34;NOBOX&#34;, &#34;MAGEA9B&#34;, &#34;MAGEA10&#34;)</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_name=grp</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_col=grp.col</span>
</span></span><span class=line><span class=cl>  <span class=c1># Pval=0.05</span>
</span></span><span class=line><span class=cl>  <span class=c1># LogFC=log2fc_threshold</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>dat</span> <span class=o>&lt;-</span> <span class=n>datset</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;FeatureID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Enrichment</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                        <span class=n>levels</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=n>group_name</span><span class=p>,</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>  <span class=c1># print(table(dat$color))</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status</span> <span class=o>&lt;-</span> <span class=nf>table</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>color</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_number</span> <span class=o>&lt;-</span> <span class=nf>as.numeric</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_name</span> <span class=o>&lt;-</span> <span class=nf>names</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>legend_label</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=n>dat_status_name[1]</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>dat_status_number[1]</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=nf>paste0</span><span class=p>(</span><span class=n>dat_status_name[2]</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>dat_status_number[2]</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>dat_status_number[3]</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>dat.signif</span> <span class=o>&lt;-</span> <span class=nf>subset</span><span class=p>(</span><span class=n>dat</span><span class=p>,</span> <span class=n>padj</span> <span class=o>&lt;</span> <span class=n>Pval</span> <span class=o>&amp;</span> <span class=nf>abs</span><span class=p>(</span><span class=n>log2FoldChange</span><span class=p>)</span> <span class=o>&gt;</span> <span class=n>LogFC</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>filter</span><span class=p>(</span><span class=n>FeatureID</span><span class=o>%in%</span><span class=n>genelist</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>print</span><span class=p>(</span><span class=nf>table</span><span class=p>(</span><span class=n>dat.signif</span><span class=o>$</span><span class=n>color</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>group_col_new</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=nf>rev</span><span class=p>(</span><span class=n>group_col</span><span class=p>),</span> <span class=s>&#34;grey80&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>group_name_new</span> <span class=o>&lt;-</span> <span class=nf>levels</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>color</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>xlabel</span> <span class=o>&lt;-</span> <span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;log2(&#34;</span><span class=p>,</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>rev</span><span class=p>(</span><span class=n>group_name</span><span class=p>),</span> <span class=n>collapse</span><span class=o>=</span><span class=s>&#34;/&#34;</span><span class=p>),</span> <span class=s>&#34;)&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=c1># Make a basic ggplot2 object with x-y values</span>
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>dat</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>log2FoldChange</span><span class=p>,</span> <span class=n>y</span><span class=o>=-</span><span class=nf>log10</span><span class=p>(</span><span class=n>padj</span><span class=p>),</span> <span class=n>color</span><span class=o>=</span><span class=n>color</span><span class=p>))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>geom_point</span><span class=p>(</span><span class=n>size</span><span class=o>=</span><span class=m>1</span><span class=p>,</span> <span class=n>alpha</span><span class=o>=</span><span class=m>1</span><span class=p>,</span> <span class=n>stroke</span><span class=o>=</span><span class=m>1</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>scale_color_manual</span><span class=p>(</span><span class=n>name</span><span class=o>=</span><span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=n>values</span><span class=o>=</span><span class=n>group_col_new</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=n>labels</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=n>legend_label</span><span class=p>,</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>xlab</span><span class=p>(</span><span class=n>xlabel</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>ylab</span><span class=p>(</span><span class=nf>expression</span><span class=p>(</span><span class=o>-</span><span class=n>log[10</span><span class=nf>]</span><span class=p>(</span><span class=s>&#34;adjusted p-value&#34;</span><span class=p>)))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>geom_hline</span><span class=p>(</span><span class=n>yintercept</span><span class=o>=-</span><span class=nf>log10</span><span class=p>(</span><span class=n>Pval</span><span class=p>),</span> <span class=n>alpha</span><span class=o>=</span><span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span><span class=o>=</span><span class=m>2</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>.7</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>geom_vline</span><span class=p>(</span><span class=n>xintercept</span><span class=o>=</span><span class=n>LogFC</span><span class=p>,</span> <span class=n>alpha</span><span class=o>=</span><span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span><span class=o>=</span><span class=m>2</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>.7</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>geom_vline</span><span class=p>(</span><span class=n>xintercept</span><span class=o>=-</span><span class=n>LogFC</span><span class=p>,</span> <span class=n>alpha</span><span class=o>=</span><span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span><span class=o>=</span><span class=m>2</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>.7</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>geom_text_repel</span><span class=p>(</span><span class=n>data</span> <span class=o>=</span> <span class=n>dat.signif</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                          <span class=nf>aes</span><span class=p>(</span><span class=n>label</span> <span class=o>=</span> <span class=n>FeatureID</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                          <span class=n>size</span> <span class=o>=</span> <span class=m>4</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                          <span class=n>max.overlaps</span> <span class=o>=</span> <span class=nf>getOption</span><span class=p>(</span><span class=s>&#34;ggrepel.max.overlaps&#34;</span><span class=p>,</span> <span class=n>default</span> <span class=o>=</span> <span class=m>80</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                          <span class=n>segment.linetype</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                          <span class=n>segment.curvature</span> <span class=o>=</span> <span class=m>-1e-20</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                          <span class=n>box.padding</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.35</span><span class=p>,</span> <span class=s>&#34;lines&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                          <span class=n>point.padding</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.3</span><span class=p>,</span> <span class=s>&#34;lines&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                          <span class=n>arrow</span> <span class=o>=</span> <span class=nf>arrow</span><span class=p>(</span><span class=n>length</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.005</span><span class=p>,</span> <span class=s>&#34;npc&#34;</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                          <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span>     <span class=c1># text color</span>
</span></span><span class=line><span class=cl>                          <span class=n>bg.color</span> <span class=o>=</span> <span class=s>&#34;white&#34;</span><span class=p>,</span> <span class=c1># shadow color</span>
</span></span><span class=line><span class=cl>                          <span class=n>bg.r</span> <span class=o>=</span> <span class=m>0.15</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span> <span class=n>x</span><span class=o>=</span><span class=nf>min</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>log2FoldChange</span><span class=p>),</span> <span class=n>y</span><span class=o>=-</span><span class=nf>log10</span><span class=p>(</span><span class=n>Pval</span><span class=p>),</span> <span class=n>label</span><span class=o>=</span><span class=n>Pval</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>6</span><span class=p>,</span> <span class=n>color</span><span class=o>=</span><span class=s>&#34;red&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span> <span class=n>x</span><span class=o>=</span><span class=n>LogFC</span><span class=p>,</span> <span class=n>y</span><span class=o>=</span><span class=m>0</span><span class=p>,</span> <span class=n>label</span><span class=o>=</span><span class=n>LogFC</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>6</span><span class=p>,</span> <span class=n>color</span><span class=o>=</span><span class=s>&#34;red&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span> <span class=n>x</span><span class=o>=-</span><span class=n>LogFC</span><span class=p>,</span> <span class=n>y</span><span class=o>=</span><span class=m>0</span><span class=p>,</span> <span class=n>label</span><span class=o>=-</span><span class=n>LogFC</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>6</span><span class=p>,</span> <span class=n>color</span><span class=o>=</span><span class=s>&#34;red&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>scale_y_continuous</span><span class=p>(</span><span class=n>trans</span> <span class=o>=</span> <span class=s>&#34;log1p&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>guides</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=nf>guide_legend</span><span class=p>(</span><span class=n>override.aes</span> <span class=o>=</span> <span class=nf>list</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>3</span><span class=p>)))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>theme_bw</span><span class=p>()</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>theme</span><span class=p>(</span><span class=n>axis.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>12</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>axis.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span><span class=o>=</span><span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>panel.grid</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>                <span class=c1>#legend.position = &#34;right&#34;,</span>
</span></span><span class=line><span class=cl>                <span class=n>legend.position</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>.15</span><span class=p>,</span> <span class=m>.1</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>legend.key.height</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.6</span><span class=p>,</span><span class=s>&#34;cm&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>legend.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>strip.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>14</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>VolcanoFun</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>DESeq_result_df</span><span class=p>,</span> <span class=n>genelist</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;SMR3B&#34;</span><span class=p>,</span> <span class=s>&#34;BPIFA2&#34;</span><span class=p>,</span> <span class=s>&#34;HTN1&#34;</span><span class=p>,</span> <span class=s>&#34;NOBOX&#34;</span><span class=p>,</span> <span class=s>&#34;MAGEA9B&#34;</span><span class=p>,</span> <span class=s>&#34;MAGEA10&#34;</span><span class=p>))</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152506.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Notes</strong>： <strong>log2Foldchange > 0</strong>是富集在Tumor组，图上表示<em>NOBOX, MAGEA9B, MAGEA10</em> 是富集在Tumor组，也可以从公式log2(Tumor/Normal)看出。为了验证上述结论，使用<strong>boxplot</strong>图展示结果。</p><h3 id=boxplot>boxplot</h3><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>BoxplotFun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>genelist</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;SMR3B&#34;</span><span class=p>,</span> <span class=s>&#34;BPIFA2&#34;</span><span class=p>,</span> <span class=s>&#34;HTN1&#34;</span><span class=p>,</span> <span class=s>&#34;NOBOX&#34;</span><span class=p>,</span> <span class=s>&#34;MAGEA9B&#34;</span><span class=p>,</span> <span class=s>&#34;MAGEA10&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                       <span class=n>group_name</span><span class=o>=</span><span class=n>grp</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>group_col</span><span class=o>=</span><span class=n>grp.col</span><span class=p>){</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=c1># datset=ExprSet_count</span>
</span></span><span class=line><span class=cl>  <span class=c1># genelist=c(&#34;SMR3B&#34;, &#34;BPIFA2&#34;, &#34;HTN1&#34;, &#34;NOBOX&#34;, &#34;MAGEA9B&#34;, &#34;MAGEA10&#34;)</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_name=grp</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_col=grp.col</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>pheno</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=nf>filter</span><span class=p>(</span><span class=n>Group</span><span class=o>%in%</span><span class=n>group_name</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>group_name</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=nf>print</span><span class=p>(</span><span class=nf>table</span><span class=p>(</span><span class=n>pheno</span><span class=o>$</span><span class=n>Group</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>edata</span> <span class=o>&lt;-</span> <span class=nf>data.frame</span><span class=p>(</span><span class=nf>exprs</span><span class=p>(</span><span class=n>datset</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>            <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=nf>rownames</span><span class=p>(</span><span class=n>pheno</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>            <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;FeatureID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>            <span class=nf>filter</span><span class=p>(</span><span class=n>FeatureID</span><span class=o>%in%</span><span class=n>genelist</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>            <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;FeatureID&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>mdat</span> <span class=o>&lt;-</span> <span class=n>pheno</span> <span class=o>%&gt;%</span> <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>Group</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>inner_join</span><span class=p>(</span><span class=nf>t</span><span class=p>(</span><span class=n>edata</span><span class=p>)</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>),</span> <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>plotdata</span> <span class=o>&lt;-</span> <span class=n>mdat</span> <span class=o>%&gt;%</span> <span class=n>tidyr</span><span class=o>::</span><span class=nf>gather</span><span class=p>(</span><span class=n>key</span><span class=o>=</span><span class=s>&#34;FeatureID&#34;</span><span class=p>,</span> <span class=n>value</span><span class=o>=</span><span class=s>&#34;value&#34;</span><span class=p>,</span> <span class=o>-</span><span class=n>Group</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>group_name</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>plotdata</span><span class=o>$</span><span class=n>FeatureID</span> <span class=o>&lt;-</span> <span class=nf>factor</span><span class=p>(</span><span class=n>plotdata</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>genelist</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>plotdata</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=n>Group</span><span class=p>,</span> <span class=n>y</span><span class=o>=</span><span class=n>value</span><span class=p>,</span> <span class=n>fill</span><span class=o>=</span><span class=n>Group</span><span class=p>))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>stat_boxplot</span><span class=p>(</span><span class=n>geom</span><span class=o>=</span><span class=s>&#34;errorbar&#34;</span><span class=p>,</span> <span class=n>width</span><span class=o>=</span><span class=m>0.15</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>position</span><span class=o>=</span><span class=nf>position_dodge</span><span class=p>(</span><span class=m>0.4</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>geom_boxplot</span><span class=p>(</span><span class=n>width</span><span class=o>=</span><span class=m>0.4</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>outlier.colour</span><span class=o>=</span><span class=s>&#34;black&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>outlier.shape</span><span class=o>=</span><span class=m>21</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>outlier.size</span><span class=o>=</span><span class=m>.5</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>scale_fill_manual</span><span class=p>(</span><span class=n>values</span><span class=o>=</span><span class=n>group_col</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=c1>#ggpubr::stat_compare_means(method = &#34;wilcox.test&#34;, comparisons = group_name)+</span>
</span></span><span class=line><span class=cl>          <span class=nf>facet_wrap</span><span class=p>(</span><span class=n>facets</span><span class=o>=</span><span class=s>&#34;FeatureID&#34;</span><span class=p>,</span> <span class=n>scales</span><span class=o>=</span><span class=s>&#34;free_y&#34;</span><span class=p>,</span> <span class=n>nrow</span><span class=o>=</span><span class=m>2</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>labs</span><span class=p>(</span><span class=n>x</span><span class=o>=</span><span class=s>&#34;&#34;</span><span class=p>,</span> <span class=n>y</span><span class=o>=</span><span class=s>&#34;Gene Count&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>guides</span><span class=p>(</span><span class=n>fill</span><span class=o>=</span><span class=s>&#34;none&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>theme_bw</span><span class=p>()</span><span class=o>+</span>
</span></span><span class=line><span class=cl>          <span class=nf>theme</span><span class=p>(</span><span class=n>axis.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>12</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>axis.text.x</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>color</span><span class=o>=</span><span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>10</span><span class=p>,</span> <span class=n>hjust</span><span class=o>=</span><span class=m>.5</span><span class=p>,</span> <span class=n>vjust</span><span class=o>=</span><span class=m>.5</span><span class=p>,</span> <span class=n>angle</span><span class=o>=</span><span class=m>60</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span><span class=o>=</span><span class=m>8</span><span class=p>,</span> <span class=n>color</span><span class=o>=</span><span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>family</span><span class=o>=</span><span class=s>&#34;serif&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                <span class=n>panel.grid</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>                <span class=n>strip.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>face</span><span class=o>=</span><span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>size</span><span class=o>=</span><span class=m>12</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>BoxplotFun</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>,</span>
</span></span><span class=line><span class=cl>           <span class=n>genelist</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;SMR3B&#34;</span><span class=p>,</span> <span class=s>&#34;BPIFA2&#34;</span><span class=p>,</span> <span class=s>&#34;HTN1&#34;</span><span class=p>,</span> <span class=s>&#34;NOBOX&#34;</span><span class=p>,</span> <span class=s>&#34;MAGEA9B&#34;</span><span class=p>,</span> <span class=s>&#34;MAGEA10&#34;</span><span class=p>))</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152535.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Notes</strong>： 从图中可以看出，最显著富集的基因在另一组的表达值(counts数目)几乎都为0，这说明这些基因是某组独有的基因（需要进一步查看出现率判断是否是独有还是仅仅低丰度而已，因为Tumor组样本数目要远远大于Normal组），而DESeq2通过标准化因子能区分出来，而用wilcox-test检验会出现<em>not enough observations</em>的警告信息，因此注释掉<em>stat_compare_means</em>函数。</p><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152556.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>总结：表达谱矩阵是count matrix的时，可以使用DESeq2包做假设检验（差异分析）。以前做扩增子分析的时候，OTU table也是count matrix，其实也可以用该包做假设检验，但是如果转换成Relative abundance则不可。除此之外，输入参数design还可以添加协变量进行校正处理，后面有时间再写另一篇专门介绍怎么用DESeq2(limma/edgeR)校正协变量的博客吧。</strong></p><h2 id=limma>limma</h2><p>limma最初是用于微阵列芯片基因表达的差异分析，但后来它提供的<em>voom</em>函数使其可以应用于转录组等差异分析（<strong>limma-voom</strong>模型），那我们肯定好奇<em>voom</em>函数的原理是什么。</p><h3 id=voom功能>Voom功能</h3><p>这里要着重介绍下voom函数的作用：</p><ol><li><p>标准化counts成CPM：log2 transform；</p></li><li><p>计算每个基因的log2CPM的线性模型残差（residuals = observed - fitted）；</p></li><li><p>通过平均表达值拟合残差标准差（residual standard deviation）；</p></li><li><p>结合拟合平滑曲线获得的每个基因的权重和log2CPM值用于后续差异分析。</p></li></ol><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152709.png alt loading=lazy data-zoomable></div></div></figure></p><h3 id=计算过程>计算过程</h3><ol><li><p>构建分组矩阵；</p></li><li><p>构建DGEList对象；</p></li><li><p>计算Counts标准化因子；</p></li><li><p>voom标准化；</p></li><li><p>线性模型计算每个基因在分组的weighted least square；</p></li><li><p>构建比较对象；</p></li><li><p>计算每个基因在比较对象间的比较结果；</p></li><li><p>在基因的平均标准误基础上，使用经典贝叶斯算法缩小基因组间比较结果的最大最小标准误差；</p></li><li><p>提取最终差异结果。</p></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>limma</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>Limma2Fun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>){</span>
</span></span><span class=line><span class=cl> 
</span></span><span class=line><span class=cl>  <span class=c1># datset=ExprSet_count</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>profile</span> <span class=o>&lt;-</span> <span class=nf>exprs</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>sid</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=nf>rownames</span><span class=p>(</span><span class=n>metadata</span><span class=p>),</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>profile</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>colData</span> <span class=o>&lt;-</span> <span class=n>metadata</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>SampleID</span><span class=p>,</span> <span class=n>Group</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>filter</span><span class=p>(</span><span class=n>SampleID</span><span class=o>%in%</span><span class=n>sid</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>grp</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>countData</span> <span class=o>&lt;-</span> <span class=n>profile</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=nf>all_of</span><span class=p>(</span><span class=n>sid</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>as.matrix</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if</span><span class=p>(</span><span class=o>!</span><span class=nf>all</span><span class=p>(</span><span class=nf>rownames</span><span class=p>(</span><span class=n>colData</span><span class=p>)</span> <span class=o>==</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>countData</span><span class=p>))){</span>
</span></span><span class=line><span class=cl>    <span class=nf>stop</span><span class=p>(</span><span class=s>&#34;Order is wrong please check your data&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group matrix </span>
</span></span><span class=line><span class=cl>  <span class=n>design</span> <span class=o>&lt;-</span> <span class=nf>model.matrix</span><span class=p>(</span><span class=o>~</span><span class=m>0</span> <span class=o>+</span> <span class=n>Group</span><span class=p>,</span> <span class=n>data</span> <span class=o>=</span> <span class=n>colData</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>design</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>levels</span><span class=p>(</span><span class=n>colData</span><span class=o>$</span><span class=n>Group</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>rownames</span><span class=p>(</span><span class=n>design</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>countData</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># DGEList object: counts-&gt; colnames:Samples; rownames: Features</span>
</span></span><span class=line><span class=cl>  <span class=n>dge_obj</span> <span class=o>&lt;-</span> <span class=n>edgeR</span><span class=o>::</span><span class=nf>DGEList</span><span class=p>(</span><span class=n>counts</span> <span class=o>=</span> <span class=n>countData</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dge_obj_factors</span> <span class=o>&lt;-</span> <span class=n>edgeR</span><span class=o>::</span><span class=nf>calcNormFactors</span><span class=p>(</span><span class=n>dge_obj</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># voom </span>
</span></span><span class=line><span class=cl>  <span class=n>dat_voom</span> <span class=o>&lt;-</span> <span class=nf>voom</span><span class=p>(</span><span class=n>counts</span> <span class=o>=</span> <span class=n>dge_obj_factors</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                   <span class=n>design</span> <span class=o>=</span> <span class=n>design</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                   <span class=n>normalize.method</span> <span class=o>=</span> <span class=s>&#34;quantile&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                   <span class=n>plot</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># Linear model to calculate weighted least squares per gene in each group</span>
</span></span><span class=line><span class=cl>  <span class=n>fit</span> <span class=o>&lt;-</span> <span class=nf>lmFit</span><span class=p>(</span><span class=n>object</span> <span class=o>=</span> <span class=n>dat_voom</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>               <span class=n>design</span> <span class=o>=</span> <span class=n>design</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># Comparison between two groups</span>
</span></span><span class=line><span class=cl>  <span class=n>contr</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>rev</span><span class=p>(</span><span class=nf>levels</span><span class=p>(</span><span class=n>colData</span><span class=o>$</span><span class=n>Group</span><span class=p>)),</span> <span class=n>collapse</span> <span class=o>=</span> <span class=s>&#34;-&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>contr.matrix</span> <span class=o>&lt;-</span> <span class=nf>makeContrasts</span><span class=p>(</span><span class=n>contrasts</span> <span class=o>=</span> <span class=n>contr</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                                <span class=n>levels</span> <span class=o>=</span> <span class=n>design</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1># Estimate contrast for each gene</span>
</span></span><span class=line><span class=cl>  <span class=n>fit.contr</span> <span class=o>&lt;-</span> <span class=nf>contrasts.fit</span><span class=p>(</span><span class=n>fit</span><span class=p>,</span> <span class=n>contr.matrix</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># Empirical Bayes smoothing of standard errors</span>
</span></span><span class=line><span class=cl>  <span class=n>fit.ebay</span> <span class=o>&lt;-</span> <span class=nf>eBayes</span><span class=p>(</span><span class=n>fit.contr</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># DE result table</span>
</span></span><span class=line><span class=cl>  <span class=n>res_table</span> <span class=o>&lt;-</span> <span class=nf>topTable</span><span class=p>(</span><span class=n>fit.ebay</span><span class=p>,</span> <span class=n>coef</span> <span class=o>=</span> <span class=n>contr</span><span class=p>,</span> <span class=n>n</span><span class=o>=</span><span class=kc>Inf</span><span class=p>,</span> <span class=n>sort.by</span> <span class=o>=</span> <span class=s>&#34;P&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>res_table</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>Limma_res</span> <span class=o>&lt;-</span> <span class=nf>Limma2Fun</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>head</span><span class=p>(</span><span class=n>Limma_res</span><span class=p>)</span>
</span></span></code></pre></div><p>查看voom拟合曲线结果：曲线拟合不够平滑，说明需要设置counts或occurrence过滤阈值过滤离群的基因</p><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152804.png alt loading=lazy data-zoomable></div></div></figure></p><p>limma差异结果：</p><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152920.png alt loading=lazy data-zoomable></div></div></figure></p><h3 id=差异分组-1>差异分组</h3><p>设置差异基因过滤阈值（logFC+adj.P.Val）</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>Limma_res_df</span> <span class=o>&lt;-</span> <span class=n>Limma_res</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> <span class=nf>arrange</span><span class=p>(</span><span class=n>logFC</span><span class=p>,</span> <span class=n>adj.P.Val</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>log2fc_threshold</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>Limma_res_df</span><span class=p>,</span> <span class=nf>mean</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>logFC</span><span class=p>))</span> <span class=o>+</span> <span class=m>1.5</span> <span class=o>*</span> <span class=nf>sd</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>logFC</span><span class=p>)))</span>
</span></span><span class=line><span class=cl><span class=nf>message</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=s>&#34;Threshold of log2Foldchange [Mean+1.5(SD)] is&#34;</span><span class=p>,</span> <span class=n>log2fc_threshold</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=n>pval</span> <span class=o>&lt;-</span> <span class=m>0.05</span>
</span></span><span class=line><span class=cl><span class=n>Limma_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=n>Limma_res_df</span><span class=o>$</span><span class=n>logFC</span> <span class=o>&gt;=</span> <span class=n>log2fc_threshold</span> <span class=o>&amp;</span> <span class=n>Limma_res_df</span><span class=o>$</span><span class=n>adj.P.Val</span> <span class=o>&lt;</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>grp[2]</span>
</span></span><span class=line><span class=cl><span class=n>Limma_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=n>Limma_res_df</span><span class=o>$</span><span class=n>logFC</span> <span class=o>&lt;=</span> <span class=o>-</span><span class=n>log2fc_threshold</span> <span class=o>&amp;</span> <span class=n>Limma_res_df</span><span class=o>$</span><span class=n>adj.P.Val</span> <span class=o>&lt;</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>grp[1]</span>
</span></span><span class=line><span class=cl><span class=n>Limma_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>Limma_res_df</span><span class=o>$</span><span class=n>logFC</span><span class=p>)</span> <span class=o>&lt;</span> <span class=n>log2fc_threshold</span> <span class=o>|</span> <span class=n>Limma_res_df</span><span class=o>$</span><span class=n>adj.P.Val</span> <span class=o>&gt;=</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignf&#34;</span>
</span></span><span class=line><span class=cl><span class=nf>table</span><span class=p>(</span><span class=n>Limma_res_df</span><span class=o>$</span><span class=n>Enrichment</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126152948.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>结果：差异显著的基因数目在Tumor和Normal组分别是583和861，火山图展示差异基因如上图修改代码即可。</strong></p><h2 id=edger>edgeR</h2><p>edgeR包也是基于线性模型做差异分析的包。</p><h3 id=计算过程-1>计算过程</h3><ol><li><p>先存储counts成DGEList对象；</p></li><li><p>过滤数据（可选择）；</p></li><li><p>标准化counts矩阵；</p></li><li><p>计算离散程度（整体可变性评估estimateGLMCommonDisp；趋势模型estimateGLMTrendedDisp；比较组estimateGLMTagwiseDisp）；</p></li><li><p>GLM 模型的likelihood ratio test进行差异分析。</p></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>edgeR</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>EdgeR2Fun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>){</span>
</span></span><span class=line><span class=cl> 
</span></span><span class=line><span class=cl>  <span class=c1># datset=ExprSet_count</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>profile</span> <span class=o>&lt;-</span> <span class=nf>exprs</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>datset</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>sid</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=nf>rownames</span><span class=p>(</span><span class=n>metadata</span><span class=p>),</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>profile</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>colData</span> <span class=o>&lt;-</span> <span class=n>metadata</span> <span class=o>%&gt;%</span> <span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>SampleID</span><span class=p>,</span> <span class=n>Group</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>filter</span><span class=p>(</span><span class=n>SampleID</span><span class=o>%in%</span><span class=n>sid</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>mutate</span><span class=p>(</span><span class=n>Group</span><span class=o>=</span><span class=nf>factor</span><span class=p>(</span><span class=n>Group</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>grp</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>countData</span> <span class=o>&lt;-</span> <span class=n>profile</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=nf>all_of</span><span class=p>(</span><span class=n>sid</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>as.matrix</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if</span><span class=p>(</span><span class=o>!</span><span class=nf>all</span><span class=p>(</span><span class=nf>rownames</span><span class=p>(</span><span class=n>colData</span><span class=p>)</span> <span class=o>==</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>countData</span><span class=p>))){</span>
</span></span><span class=line><span class=cl>    <span class=nf>stop</span><span class=p>(</span><span class=s>&#34;Order is wrong please check your data&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group matrix </span>
</span></span><span class=line><span class=cl>  <span class=n>design</span> <span class=o>&lt;-</span> <span class=nf>model.matrix</span><span class=p>(</span><span class=o>~</span><span class=m>0</span> <span class=o>+</span> <span class=n>Group</span><span class=p>,</span> <span class=n>data</span> <span class=o>=</span> <span class=n>colData</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>design</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>levels</span><span class=p>(</span><span class=n>colData</span><span class=o>$</span><span class=n>Group</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>rownames</span><span class=p>(</span><span class=n>design</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>countData</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=c1># Filter data </span>
</span></span><span class=line><span class=cl>  <span class=c1># keep &lt;- rowSums(cpm(countData) &gt; 100) &gt;= 2</span>
</span></span><span class=line><span class=cl>  <span class=c1># countData &lt;- countData[keep, ]  </span>
</span></span><span class=line><span class=cl>    
</span></span><span class=line><span class=cl>  <span class=c1># DGEList object: counts-&gt; colnames:Samples; rownames: Features</span>
</span></span><span class=line><span class=cl>  <span class=n>dge_obj</span> <span class=o>&lt;-</span> <span class=nf>DGEList</span><span class=p>(</span><span class=n>counts</span> <span class=o>=</span> <span class=n>countData</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dge_obj_factors</span> <span class=o>&lt;-</span> <span class=nf>calcNormFactors</span><span class=p>(</span><span class=n>dge_obj</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>plotMDS</span><span class=p>(</span><span class=n>dge_obj_factors</span><span class=p>,</span> <span class=n>method</span><span class=o>=</span><span class=s>&#34;bcv&#34;</span><span class=p>,</span> <span class=n>col</span><span class=o>=</span><span class=nf>as.numeric</span><span class=p>(</span><span class=n>colData</span><span class=o>$</span><span class=n>Group</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>legend</span><span class=p>(</span><span class=s>&#34;bottomleft&#34;</span><span class=p>,</span> <span class=nf>as.character</span><span class=p>(</span><span class=nf>unique</span><span class=p>(</span><span class=n>colData</span><span class=o>$</span><span class=n>Group</span><span class=p>)),</span> <span class=n>col</span><span class=o>=</span><span class=m>1</span><span class=o>:</span><span class=m>2</span><span class=p>,</span> <span class=n>pch</span><span class=o>=</span><span class=m>20</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># GLM estimates of dispersion</span>
</span></span><span class=line><span class=cl>  <span class=n>dge</span> <span class=o>&lt;-</span> <span class=nf>estimateGLMCommonDisp</span><span class=p>(</span><span class=n>dge_obj_factors</span><span class=p>,</span> <span class=n>design</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dge</span> <span class=o>&lt;-</span> <span class=nf>estimateGLMTrendedDisp</span><span class=p>(</span><span class=n>dge</span><span class=p>,</span> <span class=n>design</span><span class=p>,</span> <span class=n>method</span><span class=o>=</span><span class=s>&#34;power&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1># You can change method to &#34;auto&#34;, &#34;bin.spline&#34;, &#34;power&#34;, &#34;spline&#34;, &#34;bin.loess&#34;.</span>
</span></span><span class=line><span class=cl>  <span class=c1># The default is &#34;auto&#34; which chooses &#34;bin.spline&#34; when &gt; 200 tags and &#34;power&#34; otherwise.</span>
</span></span><span class=line><span class=cl>  <span class=n>dge</span> <span class=o>&lt;-</span> <span class=nf>estimateGLMTagwiseDisp</span><span class=p>(</span><span class=n>dge</span><span class=p>,</span> <span class=n>design</span><span class=p>)</span>  
</span></span><span class=line><span class=cl>  <span class=nf>plotBCV</span><span class=p>(</span><span class=n>dge</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># GLM testing for differential expression: </span>
</span></span><span class=line><span class=cl>  <span class=n>fit</span> <span class=o>&lt;-</span> <span class=nf>glmFit</span><span class=p>(</span><span class=n>dge</span><span class=p>,</span> <span class=n>design</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>fit2</span> <span class=o>&lt;-</span> <span class=nf>glmLRT</span><span class=p>(</span><span class=n>fit</span><span class=p>,</span> <span class=n>contrast</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=m>-1</span><span class=p>,</span> <span class=m>1</span><span class=p>))</span> <span class=c1># Normal:-1; Tumor:1</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>res_table</span> <span class=o>&lt;-</span> <span class=nf>topTags</span><span class=p>(</span><span class=n>fit2</span><span class=p>,</span> <span class=n>n</span><span class=o>=</span><span class=nf>nrow</span><span class=p>(</span><span class=n>countData</span><span class=p>))</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>res_table</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>EdgeR_res</span> <span class=o>&lt;-</span> <span class=nf>EdgeR2Fun</span><span class=p>(</span><span class=n>datset</span><span class=o>=</span><span class=n>ExprSet_count</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>head</span><span class=p>(</span><span class=n>EdgeR_res</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153035.png alt loading=lazy data-zoomable></div></div></figure></p><h3 id=差异分组-2>差异分组</h3><p>设置差异基因过滤阈值（logFC+FDR）</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>EdgeR_res_df</span> <span class=o>&lt;-</span> <span class=n>EdgeR_res</span> <span class=o>%&gt;%</span> <span class=nf>arrange</span><span class=p>(</span><span class=n>logFC</span><span class=p>,</span> <span class=n>FDR</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>log2fc_threshold</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>EdgeR_res_df</span><span class=p>,</span> <span class=nf>mean</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>logFC</span><span class=p>))</span> <span class=o>+</span> <span class=m>1.5</span> <span class=o>*</span> <span class=nf>sd</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>logFC</span><span class=p>)))</span>
</span></span><span class=line><span class=cl><span class=nf>message</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=s>&#34;Threshold of log2Foldchange [Mean+1.5(SD)] is&#34;</span><span class=p>,</span> <span class=n>log2fc_threshold</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=n>pval</span> <span class=o>&lt;-</span> <span class=m>0.05</span>
</span></span><span class=line><span class=cl><span class=n>EdgeR_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>logFC</span> <span class=o>&gt;=</span> <span class=n>log2fc_threshold</span> <span class=o>&amp;</span> <span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>FDR</span> <span class=o>&lt;</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>grp[2]</span>
</span></span><span class=line><span class=cl><span class=n>EdgeR_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>logFC</span> <span class=o>&lt;=</span> <span class=o>-</span><span class=n>log2fc_threshold</span> <span class=o>&amp;</span> <span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>FDR</span> <span class=o>&lt;</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>grp[1]</span>
</span></span><span class=line><span class=cl><span class=n>EdgeR_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>logFC</span><span class=p>)</span> <span class=o>&lt;</span> <span class=n>log2fc_threshold</span> <span class=o>|</span> <span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>FDR</span> <span class=o>&gt;=</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignf&#34;</span>
</span></span><span class=line><span class=cl><span class=nf>table</span><span class=p>(</span><span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>Enrichment</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153103.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>结果：差异显著的基因数目在Tumor和Normal组分别是535和842，火山图展示差异基因如上图修改代码即可。</strong></p><h2 id=wilcox-rank-sum-test-or-t-test>Wilcox-rank-sum test or t-test</h2><p>判断基因的表达值是否符合正态分布，是则选择t-test，否则选择wilcox-rank-sum test，log2foldchange根据scale结果的两组均值相除获取。</p><h3 id=计算差异结果-1>计算差异结果</h3><div class=highlight><pre tabindex=0 class=chroma><code class=language-fallback data-lang=fallback><span class=line><span class=cl>DEA_test &lt;- function(datset=ExprSet_count,
</span></span><span class=line><span class=cl>                     group_info=&#34;Group&#34;,
</span></span><span class=line><span class=cl>                     group_name=grp){
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  # datset=ExprSet
</span></span><span class=line><span class=cl>  # group_info=&#34;Group&#34;
</span></span><span class=line><span class=cl>  # group_name=grp
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  phenotype &lt;- pData(datset)
</span></span><span class=line><span class=cl>  profile &lt;- exprs(datset)
</span></span><span class=line><span class=cl>  phenotype_cln &lt;- phenotype %&gt;% dplyr::select(all_of(group_info))
</span></span><span class=line><span class=cl>  colnames(phenotype_cln)[which(colnames(phenotype_cln) == group_info)] &lt;- &#34;Group&#34;
</span></span><span class=line><span class=cl>  pheno &lt;- phenotype_cln %&gt;%
</span></span><span class=line><span class=cl>      rownames_to_column(&#34;SampleID&#34;) %&gt;%
</span></span><span class=line><span class=cl>      filter(Group%in%group_name) %&gt;%
</span></span><span class=line><span class=cl>      mutate(Group=factor(Group, levels = group_name)) %&gt;%
</span></span><span class=line><span class=cl>      column_to_rownames(&#34;SampleID&#34;)
</span></span><span class=line><span class=cl>  print(table(pheno$Group))
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  edata &lt;- profile[, rownames(pheno)]
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  if(length(unique(pheno$Group)) != 2){
</span></span><span class=line><span class=cl>    stop(&#34;Levels of Group must be 2 levels&#34;)
</span></span><span class=line><span class=cl>  }
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  # Checking whether the two groups that where tested were normally distributed using the Shapiro-Wilk test
</span></span><span class=line><span class=cl>  shapiro_res &lt;- apply(edata, 1, function(x, y){
</span></span><span class=line><span class=cl>    #dat &lt;- data.frame(value=as.numeric(edata[1, ]), group=pheno$Group)
</span></span><span class=line><span class=cl>    dat &lt;- data.frame(value=x, group=y)
</span></span><span class=line><span class=cl>    
</span></span><span class=line><span class=cl>    nUnique &lt;- tapply(dat$value, dat$group, function(x){length(unique(x))})
</span></span><span class=line><span class=cl>    if(nUnique[1] &lt; 5 | nUnique[2] &lt; 5){
</span></span><span class=line><span class=cl>      res &lt;- NA
</span></span><span class=line><span class=cl>    }else{
</span></span><span class=line><span class=cl>      t_res &lt;- tapply(dat$value, dat$group, shapiro.test)
</span></span><span class=line><span class=cl>      if(t_res[[1]]$p.value &lt; 0.05){
</span></span><span class=line><span class=cl>        res &lt;- FALSE
</span></span><span class=line><span class=cl>      }else{
</span></span><span class=line><span class=cl>        res &lt;- TRUE
</span></span><span class=line><span class=cl>      }      
</span></span><span class=line><span class=cl>    }
</span></span><span class=line><span class=cl>    return(res)
</span></span><span class=line><span class=cl>  }, pheno$Group) %&gt;%
</span></span><span class=line><span class=cl>    data.frame() %&gt;%
</span></span><span class=line><span class=cl>    setNames(&#34;Normal&#34;) %&gt;%
</span></span><span class=line><span class=cl>    rownames_to_column(&#34;FeatureID&#34;)
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  shapiro_res_final &lt;- shapiro_res %&gt;% filter(!is.na(Normal))
</span></span><span class=line><span class=cl>  Normal_prof &lt;- edata[rownames(edata)%in%shapiro_res_final$Normal, ]
</span></span><span class=line><span class=cl>  Non_Normal_prof &lt;- edata[!rownames(edata)%in%shapiro_res_final$Normal, ]
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  if(nrow(Normal_prof) != 0){
</span></span><span class=line><span class=cl>    # Normally distributed proteins
</span></span><span class=line><span class=cl>    Welch_res &lt;- apply(Normal_prof, 1, function(x, y){
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>      dat &lt;- data.frame(value=as.numeric(x), group=y)
</span></span><span class=line><span class=cl>      mn_GM &lt;- tapply(dat$value_scale, dat$group, compositions::geometricmean) %&gt;%
</span></span><span class=line><span class=cl>        data.frame() %&gt;% setNames(&#34;value&#34;) %&gt;%
</span></span><span class=line><span class=cl>        rownames_to_column(&#34;Group&#34;)
</span></span><span class=line><span class=cl>      mn_GM1 &lt;- with(mn_GM, mn_GM[Group%in%group_name[1], &#34;value&#34;])
</span></span><span class=line><span class=cl>      mn_GM2 &lt;- with(mn_GM, mn_GM[Group%in%group_name[2], &#34;value&#34;])
</span></span><span class=line><span class=cl>      Log2FC_GM &lt;- log2(mn_GM1/mn_GM2)
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>      # pvalue
</span></span><span class=line><span class=cl>      rest &lt;- rstatix::t_test(data = dat, value ~ group)
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>      return(c(Log2FC_GM, mn_GM1, mn_GM2,
</span></span><span class=line><span class=cl>               rest$statistic, rest$p))
</span></span><span class=line><span class=cl>    }, pheno$Group) %&gt;%
</span></span><span class=line><span class=cl>      t() %&gt;% data.frame() %&gt;%
</span></span><span class=line><span class=cl>      setNames(c(&#34;logFC&#34;, paste0(&#34;geometricmean_&#34;, group_name),
</span></span><span class=line><span class=cl>                 &#34;Statistic&#34;, &#34;P.value&#34;))
</span></span><span class=line><span class=cl>    Normal_res &lt;- Welch_res %&gt;%
</span></span><span class=line><span class=cl>      rownames_to_column(&#34;FeatureID&#34;) %&gt;%
</span></span><span class=line><span class=cl>      arrange(desc(abs(logFC)), P.value)
</span></span><span class=line><span class=cl>  }else{
</span></span><span class=line><span class=cl>    Normal_res &lt;- data.frame()
</span></span><span class=line><span class=cl>  }
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  if(nrow(Non_Normal_prof) != 0){
</span></span><span class=line><span class=cl>    # non-Normally distributed proteins
</span></span><span class=line><span class=cl>    Wilcox_res &lt;- apply(Non_Normal_prof, 1, function(x, y){
</span></span><span class=line><span class=cl>      dat &lt;- data.frame(value=as.numeric(x), group=y)
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>      # Fold2Change
</span></span><span class=line><span class=cl>      dat$value_scale &lt;- scale(dat$value, center = TRUE, scale = TRUE)
</span></span><span class=line><span class=cl>      mn_GM &lt;- tapply(dat$value_scale, dat$group, compositions::geometricmean) %&gt;%
</span></span><span class=line><span class=cl>        data.frame() %&gt;% setNames(&#34;value&#34;) %&gt;%
</span></span><span class=line><span class=cl>        rownames_to_column(&#34;Group&#34;)
</span></span><span class=line><span class=cl>      mn_GM1 &lt;- with(mn_GM, mn_GM[Group%in%group_name[1], &#34;value&#34;])
</span></span><span class=line><span class=cl>      mn_GM2 &lt;- with(mn_GM, mn_GM[Group%in%group_name[2], &#34;value&#34;])
</span></span><span class=line><span class=cl>      Log2FC_GM &lt;- log2(mn_GM1/mn_GM2)
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>      # pvalue
</span></span><span class=line><span class=cl>      rest &lt;- wilcox.test(data = dat, value ~ group)
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>      return(c(Log2FC_GM, mn_GM1, mn_GM2,
</span></span><span class=line><span class=cl>               rest$statistic, rest$p.value))
</span></span><span class=line><span class=cl>    }, pheno$Group) %&gt;%
</span></span><span class=line><span class=cl>      t() %&gt;% data.frame() %&gt;%
</span></span><span class=line><span class=cl>      setNames(c(&#34;logFC&#34;, paste0(&#34;geometricmean_&#34;, group_name),
</span></span><span class=line><span class=cl>                 &#34;Statistic&#34;, &#34;P.value&#34;))
</span></span><span class=line><span class=cl>    Non_Normal_res &lt;- Wilcox_res %&gt;%
</span></span><span class=line><span class=cl>      rownames_to_column(&#34;FeatureID&#34;) %&gt;%
</span></span><span class=line><span class=cl>      arrange(desc(abs(logFC)), P.value)
</span></span><span class=line><span class=cl>  }else{
</span></span><span class=line><span class=cl>    Non_Normal_res &lt;- data.frame()
</span></span><span class=line><span class=cl>  }
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  # Number &amp; Block
</span></span><span class=line><span class=cl>  res &lt;- rbind(Normal_res, Non_Normal_res)
</span></span><span class=line><span class=cl>  res$adj.P.Val &lt;- p.adjust(as.numeric(res$P.value), method = &#34;BH&#34;)
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  dat_status &lt;- table(pheno$Group)
</span></span><span class=line><span class=cl>  dat_status_number &lt;- as.numeric(dat_status)
</span></span><span class=line><span class=cl>  dat_status_name &lt;- names(dat_status)
</span></span><span class=line><span class=cl>  res$Block &lt;- paste(paste(dat_status_number[1], dat_status_name[1], sep = &#34;_&#34;),
</span></span><span class=line><span class=cl>                         &#34;vs&#34;,
</span></span><span class=line><span class=cl>                         paste(dat_status_number[2], dat_status_name[2], sep = &#34;_&#34;))
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  res_final &lt;- res %&gt;% dplyr::select(FeatureID, Block, adj.P.Val, P.value, everything()) %&gt;%
</span></span><span class=line><span class=cl>    arrange(adj.P.Val)
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  return(res_final)
</span></span><span class=line><span class=cl>}
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>Test_res &lt;- DEA_test(datset=ExprSet_count)
</span></span><span class=line><span class=cl>head(Test_res)
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153138.png alt loading=lazy data-zoomable></div></div></figure></p><h3 id=差异分组-3>差异分组</h3><p>设置差异基因过滤阈值（logFC+adj.P.Va）</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>Test_res_df</span> <span class=o>&lt;-</span> <span class=n>Test_res</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=o>!</span><span class=nf>is.nan</span><span class=p>(</span><span class=n>logFC</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=nf>filter</span><span class=p>(</span><span class=o>!</span><span class=nf>is.infinite</span><span class=p>(</span><span class=n>logFC</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=nf>filter</span><span class=p>(</span><span class=nf>is.finite</span><span class=p>(</span><span class=n>logFC</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=nf>arrange</span><span class=p>(</span><span class=n>logFC</span><span class=p>,</span> <span class=n>adj.P.Val</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>log2fc_threshold</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>Test_res_df</span><span class=p>,</span> <span class=nf>mean</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>logFC</span><span class=p>))</span> <span class=o>+</span> <span class=m>1.5</span> <span class=o>*</span> <span class=nf>sd</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>logFC</span><span class=p>)))</span>
</span></span><span class=line><span class=cl><span class=nf>message</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=s>&#34;Threshold of log2Foldchange [Mean+1.5(SD)] is&#34;</span><span class=p>,</span> <span class=n>log2fc_threshold</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=n>pval</span> <span class=o>&lt;-</span> <span class=m>0.05</span>
</span></span><span class=line><span class=cl><span class=n>Test_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=n>Test_res_df</span><span class=o>$</span><span class=n>logFC</span> <span class=o>&gt;=</span> <span class=n>log2fc_threshold</span> <span class=o>&amp;</span> <span class=n>Test_res_df</span><span class=o>$</span><span class=n>adj.P.Va</span> <span class=o>&lt;</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>grp[1]</span>
</span></span><span class=line><span class=cl><span class=n>Test_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=n>Test_res_df</span><span class=o>$</span><span class=n>logFC</span> <span class=o>&lt;=</span> <span class=o>-</span><span class=n>log2fc_threshold</span> <span class=o>&amp;</span> <span class=n>Test_res_df</span><span class=o>$</span><span class=n>adj.P.Va</span> <span class=o>&lt;</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>grp[2]</span>
</span></span><span class=line><span class=cl><span class=n>Test_res_df</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>Test_res_df</span><span class=o>$</span><span class=n>logFC</span><span class=p>)</span> <span class=o>&lt;</span> <span class=n>log2fc_threshold</span> <span class=o>|</span> <span class=n>Test_res_df</span><span class=o>$</span><span class=n>adj.P.Va</span> <span class=o>&gt;=</span> <span class=n>pval</span><span class=p>),</span> <span class=s>&#34;Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignf&#34;</span>
</span></span><span class=line><span class=cl><span class=nf>table</span><span class=p>(</span><span class=n>Test_res_df</span><span class=o>$</span><span class=n>Enrichment</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126170048.webp alt loading=lazy data-zoomable></div></div></figure></p><p><strong>结果：差异显著的基因数目在Tumor和Normal组分别是716和294，火山图展示差异基因如上图修改代码即可。</strong></p><h2 id=不同方法的结果比较>不同方法的结果比较</h2><p>虽然<strong>log2FoldChange</strong>值都是使用mean+1.5SD(均值+1.5倍方差)，但是会发现四种方法的最后的<strong>log2FoldChange</strong>的阈值均不相同，这也是导致差异基因数目不同的原因之一。</p><h3 id=保存结果>保存结果</h3><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>save</span><span class=p>(</span><span class=n>DESeq_result_df</span><span class=p>,</span> <span class=n>Limma_res_df</span><span class=p>,</span> <span class=n>EdgeR_res_df</span><span class=p>,</span> <span class=n>Test_res_df</span><span class=p>,</span> <span class=n>file</span> <span class=o>=</span> <span class=s>&#34;DEA.RData&#34;</span><span class=p>)</span>
</span></span></code></pre></div><h3 id=差异基因数目分布>差异基因数目分布</h3><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>DEA_status</span> <span class=o>&lt;-</span> <span class=nf>data.frame</span><span class=p>(</span><span class=n>DESeq2</span><span class=o>=</span><span class=nf>as.numeric</span><span class=p>(</span><span class=nf>table</span><span class=p>(</span><span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>Enrichment</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                         <span class=n>Limma</span><span class=o>=</span><span class=nf>as.numeric</span><span class=p>(</span><span class=nf>table</span><span class=p>(</span><span class=n>Limma_res_df</span><span class=o>$</span><span class=n>Enrichment</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                         <span class=n>EdgeR</span><span class=o>=</span><span class=nf>as.numeric</span><span class=p>(</span><span class=nf>table</span><span class=p>(</span><span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>Enrichment</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                         <span class=n>T_wilcox</span><span class=o>=</span><span class=nf>as.numeric</span><span class=p>(</span><span class=nf>table</span><span class=p>(</span><span class=n>Test_res_df</span><span class=o>$</span><span class=n>Enrichment</span><span class=p>)))</span>
</span></span><span class=line><span class=cl><span class=nf>rownames</span><span class=p>(</span><span class=n>DEA_status</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Non-significant&#34;</span><span class=p>,</span> <span class=n>grp</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=c1>#DT::datatable(DEA_status)</span>
</span></span><span class=line><span class=cl><span class=n>DEA_status</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153427.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Notes:</strong> T_wilcox方法的差异基因数目是最少的。</p><h3 id=差异基因交集venn>差异基因交集Venn</h3><p>Venn画图可参考<a href=https://www.jianshu.com/p/bac9a4d02a2f target=_blank rel=noopener>R可视化：ggplot语法的Venn 图</a></p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>DESeq_DEG</span> <span class=o>&lt;-</span> <span class=n>DESeq_result_df</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignf&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Limma_DEG</span> <span class=o>&lt;-</span> <span class=n>Limma_res_df</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignf&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>EdgeR_DEG</span> <span class=o>&lt;-</span> <span class=n>EdgeR_res_df</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignf&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Test_DEG</span> <span class=o>&lt;-</span> <span class=n>Test_res_df</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignf&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>ggVennDiagram</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>dat_DEG</span> <span class=o>&lt;-</span> <span class=nf>list</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>A</span> <span class=o>=</span> <span class=n>DESeq_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>  <span class=n>B</span> <span class=o>=</span> <span class=n>Limma_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>  <span class=n>C</span> <span class=o>=</span> <span class=n>EdgeR_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>D</span> <span class=o>=</span> <span class=n>Test_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>ggVennDiagram</span><span class=p>(</span><span class=n>dat_DEG</span><span class=p>,</span> <span class=n>label_alpha</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>category.names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;DESeq2&#34;</span><span class=p>,</span> <span class=s>&#34;Limma+Voom&#34;</span><span class=p>,</span> <span class=s>&#34;EdgeR&#34;</span><span class=p>,</span> <span class=s>&#34;T-test/WilcoxTest&#34;</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>  <span class=nf>scale_fill_gradient</span><span class=p>(</span><span class=n>low</span> <span class=o>=</span> <span class=s>&#34;blue&#34;</span><span class=p>,</span> <span class=n>high</span> <span class=o>=</span> <span class=s>&#34;yellow&#34;</span><span class=p>)</span><span class=o>+</span>
</span></span><span class=line><span class=cl>  <span class=nf>ggtitle</span><span class=p>(</span><span class=s>&#34;Comprison of Differential Expression Approaches&#34;</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153533.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Notes</strong>：</p><ol><li><p>四种方法重叠的差异基因数目是160个；</p></li><li><p>DEseq2和Limma+Voom的重叠差异基因数目：92+634+160+14；</p></li><li><p>DEseq2和EdgeR的重叠差异基因数目：634+326+57+160；</p></li><li><p>DEseq2和T-test/WilcoxTest的重叠差异基因数目：160+57+12+14；</p></li><li><p>DEseq2，Limma+Voom和EdgeR的重叠差异基因数目：634+160；</p></li><li><p>T-test/WilcoxTest和其他三种方法的重叠差异基因数目最少。</p></li></ol><h3 id=差异基因heatmapvolcano>差异基因heatmap+volcano</h3><p>参考<a href=https://www.jianshu.com/p/0165de6fac90 target=_blank rel=noopener>3大差异分析r包：DESeq2、edgeR和limma</a></p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>tinyarray</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>dat_count</span> <span class=o>&lt;-</span> <span class=nf>exprs</span><span class=p>(</span><span class=n>ExprSet_count</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>dat_group</span> <span class=o>&lt;-</span> <span class=nf>pData</span><span class=p>(</span><span class=n>ExprSet_count</span><span class=p>)</span> 
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>h1</span> <span class=o>&lt;-</span> <span class=nf>draw_heatmap</span><span class=p>(</span><span class=n>dat_count[DESeq_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>]</span><span class=p>,</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_group</span><span class=o>$</span><span class=n>Group</span><span class=p>),</span> <span class=n>n_cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>h2</span> <span class=o>&lt;-</span> <span class=nf>draw_heatmap</span><span class=p>(</span><span class=n>dat_count[Limma_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>]</span><span class=p>,</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_group</span><span class=o>$</span><span class=n>Group</span><span class=p>),</span> <span class=n>n_cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>h3</span> <span class=o>&lt;-</span> <span class=nf>draw_heatmap</span><span class=p>(</span><span class=n>dat_count[EdgeR_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>]</span><span class=p>,</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_group</span><span class=o>$</span><span class=n>Group</span><span class=p>),</span> <span class=n>n_cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>h4</span> <span class=o>&lt;-</span> <span class=nf>draw_heatmap</span><span class=p>(</span><span class=n>dat_count[Test_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>]</span><span class=p>,</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_group</span><span class=o>$</span><span class=n>Group</span><span class=p>),</span> <span class=n>n_cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>m2d</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>x</span><span class=p>){</span>
</span></span><span class=line><span class=cl>  <span class=nf>mean</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>x</span><span class=p>))</span> <span class=o>+</span> <span class=m>1.5</span> <span class=o>*</span> <span class=nf>sd</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>x</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>v1</span> <span class=o>&lt;-</span> <span class=nf>draw_volcano</span><span class=p>(</span><span class=n>DESeq_result_df</span> <span class=o>%&gt;%</span> <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;FeatureID&#34;</span><span class=p>),</span> <span class=n>pkg</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>logFC_cutoff</span> <span class=o>=</span> <span class=nf>m2d</span><span class=p>(</span><span class=n>DESeq_result_df</span><span class=o>$</span><span class=n>log2FoldChange</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=n>v2</span> <span class=o>&lt;-</span> <span class=nf>draw_volcano</span><span class=p>(</span><span class=n>Limma_res_df</span> <span class=o>%&gt;%</span> <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;FeatureID&#34;</span><span class=p>),</span> <span class=n>pkg</span> <span class=o>=</span> <span class=m>3</span><span class=p>,</span> <span class=n>logFC_cutoff</span> <span class=o>=</span> <span class=nf>m2d</span><span class=p>(</span><span class=n>Limma_res_df</span><span class=o>$</span><span class=n>logFC</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=n>v3</span> <span class=o>&lt;-</span> <span class=nf>draw_volcano</span><span class=p>(</span><span class=n>EdgeR_res_df</span> <span class=o>%&gt;%</span> <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;FeatureID&#34;</span><span class=p>),</span> <span class=n>pkg</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>logFC_cutoff</span> <span class=o>=</span> <span class=nf>m2d</span><span class=p>(</span><span class=n>EdgeR_res_df</span><span class=o>$</span><span class=n>logFC</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># convert Test_res_df format into DESeq_result_df format</span>
</span></span><span class=line><span class=cl><span class=n>Test_res_df_DESeq</span> <span class=o>&lt;-</span> <span class=n>Test_res_df</span> <span class=o>%&gt;%</span> <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>baseMean</span><span class=o>=</span><span class=n>geometricmean_Tumor</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                                                   <span class=n>log2FoldChange</span><span class=o>=</span><span class=n>logFC</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                                                   <span class=n>lfcSE</span><span class=o>=</span><span class=n>geometricmean_Normal</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                                                   <span class=n>stat</span><span class=o>=</span><span class=n>Statistic</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                                                   <span class=n>pvalue</span><span class=o>=</span><span class=n>P.value</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                                                   <span class=n>padj</span><span class=o>=</span><span class=n>adj.P.Val</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>baseMean</span><span class=p>,</span> <span class=n>log2FoldChange</span><span class=p>,</span> <span class=n>lfcSE</span><span class=p>,</span> <span class=n>stat</span><span class=p>,</span> <span class=n>pvalue</span><span class=p>,</span> <span class=n>padj</span><span class=p>,</span> <span class=n>Enrichment</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>v4</span> <span class=o>&lt;-</span> <span class=nf>draw_volcano</span><span class=p>(</span><span class=n>Test_res_df_DESeq</span> <span class=o>%&gt;%</span> <span class=nf>column_to_rownames</span><span class=p>(</span><span class=s>&#34;FeatureID&#34;</span><span class=p>),</span> <span class=n>pkg</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>logFC_cutoff</span> <span class=o>=</span> <span class=nf>m2d</span><span class=p>(</span><span class=n>Test_res_df_DESeq</span><span class=o>$</span><span class=n>log2FoldChange</span><span class=p>))</span><span class=o>+</span>
</span></span><span class=line><span class=cl>  <span class=nf>xlab</span><span class=p>(</span><span class=s>&#34;T-test/WilcoxTest&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=p>(</span><span class=n>h1</span> <span class=o>+</span> <span class=n>h2</span> <span class=o>+</span> <span class=n>h3</span> <span class=o>+</span> <span class=n>h4</span> <span class=o>+</span> <span class=n>v1</span> <span class=o>+</span> <span class=n>v2</span> <span class=o>+</span> <span class=n>v3</span> <span class=o>+</span> <span class=n>v4</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>  <span class=nf>plot_layout</span><span class=p>(</span><span class=n>guides</span> <span class=o>=</span> <span class=s>&#34;collect&#34;</span><span class=p>,</span> <span class=n>nrow</span> <span class=o>=</span> <span class=m>2</span><span class=p>)</span> <span class=o>&amp;</span> <span class=nf>theme</span><span class=p>(</span><span class=n>legend.position</span> <span class=o>=</span> <span class=s>&#34;none&#34;</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153632.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Notes: 火山图和热图也表明前三种差异检验方法差异基因趋势相对一致，而最后一种与前面三种均存在差异，这提示我们用counts矩阵在t-test/wilcox-rank-sum test中做假设检验时候要非常小心注意（也说明测序深度对假设检验结果影响较大）。</strong></p><h3 id=共有差异基因venn>共有差异基因venn</h3><p>综上，我们提取前三种方法（DEseq2，Limma+Voom和EdgeR）的重叠差异基因做后续分析，鉴别三种方法的优劣。</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>DESeq_DEG</span> <span class=o>&lt;-</span> <span class=n>DESeq_result_df</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignf&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Limma_DEG</span> <span class=o>&lt;-</span> <span class=n>Limma_res_df</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignf&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>EdgeR_DEG</span> <span class=o>&lt;-</span> <span class=n>EdgeR_res_df</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignf&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>ExtractDEG</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>datDEG</span><span class=o>=</span><span class=n>DESeq_DEG</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>tag</span><span class=o>=</span><span class=n>grp[1]</span><span class=p>){</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>tmp</span> <span class=o>&lt;-</span> <span class=n>datDEG</span> <span class=o>%&gt;%</span> <span class=nf>filter</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>==</span> <span class=n>tag</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>res</span> <span class=o>&lt;-</span> <span class=n>tmp</span><span class=o>$</span><span class=n>FeatureID</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>res</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=n>DESeq_DEG_normal</span> <span class=o>&lt;-</span> <span class=nf>ExtractDEG</span><span class=p>(</span><span class=n>datDEG</span><span class=o>=</span><span class=n>DESeq_DEG</span><span class=p>,</span> <span class=n>tag</span><span class=o>=</span><span class=n>grp[1]</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>DESeq_DEG_tumor</span> <span class=o>&lt;-</span> <span class=nf>ExtractDEG</span><span class=p>(</span><span class=n>datDEG</span><span class=o>=</span><span class=n>DESeq_DEG</span><span class=p>,</span> <span class=n>tag</span><span class=o>=</span><span class=n>grp[2]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>Limma_DEG_normal</span> <span class=o>&lt;-</span> <span class=nf>ExtractDEG</span><span class=p>(</span><span class=n>datDEG</span><span class=o>=</span><span class=n>Limma_DEG</span><span class=p>,</span> <span class=n>tag</span><span class=o>=</span><span class=n>grp[1]</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>Limma_DEG_tumor</span> <span class=o>&lt;-</span> <span class=nf>ExtractDEG</span><span class=p>(</span><span class=n>datDEG</span><span class=o>=</span><span class=n>Limma_DEG</span><span class=p>,</span> <span class=n>tag</span><span class=o>=</span><span class=n>grp[2]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>EdgeR_DEG_normal</span> <span class=o>&lt;-</span> <span class=nf>ExtractDEG</span><span class=p>(</span><span class=n>datDEG</span><span class=o>=</span><span class=n>EdgeR_DEG</span><span class=p>,</span> <span class=n>tag</span><span class=o>=</span><span class=n>grp[1]</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>EdgeR_DEG_tumor</span> <span class=o>&lt;-</span> <span class=nf>ExtractDEG</span><span class=p>(</span><span class=n>datDEG</span><span class=o>=</span><span class=n>EdgeR_DEG</span><span class=p>,</span> <span class=n>tag</span><span class=o>=</span><span class=n>grp[2]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>DEG_normal</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>DESeq_DEG_normal</span><span class=p>,</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>Limma_DEG_normal</span><span class=p>,</span> <span class=n>EdgeR_DEG_normal</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=n>DEG_tumor</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>DESeq_DEG_tumor</span><span class=p>,</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>Limma_DEG_tumor</span><span class=p>,</span> <span class=n>EdgeR_DEG_tumor</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>draw_venn</span><span class=p>(</span><span class=nf>list</span><span class=p>(</span><span class=n>DESeq2</span><span class=o>=</span><span class=n>DESeq_DEG_normal</span><span class=p>,</span>
</span></span><span class=line><span class=cl>               <span class=n>Limma</span><span class=o>=</span><span class=n>Limma_DEG_normal</span><span class=p>,</span>
</span></span><span class=line><span class=cl>               <span class=n>EdgeR</span><span class=o>=</span><span class=n>EdgeR_DEG_normal</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=s>&#34;DEG in Normal&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl><span class=nf>draw_venn</span><span class=p>(</span><span class=nf>list</span><span class=p>(</span><span class=n>DESeq2</span><span class=o>=</span><span class=n>DESeq_DEG_tumor</span><span class=p>,</span>
</span></span><span class=line><span class=cl>               <span class=n>Limma</span><span class=o>=</span><span class=n>Limma_DEG_tumor</span><span class=p>,</span>
</span></span><span class=line><span class=cl>               <span class=n>EdgeR</span><span class=o>=</span><span class=n>EdgeR_DEG_tumor</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=s>&#34;DEG in Tumor&#34;</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153713.png alt loading=lazy data-zoomable></div></div></figure></p><p><strong>Notes: 三种方法富集在Normal和Tumor组的差异基因还是有区别的。接下来我们用相关性图展示三种方法的差异基因的相关性。</strong></p><h3 id=差异基因log2foldchange相关性分析>差异基因Log2Foldchange相关性分析</h3><p>差异基因的Log2Foldchange相关性分析 (参考文章<a href=https://www.jianshu.com/p/517167c50a5f target=_blank rel=noopener>转录组分析5——差异表达分析</a>)</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>corrplot</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>All_DEG</span> <span class=o>&lt;-</span> <span class=nf>unique</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=n>DESeq_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>Limma_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>EdgeR_DEG</span><span class=o>$</span><span class=n>FeatureID</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=nf>length</span><span class=p>(</span><span class=n>All_DEG</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>All_DEG_lg2FC</span> <span class=o>&lt;-</span> <span class=nf>data.frame</span><span class=p>(</span><span class=n>DESeq2</span><span class=o>=</span><span class=n>DESeq_result_df[DESeq_result_df</span><span class=o>$</span><span class=n>FeatureID</span><span class=o>%in%</span><span class=n>All_DEG</span><span class=p>,</span> <span class=m>3</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                            <span class=n>Limma</span><span class=o>=</span><span class=n>Limma_res_df[Limma_res_df</span><span class=o>$</span><span class=n>FeatureID</span><span class=o>%in%</span><span class=n>All_DEG</span><span class=p>,</span> <span class=m>2</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                            <span class=n>EdgeR</span><span class=o>=</span><span class=n>EdgeR_res_df[EdgeR_res_df</span><span class=o>$</span><span class=n>FeatureID</span><span class=o>%in%</span><span class=n>All_DEG</span><span class=p>,</span> <span class=m>2</span><span class=n>]</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=nf>na.omit</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl><span class=n>All_DEG_lg2FC_cor</span> <span class=o>&lt;-</span> <span class=nf>cor</span><span class=p>(</span><span class=n>All_DEG_lg2FC</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>All_DEG_lg2FC_cor</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>corrplot</span><span class=p>(</span><span class=n>All_DEG_lg2FC_cor</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153739.png alt loading=lazy data-zoomable></div></div></figure></p><h3 id=共有差异基因总结>共有差异基因总结</h3><p>三种方法的差异基因的PCA+heatmap+Venn（参考<a href=https://www.jianshu.com/p/0165de6fac90 target=_blank rel=noopener>3大差异分析r包：DESeq2、edgeR和limma</a>）</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>DEG_normal</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>DESeq_DEG_normal</span><span class=p>,</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>Limma_DEG_normal</span><span class=p>,</span> <span class=n>EdgeR_DEG_normal</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=n>DEG_tumor</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>DESeq_DEG_tumor</span><span class=p>,</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>Limma_DEG_tumor</span><span class=p>,</span> <span class=n>EdgeR_DEG_tumor</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>dat_count_log</span> <span class=o>&lt;-</span> <span class=nf>log2</span><span class=p>(</span><span class=nf>cpm</span><span class=p>(</span><span class=n>dat_count</span><span class=p>)</span><span class=m>+1</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>DEG_heatmap</span> <span class=o>&lt;-</span> <span class=nf>draw_heatmap</span><span class=p>(</span><span class=n>dat_count_log</span><span class=nf>[c</span><span class=p>(</span><span class=n>DEG_normal</span><span class=p>,</span> <span class=n>DEG_tumor</span><span class=p>),</span> <span class=n>]</span><span class=p>,</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_group</span><span class=o>$</span><span class=n>Group</span><span class=p>),</span> <span class=n>n_cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>DEG_pca</span> <span class=o>&lt;-</span> <span class=nf>draw_pca</span><span class=p>(</span><span class=n>dat_count_log</span><span class=nf>[c</span><span class=p>(</span><span class=n>DEG_normal</span><span class=p>,</span> <span class=n>DEG_tumor</span><span class=p>),</span> <span class=n>]</span><span class=p>,</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_group</span><span class=o>$</span><span class=n>Group</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>DEG_Normal_venn</span> <span class=o>&lt;-</span> <span class=nf>draw_venn</span><span class=p>(</span><span class=nf>list</span><span class=p>(</span><span class=n>DESeq2</span><span class=o>=</span><span class=n>DESeq_DEG_normal</span><span class=p>,</span>
</span></span><span class=line><span class=cl>               <span class=n>Limma</span><span class=o>=</span><span class=n>Limma_DEG_normal</span><span class=p>,</span>
</span></span><span class=line><span class=cl>               <span class=n>EdgeR</span><span class=o>=</span><span class=n>EdgeR_DEG_normal</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=s>&#34;DEG in Normal&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>DEG_Tumor_venn</span> <span class=o>&lt;-</span> <span class=nf>draw_venn</span><span class=p>(</span><span class=nf>list</span><span class=p>(</span><span class=n>DESeq2</span><span class=o>=</span><span class=n>DESeq_DEG_tumor</span><span class=p>,</span>
</span></span><span class=line><span class=cl>               <span class=n>Limma</span><span class=o>=</span><span class=n>Limma_DEG_tumor</span><span class=p>,</span>
</span></span><span class=line><span class=cl>               <span class=n>EdgeR</span><span class=o>=</span><span class=n>EdgeR_DEG_tumor</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=s>&#34;DEG in Tumor&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=p>(</span><span class=n>DEG_pca</span> <span class=o>+</span> <span class=n>DEG_heatmap</span> <span class=o>+</span> <span class=n>DEG_Normal_venn</span> <span class=o>+</span> <span class=n>DEG_Tumor_venn</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>  <span class=nf>plot_layout</span><span class=p>(</span><span class=n>nrow</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>guides</span> <span class=o>=</span> <span class=s>&#34;collect&#34;</span><span class=p>)</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/20211126153803.png alt loading=lazy data-zoomable></div></div></figure></p><h2 id=总结>总结</h2><ol><li><p>有文章说edgeR分析速度快，但是从这次分析看，它反而是最慢的，另外edgeR的假阳性太高；</p></li><li><p>DESeq2在计算标准化因子时耗时太久，但它的标准化因子相对来说最合理；</p></li><li><p>三种方法得到的差异基因不是完全重叠的，但再提取它们所有的差异基因<em>log2Foldchange</em>值做相关性分析，发现相似度极高，也就是说明三种方法差异不是很大；</p></li><li><p>最后大家选择适合自己的方法做差异分析即可。</p></li></ol><h2 id=systemic-information>systemic information</h2><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>sessionInfo</span><span class=p>()</span>
</span></span></code></pre></div><div class=highlight><pre tabindex=0 class=chroma><code class=language-markdown data-lang=markdown><span class=line><span class=cl>R version 4.0.2 (2020-06-22)
</span></span><span class=line><span class=cl>Platform: x86_64-conda_cos6-linux-gnu (64-bit)
</span></span><span class=line><span class=cl>Running under: CentOS Linux 8 (Core)
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>Matrix products: default
</span></span><span class=line><span class=cl>BLAS/LAPACK: /disk/share/anaconda3/lib/libopenblasp-r0.3.10.so
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>locale:
</span></span><span class=line><span class=cl> [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
</span></span><span class=line><span class=cl> [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
</span></span><span class=line><span class=cl> [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>attached base packages:
</span></span><span class=line><span class=cl>[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>other attached packages:
</span></span><span class=line><span class=cl> [1] corrplot_0.84               tinyarray_2.2.7             ggVennDiagram_0.5.0         edgeR_3.32.1               
</span></span><span class=line><span class=cl> [5] ggrepel_0.9.1.9999          DESeq2_1.30.1               SummarizedExperiment_1.20.0 MatrixGenerics_1.2.1       
</span></span><span class=line><span class=cl> [9] matrixStats_0.59.0          GenomicRanges_1.42.0        GenomeInfoDb_1.26.4         IRanges_2.24.1             
</span></span><span class=line><span class=cl>[13] S4Vectors_0.28.1            convert_1.64.0              marray_1.68.0               limma_3.46.0               
</span></span><span class=line><span class=cl>[17] Biobase_2.50.0              BiocGenerics_0.36.0         cowplot_1.1.0               patchwork_1.0.1            
</span></span><span class=line><span class=cl>[21] ggplot2_3.3.5               data.table_1.14.0           tibble_3.1.5                dplyr_1.0.7                
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>loaded via a namespace (and not attached):
</span></span><span class=line><span class=cl>  [1] utf8_1.2.1             reticulate_1.18        tidyselect_1.1.1       RSQLite_2.2.5          AnnotationDbi_1.52.0  
</span></span><span class=line><span class=cl>  [6] htmlwidgets_1.5.3      FactoMineR_2.4         grid_4.0.2             BiocParallel_1.24.1    scatterpie_0.1.5      
</span></span><span class=line><span class=cl> [11] munsell_0.5.0          units_0.7-2            DT_0.18                umap_0.2.7.0           withr_2.4.2           
</span></span><span class=line><span class=cl> [16] colorspace_2.0-2       GOSemSim_2.16.1        knitr_1.33             leaps_3.1              rstudioapi_0.13       
</span></span><span class=line><span class=cl> [21] robustbase_0.93-6      bayesm_3.1-4           ggsignif_0.6.0         DOSE_3.16.0            labeling_0.4.2        
</span></span><span class=line><span class=cl> [26] GenomeInfoDbData_1.2.4 KMsurv_0.1-5           polyclip_1.10-0        bit64_4.0.5            farver_2.1.0          
</span></span><span class=line><span class=cl> [31] pheatmap_1.0.12        downloader_0.4         vctrs_0.3.8            generics_0.1.0         lambda.r_1.2.4        
</span></span><span class=line><span class=cl> [36] xfun_0.24              R6_2.5.0               graphlayouts_0.7.1     locfit_1.5-9.4         gridGraphics_0.5-1    
</span></span><span class=line><span class=cl> [41] bitops_1.0-7           cachem_1.0.5           fgsea_1.16.0           DelayedArray_0.16.3    assertthat_0.2.1      
</span></span><span class=line><span class=cl> [46] scales_1.1.1           ggraph_2.0.4           enrichplot_1.10.1      gtable_0.3.0           tidygraph_1.2.0       
</span></span><span class=line><span class=cl> [51] rlang_0.4.11           genefilter_1.72.0      scatterplot3d_0.3-41   splines_4.0.2          rstatix_0.7.0         
</span></span><span class=line><span class=cl> [56] broom_0.7.9            BiocManager_1.30.16    yaml_2.2.1             reshape2_1.4.4         abind_1.4-5           
</span></span><span class=line><span class=cl> [61] crosstalk_1.1.1        backports_1.2.1        qvalue_2.22.0          clusterProfiler_3.18.0 tensorA_0.36.2        
</span></span><span class=line><span class=cl> [66] tools_4.0.2            ggplotify_0.0.5        ellipsis_0.3.2         jquerylib_0.1.4        RColorBrewer_1.1-2    
</span></span><span class=line><span class=cl> [71] proxy_0.4-26           Rcpp_1.0.7             plyr_1.8.6             zlibbioc_1.36.0        classInt_0.4-3        
</span></span><span class=line><span class=cl> [76] purrr_0.3.4            RCurl_1.98-1.3         ggpubr_0.4.0           openssl_1.4.4          viridis_0.6.1         
</span></span><span class=line><span class=cl> [81] zoo_1.8-8              cluster_2.1.0          haven_2.3.1            factoextra_1.0.7       tinytex_0.32          
</span></span><span class=line><span class=cl> [86] magrittr_2.0.1         RSpectra_0.16-0        futile.options_1.0.1   DO.db_2.9              openxlsx_4.2.3        
</span></span><span class=line><span class=cl> [91] survminer_0.4.9        hms_1.1.0              evaluate_0.14          xtable_1.8-4           XML_3.99-0.6          
</span></span><span class=line><span class=cl> [96] VennDiagram_1.6.20     rio_0.5.16             readxl_1.3.1           gridExtra_2.3          compiler_4.0.2        
</span></span><span class=line><span class=cl>[101] KernSmooth_2.23-18     crayon_1.4.1           shadowtext_0.0.7       htmltools_0.5.1.1      tidyr_1.1.4           
</span></span><span class=line><span class=cl>[106] geneplotter_1.66.0     DBI_1.1.1              tweenr_1.0.1           formatR_1.11           MASS_7.3-54           
</span></span><span class=line><span class=cl>[111] sf_0.9-5               compositions_2.0-2     Matrix_1.3-4           car_3.0-10             cli_3.0.1             
</span></span><span class=line><span class=cl>[116] igraph_1.2.6           km.ci_0.5-2            forcats_0.5.0          pkgconfig_2.0.3        flashClust_1.01-2     
</span></span><span class=line><span class=cl>[121] rvcheck_0.1.8          foreign_0.8-81         annotate_1.68.0        bslib_0.2.5.1          XVector_0.30.0        
</span></span><span class=line><span class=cl>[126] stringr_1.4.0          digest_0.6.27          rmarkdown_2.9          cellranger_1.1.0       fastmatch_1.1-0       
</span></span><span class=line><span class=cl>[131] survMisc_0.5.5         curl_4.3.2             lifecycle_1.0.0        jsonlite_1.7.2         carData_3.0-4         
</span></span><span class=line><span class=cl>[136] futile.logger_1.4.3    viridisLite_0.4.0      askpass_1.1            fansi_0.5.0            pillar_1.6.4          
</span></span><span class=line><span class=cl>[141] lattice_0.20-44        fastmap_1.1.0          httr_1.4.2             DEoptimR_1.0-8         survival_3.2-11       
</span></span><span class=line><span class=cl>[146] GO.db_3.12.1           glue_1.4.2             zip_2.1.1              bit_4.0.4              ggforce_0.3.2         
</span></span><span class=line><span class=cl>[151] class_7.3-19           stringi_1.4.6          sass_0.4.0             blob_1.2.1             org.Hs.eg.db_3.12.0   
</span></span><span class=line><span class=cl>[156] memoise_2.0.0          e1071_1.7-7 
</span></span></code></pre></div><h2 id=reference>Reference</h2><ol><li><p><a href=https://www.jianshu.com/p/1940c5954c81 target=_blank rel=noopener>RPKM、FPKM、TPM详解</a></p></li><li><p><a href=http://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/ target=_blank rel=noopener>rpkm-fpkm-and-tpm-clearly-explained</a></p></li><li><p><a href=http://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html target=_blank rel=noopener>DESeq2</a></p></li><li><p><a href=https://support.bioconductor.org/p/p132731/ target=_blank rel=noopener>DESeq2 DFrame error</a></p></li><li><p><a href=https://ucdavis-bioinformatics-training.github.io/2018-June-RNA-Seq-Workshop/thursday/DE.html target=_blank rel=noopener>Differential Expression with Limma-Voom</a></p></li><li><p><a href=https://web.stanford.edu/class/bios221/labs/rnaseq/lab_4_rnaseq.html target=_blank rel=noopener>RNA Sequence Analysis in R: edgeR</a></p></li></ol><p>参考文章如引起任何侵权问题，可以与我<a href=mailto:zouhua1@outlook.com>联系</a>，谢谢。</p></div><div class=article-tags><a class="badge badge-light" href=/tag/differential-analysis/>Differential Analysis</a>
<a class="badge badge-light" href=/tag/rna-seq/>RNA-seq</a></div><div class=share-box><ul class=share><li><a href="https://twitter.com/intent/tweet?url=https%3A%2F%2Fzouhua.top%2Fpost%2Fmath_statistics%2F2022-10-27-da-methods-comparsion%2F&text=%E8%BD%AC%E5%BD%95%E7%BB%84%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90%EF%BC%88DESeq2%2Blimma%2BedgeR%2Bt-test%2Fwilcox-test%EF%BC%89%E6%80%BB%E7%BB%93" target=_blank rel=noopener class=share-btn-twitter aria-label=twitter><i class="fab fa-twitter"></i></a></li><li><a href="https://www.facebook.com/sharer.php?u=https%3A%2F%2Fzouhua.top%2Fpost%2Fmath_statistics%2F2022-10-27-da-methods-comparsion%2F&t=%E8%BD%AC%E5%BD%95%E7%BB%84%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90%EF%BC%88DESeq2%2Blimma%2BedgeR%2Bt-test%2Fwilcox-test%EF%BC%89%E6%80%BB%E7%BB%93" target=_blank rel=noopener class=share-btn-facebook aria-label=facebook><i class="fab fa-facebook"></i></a></li><li><a href="mailto:?subject=%E8%BD%AC%E5%BD%95%E7%BB%84%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90%EF%BC%88DESeq2%2Blimma%2BedgeR%2Bt-test%2Fwilcox-test%EF%BC%89%E6%80%BB%E7%BB%93&body=https%3A%2F%2Fzouhua.top%2Fpost%2Fmath_statistics%2F2022-10-27-da-methods-comparsion%2F" target=_blank rel=noopener class=share-btn-email aria-label=envelope><i class="fas fa-envelope"></i></a></li><li><a href="https://www.linkedin.com/shareArticle?url=https%3A%2F%2Fzouhua.top%2Fpost%2Fmath_statistics%2F2022-10-27-da-methods-comparsion%2F&title=%E8%BD%AC%E5%BD%95%E7%BB%84%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90%EF%BC%88DESeq2%2Blimma%2BedgeR%2Bt-test%2Fwilcox-test%EF%BC%89%E6%80%BB%E7%BB%93" target=_blank rel=noopener class=share-btn-linkedin aria-label=linkedin-in><i class="fab fa-linkedin-in"></i></a></li><li><a href="whatsapp://send?text=%E8%BD%AC%E5%BD%95%E7%BB%84%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90%EF%BC%88DESeq2%2Blimma%2BedgeR%2Bt-test%2Fwilcox-test%EF%BC%89%E6%80%BB%E7%BB%93%20https%3A%2F%2Fzouhua.top%2Fpost%2Fmath_statistics%2F2022-10-27-da-methods-comparsion%2F" target=_blank rel=noopener class=share-btn-whatsapp aria-label=whatsapp><i class="fab fa-whatsapp"></i></a></li><li><a href="https://service.weibo.com/share/share.php?url=https%3A%2F%2Fzouhua.top%2Fpost%2Fmath_statistics%2F2022-10-27-da-methods-comparsion%2F&title=%E8%BD%AC%E5%BD%95%E7%BB%84%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90%EF%BC%88DESeq2%2Blimma%2BedgeR%2Bt-test%2Fwilcox-test%EF%BC%89%E6%80%BB%E7%BB%93" target=_blank rel=noopener class=share-btn-weibo aria-label=weibo><i class="fab fa-weibo"></i></a></li></ul></div></div></article></div><div class=page-footer><div class=container><footer class=site-footer><p class=powered-by>© 2023 Hua Zou &#183;
Rendered by <a href=https://cran.r-project.org/ target=_blank rel=noopener><i class="fab fa-r-project"></i> </a><a href=https://github.com/rstudio/blogdown target=_blank rel=noopener>blogdown</a> ,
<a href=https://sourcethemes.com/academic/ target=_blank rel=noopener>Academic theme</a> and
<a href=https://gohugo.io target=_blank rel=noopener>Hugo</a>.
<span class=float-right aria-hidden=true><a href=# id=back_to_top><span class=button_icon><i class="fas fa-chevron-up fa-2x"></i></span></a></span></p></footer><script type=text/javascript src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script></div></div><script src=/js/vendor-bundle.min.d26509351aa0ff874abbee824e982e9b.js></script>
<script id=page-data type=application/json>{"use_headroom":true}</script><script src=/js/wowchemy-headroom.db4755770454eb63685f8de785c0a172.js type=module></script>
<script src=/en/js/wowchemy.min.e973d49cb2d568aa6f8eaf3638337473.js></script><div id=modal class="modal fade" role=dialog><div class=modal-dialog><div class=modal-content><div class=modal-header><h5 class=modal-title>Cite</h5><button type=button class=close data-dismiss=modal aria-label=Close>
<span aria-hidden=true>&#215;</span></button></div><div class=modal-body><pre><code></code></pre></div><div class=modal-footer><a class="btn btn-outline-primary my-1 js-copy-cite" href=# target=_blank><i class="fas fa-copy"></i> Copy</a>
<a class="btn btn-outline-primary my-1 js-download-cite" href=# target=_blank><i class="fas fa-download"></i> Download</a><div id=modal-error></div></div></div></div></div><script src=/js/wowchemy-publication.68f8d7090562ca65fc6d3cb3f8f2d2cb.js type=module></script></body></html>