<!doctype html><!-- This site was created with Wowchemy. https://www.wowchemy.com --><!-- Last Published: July 14, 2023 --><html lang=en-us><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta http-equiv=x-ua-compatible content="IE=edge"><meta name=generator content="Wowchemy 5.7.0 for Hugo"><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=preload as=style href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap"><link rel=stylesheet href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media=print onload='this.media="all"'><link rel=stylesheet href=/css/vendor-bundle.min.16f785cdb553c8c4431db6775122af35.css media=print onload='this.media="all"'><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/academicons@1.9.2/css/academicons.min.css integrity="sha512-KlJCpRsLf+KKu2VQa5vmRuClRFjxc5lXO03ixZt82HZUk41+1I0bD8KBSA0fY290ayMfWYI9udIqeOWSu1/uZg==" crossorigin=anonymous media=print onload='this.media="all"'><link rel=stylesheet href=/css/wowchemy.e5d7adca760216d3b7e28ea434e81f6f.css><link rel=stylesheet href=/css/libs/chroma/github-light.min.css title=hl-light media=print onload='this.media="all"'><link rel=stylesheet href=/css/libs/chroma/dracula.min.css title=hl-dark media=print onload='this.media="all"' disabled><script async src="https://www.googletagmanager.com/gtag/js?id=UA-162525500-1"></script>
<script>window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments)}function trackOutboundLink(e,t){gtag("event","click",{event_category:"outbound",event_label:e,transport_type:"beacon",event_callback:function(){t!=="_blank"&&(document.location=e)}}),console.debug("Outbound link clicked: "+e)}function onClickCallback(e){if(e.target.tagName!=="A"||e.target.host===window.location.host)return;trackOutboundLink(e.target,e.target.getAttribute("target"))}gtag("js",new Date),gtag("config","UA-162525500-1",{}),gtag("set",{cookie_flags:"SameSite=None;Secure"}),document.addEventListener("click",onClickCallback,!1)</script><script>var _hmt=_hmt||[];(function(){var e,t=document.createElement("script");t.src="https://hm.baidu.com/hm.js?67583aee7371a754cff04fa0a471bca3",e=document.getElementsByTagName("script")[0],e.parentNode.insertBefore(t,e)})()</script><meta name=author content="Hua Zou"><meta name=description content="preparing input data for machine learning."><link rel=alternate hreflang=en-us href=https://zouhua.top/post/machine_learning/2022-10-26-ml001-inputdata/><link rel=canonical href=https://zouhua.top/post/machine_learning/2022-10-26-ml001-inputdata/><link rel=manifest href=/manifest.webmanifest><link rel=icon type=image/png href=/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_32x32_fill_lanczos_center_3.png><link rel=apple-touch-icon type=image/png href=/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_180x180_fill_lanczos_center_3.png><meta name=theme-color content="#1565c0"><meta property="twitter:card" content="summary"><meta property="twitter:site" content="@Hua Zou"><meta property="twitter:creator" content="@Hua Zou"><meta property="twitter:image" content="https://zouhua.top/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_512x512_fill_lanczos_center_3.png"><meta property="og:site_name" content="Hua's Cabin"><meta property="og:url" content="https://zouhua.top/post/machine_learning/2022-10-26-ml001-inputdata/"><meta property="og:title" content="Machine Learning on gut microbiota of patients with Colorectal cancer (1) | Hua's Cabin"><meta property="og:description" content="preparing input data for machine learning."><meta property="og:image" content="https://zouhua.top/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_512x512_fill_lanczos_center_3.png"><meta property="og:locale" content="en-us"><meta property="article:published_time" content="2022-10-26T20:13:14+00:00"><meta property="article:modified_time" content="2022-10-26T22:13:14+00:00"><title>Machine Learning on gut microbiota of patients with Colorectal cancer (1) | Hua's Cabin</title></head><body id=top data-spy=scroll data-offset=70 data-target=#TableOfContents class=page-wrapper data-wc-page-id=c96342c22d42734e2730080b326d735b><script src=/js/wowchemy-init.min.ec9d49ca50e4b80bdb08f0417a28ed84.js></script><div class="page-header header--fixed"><header><nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id=navbar-main><div class=container-xl><div class="d-none d-lg-inline-flex"><a class=navbar-brand href=/>Hua's Cabin</a></div><button type=button class=navbar-toggler data-toggle=collapse data-target=#navbar-content aria-controls=navbar-content aria-expanded=false aria-label="Toggle navigation">
<span><i class="fas fa-bars"></i></span></button><div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none"><a class=navbar-brand href=/>Hua's Cabin</a></div><div class="navbar-collapse main-menu-item collapse justify-content-start" id=navbar-content><ul class="navbar-nav d-md-inline-flex"><li class=nav-item><a class=nav-link href=/#about><span>Home</span></a></li><li class=nav-item><a class=nav-link href=/#posts><span>Posts</span></a></li><li class=nav-item><a class=nav-link href=/#experience><span>Experience</span></a></li><li class=nav-item><a class=nav-link href=/#publications><span>Publications</span></a></li><li class=nav-item><a class=nav-link href=/#contact><span>Contact</span></a></li><li class=nav-item><a class=nav-link href=/#bookdown><span>Bookdown</span></a></li></ul></div><ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2"><li class=nav-item><a class="nav-link js-search" href=# aria-label=Search><i class="fas fa-search" aria-hidden=true></i></a></li><li class="nav-item dropdown theme-dropdown"><a href=# class=nav-link data-toggle=dropdown aria-haspopup=true aria-label="Display preferences"><i class="fas fa-moon" aria-hidden=true></i></a><div class=dropdown-menu><a href=# class="dropdown-item js-set-theme-light"><span>Light</span></a>
<a href=# class="dropdown-item js-set-theme-dark"><span>Dark</span></a>
<a href=# class="dropdown-item js-set-theme-auto"><span>Automatic</span></a></div></li></ul></div></nav></header></div><div class=page-body><div class="container-fluid docs"><div class="row flex-xl-nowrap"><div class="col-12 col-md-3 col-xl-2 docs-sidebar"><form class="docs-search d-flex align-items-center"><button class="btn docs-toggle d-md-none p-0 mr-md-3 w-100" type=button data-toggle=collapse data-target=#docs-nav aria-controls=docs-nav aria-expanded=false aria-label="Toggle section navigation"><div class=d-flex><span class="d-md-none pl-1 flex-grow-1 text-left overflow-hidden">Posts</span>
<span><i class="fas fa-chevron-down"></i></span></div></button></form><nav class="collapse docs-links" id=docs-nav><ul class="nav docs-sidenav"><li><a href=/post/><i class="fas fa-arrow-left pr-1"></i>Posts</a></li></ul><li class=active><a href=/post/machine_learning/2022-10-26-ml001-inputdata/>Machine Learning on gut microbiota of patients with Colorectal cancer (1)</a></li></nav></div><div class="d-none d-xl-block col-xl-2 docs-toc"><ul class="nav toc-top"><li><a href=# id=back_to_top class=docs-toc-title>Contents</a></li></ul><nav id=TableOfContents><ul><li><a href=#loading-packages>Loading packages</a></li><li><a href=#investigating-potential-response-variables>Investigating potential response variables</a></li><li><a href=#write-relative-abundance-datasets-to-disk>Write relative abundance datasets to disk</a></li><li><a href=#preparing-for-machine-learning>Preparing for machine learning</a></li><li><a href=#output>output</a></li><li><a href=#session-info>Session info</a></li><li><a href=#reference>Reference</a></li></ul></nav></div><main class="col-12 col-md-9 col-xl-8 py-md-3 pl-md-5 docs-content" role=main><div class=docs-article-container></div><div class=docs-article-container><h1>Machine Learning on gut microbiota of patients with Colorectal cancer (1)</h1><article class=article-style><h1 id=obtaining-inpudata>Obtaining inpudata</h1><p><a href=https://github.com/HuaZou/Machine-Learning-on-gut-microbiota-of-patients-with-Colorectal-cancer/blob/main/00.Obtain_inputs_for_ML.Rmd target=_blank rel=noopener>Rmarkdown 1</a>: Downloading data.</p><p>Downloading datasets using <strong>curatedMetagenomicData</strong>, which contains the HUMANN or Metaphlan results.</p><h2 id=loading-packages>Loading packages</h2><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>knitr</span><span class=o>::</span><span class=n>opts_chunk</span><span class=o>$</span><span class=nf>set</span><span class=p>(</span><span class=n>warning</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>dplyr</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>tibble</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>curatedMetagenomicData</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=c1>#library(curatedMetagenomicAnalyses)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># rm(list = ls())</span>
</span></span><span class=line><span class=cl><span class=nf>options</span><span class=p>(</span><span class=n>stringsAsFactors</span> <span class=o>=</span> <span class=bp>F</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>options</span><span class=p>(</span><span class=n>future.globals.maxSize</span> <span class=o>=</span> <span class=m>1000</span> <span class=o>*</span> <span class=m>1024</span><span class=n>^2</span><span class=p>)</span>
</span></span></code></pre></div><h2 id=investigating-potential-response-variables>Investigating potential response variables</h2><ul><li>These are the 10 study conditions most commonly found in curatedMetagenomicData:</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>data</span><span class=p>(</span><span class=s>&#34;sampleMetadata&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>availablediseases</span> <span class=o>&lt;-</span> <span class=nf>pull</span><span class=p>(</span><span class=n>sampleMetadata</span><span class=p>,</span> <span class=n>study_condition</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=nf>table</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=nf>sort</span><span class=p>(</span><span class=n>decreasing</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>availablediseases</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/image-20221101155153295.png alt loading=lazy data-zoomable></div></div></figure></p><ul><li>And the number of studies they are found in:</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>studies</span> <span class=o>&lt;-</span> <span class=nf>lapply</span><span class=p>(</span><span class=nf>names</span><span class=p>(</span><span class=n>availablediseases</span><span class=p>),</span> <span class=nf>function</span><span class=p>(</span><span class=n>x</span><span class=p>){</span>
</span></span><span class=line><span class=cl>  <span class=nf>filter</span><span class=p>(</span><span class=n>sampleMetadata</span><span class=p>,</span> <span class=n>study_condition</span> <span class=o>%in%</span> <span class=n>x</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>pull</span><span class=p>(</span><span class=n>study_name</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>unique</span><span class=p>()</span>
</span></span><span class=line><span class=cl><span class=p>})</span>
</span></span><span class=line><span class=cl><span class=nf>names</span><span class=p>(</span><span class=n>studies</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>names</span><span class=p>(</span><span class=n>availablediseases</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>studies</span> <span class=o>&lt;-</span> <span class=n>studies[</span><span class=o>-</span><span class=nf>grep</span><span class=p>(</span><span class=s>&#34;control&#34;</span><span class=p>,</span> <span class=nf>names</span><span class=p>(</span><span class=n>studies</span><span class=p>))</span><span class=n>]</span> <span class=c1>#get rid of controls</span>
</span></span><span class=line><span class=cl><span class=n>studies</span> <span class=o>&lt;-</span> <span class=n>studies</span><span class=nf>[sapply</span><span class=p>(</span><span class=n>studies</span><span class=p>,</span> <span class=n>length</span><span class=p>)</span> <span class=o>&gt;</span> <span class=m>1</span><span class=n>]</span> <span class=c1>#available in more than one study</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>studies</span>
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/image-20221101155753870.png alt loading=lazy data-zoomable></div></div></figure></p><ul><li>Each of these datasets has six data types associated with it; for example:</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>curatedMetagenomicData</span><span class=p>(</span><span class=n>pattern</span> <span class=o>=</span> <span class=s>&#34;YachidaS_2019.+&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>dryrun</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>counts</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>rownames</span> <span class=o>=</span> <span class=s>&#34;long&#34;</span><span class=p>)</span> 
</span></span></code></pre></div><p><figure><div class="d-flex justify-content-center"><div class=w-100><img src=https://raw.githubusercontent.com/HuaZou/Image_Host/main/img/image-20221101155835724.png alt loading=lazy data-zoomable></div></div></figure></p><p>The metagenomics datasets contain more than 13 types data, which comprising taxonomic and functional profile with relative and absolute abundance matrix.</p><ul><li>Relative abundance: storing into <em>TreeSummarizedExperiment</em> object</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>YachidaS_2019_dataset</span> <span class=o>&lt;-</span> <span class=nf>curatedMetagenomicData</span><span class=p>(</span><span class=n>pattern</span> <span class=o>=</span> <span class=s>&#34;YachidaS_2019.+relative_abundance&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>dryrun</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>counts</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>rownames</span> <span class=o>=</span> <span class=s>&#34;long&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>YachidaS_2019_RB_TSE</span> <span class=o>&lt;-</span> <span class=n>YachidaS_2019_dataset</span><span class=o>$</span><span class=n>`2021-10-14.YachidaS_2019.relative_abundance`</span>
</span></span><span class=line><span class=cl><span class=n>YachidaS_2019_RB_TSE</span>
</span></span></code></pre></div><h2 id=write-relative-abundance-datasets-to-disk>Write relative abundance datasets to disk</h2><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>if </span><span class=p>(</span><span class=m>0</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl><span class=nf>for </span><span class=p>(</span><span class=n>i</span> <span class=n>in</span> <span class=nf>seq_along</span><span class=p>(</span><span class=n>studies</span><span class=p>)){</span>
</span></span><span class=line><span class=cl>  <span class=n>cond</span> <span class=o>&lt;-</span> <span class=nf>names</span><span class=p>(</span><span class=n>studies</span><span class=p>)</span><span class=n>[i]</span>
</span></span><span class=line><span class=cl>  <span class=n>se</span> <span class=o>&lt;-</span>
</span></span><span class=line><span class=cl>    <span class=n>curatedMetagenomicAnalyses</span><span class=o>::</span><span class=nf>makeSEforCondition</span><span class=p>(</span><span class=n>cond</span><span class=p>,</span> <span class=n>removestudies</span> <span class=o>=</span> <span class=s>&#34;HMP_2019_ibdmdb&#34;</span><span class=p>,</span> <span class=n>dataType</span> <span class=o>=</span> <span class=s>&#34;relative_abundance&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>print</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=s>&#34;Next study condition:&#34;</span><span class=p>,</span> <span class=n>cond</span><span class=p>,</span> <span class=s>&#34; /// Body site: &#34;</span><span class=p>,</span> <span class=nf>unique</span><span class=p>(</span><span class=nf>colData</span><span class=p>(</span><span class=n>se</span><span class=p>)</span><span class=o>$</span><span class=n>body_site</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>  <span class=nf>print</span><span class=p>(</span><span class=nf>with</span><span class=p>(</span><span class=nf>colData</span><span class=p>(</span><span class=n>se</span><span class=p>),</span> <span class=nf>table</span><span class=p>(</span><span class=n>study_name</span><span class=p>,</span> <span class=n>study_condition</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>  <span class=nf>cat</span><span class=p>(</span><span class=s>&#34;\n \n&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>save</span><span class=p>(</span><span class=n>se</span><span class=p>,</span> <span class=n>file</span> <span class=o>=</span> <span class=nf>paste0</span><span class=p>(</span><span class=n>cond</span><span class=p>,</span> <span class=s>&#34;.rda&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>flattext</span> <span class=o>&lt;-</span> <span class=nf>select</span><span class=p>(</span><span class=nf>as.data.frame</span><span class=p>(</span><span class=nf>colData</span><span class=p>(</span><span class=n>se</span><span class=p>)),</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;study_name&#34;</span><span class=p>,</span> <span class=s>&#34;study_condition&#34;</span><span class=p>,</span> <span class=s>&#34;subject_id&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>rownames</span><span class=p>(</span><span class=n>flattext</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>colData</span><span class=p>(</span><span class=n>se</span><span class=p>)</span><span class=o>$</span><span class=n>sample_id</span>
</span></span><span class=line><span class=cl>  <span class=n>flattext</span> <span class=o>&lt;-</span> <span class=nf>cbind</span><span class=p>(</span><span class=n>flattext</span><span class=p>,</span> <span class=nf>data.frame</span><span class=p>(</span><span class=nf>t</span><span class=p>(</span><span class=nf>assay</span><span class=p>(</span><span class=n>se</span><span class=p>))))</span>
</span></span><span class=line><span class=cl>  <span class=nf>write.csv</span><span class=p>(</span><span class=n>flattext</span><span class=p>,</span> <span class=n>file</span> <span class=o>=</span> <span class=nf>paste0</span><span class=p>(</span><span class=n>cond</span><span class=p>,</span> <span class=s>&#34;.csv&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>system</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;gzip &#34;</span><span class=p>,</span> <span class=n>cond</span><span class=p>,</span> <span class=s>&#34;.csv&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span></code></pre></div><h2 id=preparing-for-machine-learning>Preparing for machine learning</h2><ul><li><p>metadata</p></li><li><p>relative abundance profile</p></li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>metadata</span> <span class=o>&lt;-</span> <span class=nf>colData</span><span class=p>(</span><span class=n>YachidaS_2019_RB_TSE</span><span class=p>)</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>  <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl><span class=n>phenotype</span> <span class=o>&lt;-</span> <span class=n>metadata</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>disease</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>tibble</span><span class=o>::</span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;SampleID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>disease</span> <span class=o>%in%</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;CRC&#34;</span><span class=p>,</span> <span class=s>&#34;healthy&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>profile</span> <span class=o>&lt;-</span> <span class=nf>assay</span><span class=p>(</span><span class=n>YachidaS_2019_RB_TSE</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>sid</span> <span class=o>&lt;-</span> <span class=nf>intersect</span><span class=p>(</span><span class=n>phenotype</span><span class=o>$</span><span class=n>SampleID</span><span class=p>,</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>profile</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>prof</span> <span class=o>&lt;-</span> <span class=n>profile</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>  <span class=n>tibble</span><span class=o>::</span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;TaxaID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>group_by</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>TaxaID_new</span> <span class=o>=</span> <span class=nf>unlist</span><span class=p>(</span><span class=nf>strsplit</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=s>&#34;\\|&#34;</span><span class=p>))</span><span class=n>[7]</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>TaxaID_new</span><span class=p>,</span> <span class=nf>all_of</span><span class=p>(</span><span class=n>sid</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>ungroup</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=o>-</span><span class=n>TaxaID</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>TaxaID</span> <span class=o>=</span> <span class=n>TaxaID_new</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>phen</span> <span class=o>&lt;-</span> <span class=n>phenotype</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>  <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>SampleID</span> <span class=o>%in%</span> <span class=n>sid</span><span class=p>)</span>
</span></span></code></pre></div><h2 id=output>output</h2><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=nf>dir.exists</span><span class=p>(</span><span class=s>&#34;./dataset&#34;</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  <span class=nf>dir.create</span><span class=p>(</span><span class=s>&#34;./dataset&#34;</span><span class=p>,</span> <span class=n>recursive</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>write.csv</span><span class=p>(</span><span class=n>phen</span><span class=p>,</span> <span class=s>&#34;./dataset/metadata.csv&#34;</span><span class=p>,</span> <span class=n>row.names</span> <span class=o>=</span> <span class=bp>F</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>write.table</span><span class=p>(</span><span class=n>prof</span><span class=p>,</span> <span class=s>&#34;./dataset/species.tsv&#34;</span><span class=p>,</span> <span class=n>sep</span> <span class=o>=</span> <span class=s>&#34;\t&#34;</span><span class=p>,</span> <span class=n>quote</span> <span class=o>=</span> <span class=bp>F</span><span class=p>,</span> <span class=n>row.names</span> <span class=o>=</span> <span class=bp>F</span><span class=p>)</span>
</span></span></code></pre></div><h2 id=session-info>Session info</h2><div class=highlight><pre tabindex=0 class=chroma><code class=language-R data-lang=R><span class=line><span class=cl><span class=n>devtools</span><span class=o>::</span><span class=nf>session_info</span><span class=p>()</span>
</span></span></code></pre></div><h2 id=reference>Reference</h2><ul><li><a href=https://waldronlab.io/curatedMetagenomicDataAnalyses/articles/MLdatasets.html target=_blank rel=noopener>Create datasets for machine learning</a></li></ul></article><style>.btn-feedback{display:inline-block}.btn-feedback-negative{margin-left:1em}.feedback--response{display:none;margin-top:1em}.feedback--response__visible{display:block}</style><div class="d-print-none widget--feedback"><h2 class=feedback--title>Feedback</h2><p class=feedback--question>Was this page helpful?</p><p class="feedback--response feedback--response-positive">🙏</p><p class="feedback--response feedback--response-negative">🙏</p><button class="btn btn-primary mb-4 btn-feedback btn-feedback-positive">
😍 Yes</button>
<button class="btn btn-primary mb-4 btn-feedback btn-feedback-negative">
😡 No</button></div><script>const btnYes=document.querySelector(".btn-feedback-positive"),btnNo=document.querySelector(".btn-feedback-negative"),responseYes=document.querySelector(".feedback--response-positive"),responseNo=document.querySelector(".feedback--response-negative"),disableButtons=()=>{btnYes.disabled=!0,btnNo.disabled=!0},sendFeedback=e=>{if(typeof gtag!="function")return;gtag("event","click",{event_category:"page_rating",event_label:window.location.pathname,value:e,transport_type:"beacon",event_callback:function(){console.debug(`✅ Feedback sent ${e}`)}})};btnYes.addEventListener("click",()=>{console.debug("Feedback response: 😍"),responseYes.classList.add("feedback--response__visible"),disableButtons(),sendFeedback(1)}),btnNo.addEventListener("click",()=>{console.debug("Feedback response: 😡"),responseNo.classList.add("feedback--response__visible"),disableButtons(),sendFeedback(0)})</script><div class=article-tags><a class="badge badge-light" href=/tag/machine-learning/>machine learning</a>
<a class="badge badge-light" href=/tag/microbiota/>microbiota</a></div><div class=article-widget><div class=post-nav><div class=post-nav-item><div class=meta-nav>Previous</div><a href=/post/machine_learning/2022-10-27-ml002-dataclean/ rel=next>Machine Learning on gut microbiota of patients with Colorectal cancer (2)</a></div><div class=post-nav-item><div class=meta-nav>Next</div><a href=/post/math_statistics/2022-10-27-da-methods-comparsion/ rel=prev>转录组差异分析（DESeq2+limma+edgeR+t-test/wilcox-test）总结</a></div></div></div></div><div class=body-footer><p>Last updated on Oct 26, 2022</p></div><footer class=site-footer><p class=powered-by>© 2023 Hua Zou &#183;
Rendered by <a href=https://cran.r-project.org/ target=_blank rel=noopener><i class="fab fa-r-project"></i> </a><a href=https://github.com/rstudio/blogdown target=_blank rel=noopener>blogdown</a> ,
<a href=https://sourcethemes.com/academic/ target=_blank rel=noopener>Academic theme</a> and
<a href=https://gohugo.io target=_blank rel=noopener>Hugo</a>.
<span class=float-right aria-hidden=true><a href=# id=back_to_top><span class=button_icon><i class="fas fa-chevron-up fa-2x"></i></span></a></span></p></footer><script type=text/javascript src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script></main></div></div></div><div class=page-footer></div><script src=/js/vendor-bundle.min.d26509351aa0ff874abbee824e982e9b.js></script>
<script src=https://cdn.jsdelivr.net/gh/bryanbraun/anchorjs@4.2.2/anchor.min.js integrity="sha512-I7w3ZdSFzw5j3jU3ZkNikBNeIrl3i+hEuEdwNmqUJvwNcaBUNcijnP2gd9DtGlgVYDplfjGoD8vTNsID+lCjqg==" crossorigin=anonymous></script>
<script>anchors.add()</script><script id=page-data type=application/json>{"use_headroom":false}</script><script src=/en/js/wowchemy.min.e973d49cb2d568aa6f8eaf3638337473.js></script><div id=modal class="modal fade" role=dialog><div class=modal-dialog><div class=modal-content><div class=modal-header><h5 class=modal-title>Cite</h5><button type=button class=close data-dismiss=modal aria-label=Close>
<span aria-hidden=true>&#215;</span></button></div><div class=modal-body><pre><code></code></pre></div><div class=modal-footer><a class="btn btn-outline-primary my-1 js-copy-cite" href=# target=_blank><i class="fas fa-copy"></i> Copy</a>
<a class="btn btn-outline-primary my-1 js-download-cite" href=# target=_blank><i class="fas fa-download"></i> Download</a><div id=modal-error></div></div></div></div></div><script src=/js/wowchemy-publication.68f8d7090562ca65fc6d3cb3f8f2d2cb.js type=module></script></body></html>