<!doctype html><!-- This site was created with Wowchemy. https://www.wowchemy.com --><!-- Last Published: February 7, 2024 --><html lang=en-us><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta http-equiv=x-ua-compatible content="IE=edge"><meta name=generator content="Wowchemy 5.7.0 for Hugo"><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=preload as=style href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap"><link rel=stylesheet href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&family=Roboto+Mono&family=Roboto:wght@400;700&display=swap" media=print onload='this.media="all"'><link rel=stylesheet href=/css/vendor-bundle.min.16f785cdb553c8c4431db6775122af35.css media=print onload='this.media="all"'><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/academicons@1.9.2/css/academicons.min.css integrity="sha512-KlJCpRsLf+KKu2VQa5vmRuClRFjxc5lXO03ixZt82HZUk41+1I0bD8KBSA0fY290ayMfWYI9udIqeOWSu1/uZg==" crossorigin=anonymous media=print onload='this.media="all"'><link rel=stylesheet href=/css/wowchemy.e5d7adca760216d3b7e28ea434e81f6f.css><link rel=stylesheet href=/css/libs/chroma/github-light.min.css title=hl-light media=print onload='this.media="all"'><link rel=stylesheet href=/css/libs/chroma/dracula.min.css title=hl-dark media=print onload='this.media="all"' disabled><script async src="https://www.googletagmanager.com/gtag/js?id=UA-162525500-1"></script>
<script>window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments)}function trackOutboundLink(e,t){gtag("event","click",{event_category:"outbound",event_label:e,transport_type:"beacon",event_callback:function(){t!=="_blank"&&(document.location=e)}}),console.debug("Outbound link clicked: "+e)}function onClickCallback(e){if(e.target.tagName!=="A"||e.target.host===window.location.host)return;trackOutboundLink(e.target,e.target.getAttribute("target"))}gtag("js",new Date),gtag("config","UA-162525500-1",{}),gtag("set",{cookie_flags:"SameSite=None;Secure"}),document.addEventListener("click",onClickCallback,!1)</script><script>var _hmt=_hmt||[];(function(){var e,t=document.createElement("script");t.src="https://hm.baidu.com/hm.js?67583aee7371a754cff04fa0a471bca3",e=document.getElementsByTagName("script")[0],e.parentNode.insertBefore(t,e)})()</script><meta name=author content="Hua Zou"><meta name=description content="Identifying the potential biomarkers by multiple differentail methods"><link rel=alternate hreflang=en-us href=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/><link rel=canonical href=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/><link rel=manifest href=/manifest.webmanifest><link rel=icon type=image/png href=/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_32x32_fill_lanczos_center_3.png><link rel=apple-touch-icon type=image/png href=/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_180x180_fill_lanczos_center_3.png><meta name=theme-color content="#1565c0"><meta property="twitter:card" content="summary_large_image"><meta property="twitter:site" content="@Hua Zou"><meta property="twitter:creator" content="@Hua Zou"><meta property="twitter:image" content="https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/featured.png"><meta property="og:site_name" content="Hua's Cabin"><meta property="og:url" content="https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/"><meta property="og:title" content="Multiple approaches for identification of microbial biomarkers | Hua's Cabin"><meta property="og:description" content="Identifying the potential biomarkers by multiple differentail methods"><meta property="og:image" content="https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/featured.png"><meta property="og:locale" content="en-us"><meta property="article:published_time" content="2024-02-06T00:00:00+00:00"><meta property="article:modified_time" content="2024-02-07T14:05:27+08:00"><script type=application/ld+json>{"@context":"https://schema.org","@type":"BlogPosting","mainEntityOfPage":{"@type":"WebPage","@id":"https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/"},"headline":"Multiple approaches for identification of microbial biomarkers","image":["https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/featured.png"],"datePublished":"2024-02-06T00:00:00Z","dateModified":"2024-02-07T14:05:27+08:00","author":{"@type":"Person","name":"Hua Zou"},"publisher":{"@type":"Organization","name":"Hua's Cabin","logo":{"@type":"ImageObject","url":"https://zouhua.top/media/icon_hub9a637e5d84e6d4c134f91fa57903eec_55838_192x192_fill_lanczos_center_3.png"}},"description":"Identifying the potential biomarkers by multiple differentail methods"}</script><title>Multiple approaches for identification of microbial biomarkers | Hua's Cabin</title></head><body id=top data-spy=scroll data-offset=70 data-target=#TableOfContents class=page-wrapper data-wc-page-id=4195464a7af4c20d99017ff58a97c35f><script src=/js/wowchemy-init.min.ec9d49ca50e4b80bdb08f0417a28ed84.js></script><div class="page-header header--fixed"><header><nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id=navbar-main><div class=container-xl><div class="d-none d-lg-inline-flex"><a class=navbar-brand href=/>Hua's Cabin</a></div><button type=button class=navbar-toggler data-toggle=collapse data-target=#navbar-content aria-controls=navbar-content aria-expanded=false aria-label="Toggle navigation">
<span><i class="fas fa-bars"></i></span></button><div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none"><a class=navbar-brand href=/>Hua's Cabin</a></div><div class="navbar-collapse main-menu-item collapse justify-content-start" id=navbar-content><ul class="navbar-nav d-md-inline-flex"><li class=nav-item><a class=nav-link href=/#about><span>Home</span></a></li><li class=nav-item><a class=nav-link href=/#posts><span>Posts</span></a></li><li class=nav-item><a class=nav-link href=/#experience><span>Experience</span></a></li><li class=nav-item><a class=nav-link href=/#publications><span>Publications</span></a></li><li class=nav-item><a class=nav-link href=/#contact><span>Contact</span></a></li><li class=nav-item><a class=nav-link href=/#bookdown><span>Bookdown</span></a></li></ul></div><ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2"><li class=nav-item><a class="nav-link js-search" href=# aria-label=Search><i class="fas fa-search" aria-hidden=true></i></a></li><li class="nav-item dropdown theme-dropdown"><a href=# class=nav-link data-toggle=dropdown aria-haspopup=true aria-label="Display preferences"><i class="fas fa-moon" aria-hidden=true></i></a><div class=dropdown-menu><a href=# class="dropdown-item js-set-theme-light"><span>Light</span></a>
<a href=# class="dropdown-item js-set-theme-dark"><span>Dark</span></a>
<a href=# class="dropdown-item js-set-theme-auto"><span>Automatic</span></a></div></li></ul></div></nav></header></div><div class=page-body><article class=article><div class="article-container pt-3"><h1>Multiple approaches for identification of microbial biomarkers</h1><p class=page-subtitle>Differential analysis in metagenomics</p><div class=article-metadata><span class=article-date>Last updated on
Feb 7, 2024</span>
<span class=middot-divider></span>
<span class=article-reading-time>34 min read</span>
<span class=middot-divider></span>
<span class=article-categories><i class="fas fa-folder mr-1"></i><a href=/category/statistics/>Statistics</a></span></div></div><div class="article-header article-container featured-image-wrapper mt-4 mb-4" style=max-width:720px;max-height:504px><div style=position:relative><img src=/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/featured_huef5dc820ca637b470ded6a47ec006f94_104101_720x2500_fit_q75_h2_lanczos_3.webp width=720 height=504 alt class=featured-image></div></div><div class=article-container><div class=article-style><h2 id=introduction>Introduction</h2><p>Identifying species that differ between groups is a common data analysis in the field of microbiology. We used three different types of difference analysis to discover significantly different microbial species (<a href=#ref-derosa2022intestinal>Derosa et al. 2022</a>), they are:</p><ul><li><p>Linear Discriminant Analysis of Effect Size (LefSe).</p></li><li><p>Microbiome Multivariable Association with Linear Models (Maaslin2).</p></li><li><p>Analysis of Composition of Microbiomes with Bias Correction (ANCOM-BC).</p></li></ul><p><strong>LefSe</strong> is a method based on supervised linear discriminant screening of discrepant species; <strong>Maaslin2</strong> is a method based on a linear regression algorithm for identifying discrepant species; and <strong>ANCOM-BC</strong> is a method for correcting for absolute abundance bias between samples and zero-value inflation, among others, before discovering discrepant species.</p><p>We will use the differential species identified by <strong>LefSe</strong> as a benchmark, merge the differential results of the other two methods, and finally present the results in a bar chart.</p><h2 id=microbiomeanalysis-installation>MicrobiomeAnalysis installation</h2><p><strong>MicrobiomeAnalysis</strong> will provide the following functions for downstream data analysis.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=nf>requireNamespace</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;remotes&#34;</span><span class=p>,</span> <span class=s>&#34;devtools&#34;</span><span class=p>),</span> <span class=n>quietly</span><span class=o>=</span><span class=kc>TRUE</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  <span class=nf>install.packages</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;devtools&#34;</span><span class=p>,</span> <span class=s>&#34;remotes&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=n>remotes</span><span class=o>::</span><span class=nf>install_github</span><span class=p>(</span><span class=s>&#34;HuaZou/MicrobiomeAnalysis&#34;</span><span class=p>)</span>
</span></span></code></pre></div><h2 id=loading-required-packages>Loading required packages</h2><p>R packages in this tutorial.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>knitr</span><span class=o>::</span><span class=n>opts_chunk</span><span class=o>$</span><span class=nf>set</span><span class=p>(</span><span class=n>message</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span> <span class=n>warning</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>MicrobiomeAnalysis</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>Maaslin2</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>ANCOMBC</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>phyloseq</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>ggpubr</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># rm(list = ls())</span>
</span></span><span class=line><span class=cl><span class=nf>options</span><span class=p>(</span><span class=n>stringsAsFactors</span> <span class=o>=</span> <span class=bp>F</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>options</span><span class=p>(</span><span class=n>future.globals.maxSize</span> <span class=o>=</span> <span class=m>1000</span> <span class=o>*</span> <span class=m>1024</span><span class=n>^2</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># group &amp; color</span>
</span></span><span class=line><span class=cl><span class=n>sex_grp</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Male&#34;</span><span class=p>,</span> <span class=s>&#34;Female&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>sex_col</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;#F28880&#34;</span><span class=p>,</span> <span class=s>&#34;#60C4D3&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>lf_grp</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>,</span> <span class=s>&#34;Moderate&#34;</span><span class=p>,</span> <span class=s>&#34;Severe&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>lf_col</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;#803C08&#34;</span><span class=p>,</span> <span class=s>&#34;#F1A340&#34;</span><span class=p>,</span> <span class=s>&#34;#2C0a4B&#34;</span><span class=p>,</span> <span class=s>&#34;#998EC3&#34;</span><span class=p>)</span>
</span></span></code></pre></div><h2 id=data-preparation>Data preparation</h2><p>Data were obtained from the gut microbiology data of <em>Zeybel_2022</em> and can be accessed in <strong>MicrobiomeAnalysis</strong>.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>phy</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=n>Zeybel_2022_gut</span>
</span></span><span class=line><span class=cl><span class=n>phy</span>
</span></span></code></pre></div><pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 547 taxa and 42 samples ]
## sample_data() Sample Data:       [ 42 samples by 46 sample variables ]
## tax_table()   Taxonomy Table:    [ 547 taxa by 7 taxonomic ranks ]
</code></pre><p>The data object <em>phy</em> consists of three components: an otu table in species level, a metadata table, and a table with multiple taxonomic levels.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=nf>colnames</span><span class=p>(</span><span class=n>phy</span><span class=o>@</span><span class=n>sam_data</span><span class=p>)</span><span class=n>[1</span><span class=o>:</span><span class=m>4</span><span class=n>]</span>
</span></span></code></pre></div><pre><code>## [1] &quot;Gender&quot;             &quot;Age&quot;                &quot;LiverFatClass&quot;     
## [4] &quot;AlcoholConsumption&quot;
</code></pre><p>Using <em>LiverFatClass</em> as the comparison group and <em>Gender</em>, <em>Age</em> as the confounding factors.</p><p>Because there is no absolute abundance table, the relative abundance table is converted to absolute abundance table for the time being (note: accurate absolute abundance table is required, amplicons can be used with counts, Metaphlan series can be used with counts such as FPKM).</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>phy_count</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>transform_sample_counts</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>phy</span><span class=p>,</span> <span class=nf>function</span><span class=p>(</span><span class=n>x</span><span class=p>){</span><span class=nf>round</span><span class=p>(</span><span class=n>x</span><span class=o>*</span><span class=m>10</span><span class=n>^7</span><span class=p>,</span> <span class=m>0</span><span class=p>)})</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># head(otu_table(phy_count), 4)</span>
</span></span></code></pre></div><h2 id=linear-discriminant-analysis-of-effect-size-lefse>Linear Discriminant Analysis of Effect Size (LefSe)</h2><p>Provide encapsulated function for calculation and plotting.</p><ol><li><p><em>get_lefse</em>: Obtaining the LefSe result;</p></li><li><p><em>get_lefse_pl</em>: Plotting the result;</p></li><li><p><em>get_boxplot</em>: the boxplot between groups.</p></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>get_lefse</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>taxa_level</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=kc>NULL</span><span class=p>,</span> <span class=s>&#34;Phylum&#34;</span><span class=p>,</span> <span class=s>&#34;Class&#34;</span><span class=p>,</span> <span class=s>&#34;Order&#34;</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                 <span class=s>&#34;Family&#34;</span><span class=p>,</span> <span class=s>&#34;Genus&#34;</span><span class=p>,</span> <span class=s>&#34;Species&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>filterCol</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterVars</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span> <span class=s>&#34;Gender&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>,</span> <span class=s>&#34;Moderate&#34;</span><span class=p>,</span> <span class=s>&#34;Severe&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=s>&#34;Male&#34;</span><span class=p>,</span> <span class=s>&#34;Female&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>prev_cutoff</span> <span class=o>=</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>mean_threshold</span> <span class=o>=</span> <span class=m>0.0001</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>one_threshold</span> <span class=o>=</span> <span class=m>0.001</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>LDA_cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># ps = phy</span>
</span></span><span class=line><span class=cl>  <span class=c1># taxa_level = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># filterCol = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># filterVars = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># group = &#34;LiverFatClass&#34;</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_names = c(&#34;None&#34;, &#34;Mild&#34;)</span>
</span></span><span class=line><span class=cl>  <span class=c1># prev_cutoff = 0.1</span>
</span></span><span class=line><span class=cl>  <span class=c1># mean_threshold = 0.0001</span>
</span></span><span class=line><span class=cl>  <span class=c1># one_threshold = 0.001</span>
</span></span><span class=line><span class=cl>  <span class=c1># LDA_cutoff = 0</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=nf>is.null</span><span class=p>(</span><span class=n>taxa_level</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>ps_taxa</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>aggregate_taxa</span><span class=p>(</span>
</span></span><span class=line><span class=cl>      <span class=n>x</span> <span class=o>=</span> <span class=n>ps</span><span class=p>,</span> <span class=n>level</span> <span class=o>=</span> <span class=n>taxa_level</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>ps_taxa</span> <span class=o>&lt;-</span> <span class=n>ps</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=n>ps_taxa</span><span class=o>@</span><span class=n>sam_data</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=nf>is.null</span><span class=p>(</span><span class=n>filterCol</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln</span> <span class=o>&lt;-</span> <span class=n>metadata</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span> <span class=o>==</span> <span class=n>filterCol</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;FiltCol&#34;</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln</span> <span class=o>&lt;-</span> <span class=n>metadata</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>FiltCol</span> <span class=o>%in%</span> <span class=n>filterVars</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span> <span class=o>==</span> <span class=s>&#34;FiltCol&#34;</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>filterCol</span>   
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group for test </span>
</span></span><span class=line><span class=cl>  <span class=n>dat_cln2</span> <span class=o>&lt;-</span> <span class=n>dat_cln</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln2</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln2</span><span class=p>)</span> <span class=o>==</span> <span class=n>group</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Group_new&#34;</span>
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=n>group_names[1]</span> <span class=o>==</span> <span class=s>&#34;all&#34;</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln3</span> <span class=o>&lt;-</span> <span class=n>dat_cln2</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln3</span> <span class=o>&lt;-</span> <span class=n>dat_cln2</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>Group_new</span> <span class=o>%in%</span> <span class=n>group_names</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>dat_cln3</span><span class=o>$</span><span class=n>Group_new</span> <span class=o>&lt;-</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_cln3</span><span class=o>$</span><span class=n>Group_new</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>group_names</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span> <span class=o>==</span> <span class=s>&#34;Group_new&#34;</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>ps_temp</span> <span class=o>&lt;-</span> <span class=n>ps_taxa</span>
</span></span><span class=line><span class=cl>  <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>ps_temp</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># trim &amp; filter</span>
</span></span><span class=line><span class=cl>  <span class=n>ps_trim</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>trim_prevalence</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>object</span> <span class=o>=</span> <span class=n>ps_temp</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff</span> <span class=o>=</span> <span class=n>prev_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>trim</span> <span class=o>=</span> <span class=s>&#34;feature&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>ps_filter</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>filter_abundance</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>object</span> <span class=o>=</span> <span class=n>ps_trim</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff_mean</span> <span class=o>=</span> <span class=n>mean_threshold</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff_one</span> <span class=o>=</span> <span class=n>one_threshold</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># run lefse</span>
</span></span><span class=line><span class=cl>  <span class=nf>set.seed</span><span class=p>(</span><span class=m>123</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>res_lefse</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>run_lefse</span><span class=p>(</span>
</span></span><span class=line><span class=cl>                    <span class=n>ps</span> <span class=o>=</span> <span class=n>ps_filter</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>group</span> <span class=o>=</span> <span class=n>group</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>taxa_rank</span> <span class=o>=</span> <span class=s>&#34;none&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>norm</span> <span class=o>=</span> <span class=s>&#34;CPM&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>lda_cutoff</span> <span class=o>=</span> <span class=n>LDA_cutoff</span><span class=p>)</span> 
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>res_lda</span> <span class=o>&lt;-</span> <span class=n>res_lefse</span><span class=o>@</span><span class=n>marker_table</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>inner_join</span><span class=p>(</span><span class=n>ps_filter</span><span class=o>@</span><span class=n>tax_table</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>                        <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>                        <span class=n>tibble</span><span class=o>::</span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;feature&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                      <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;feature&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1># Number of Group</span>
</span></span><span class=line><span class=cl>  <span class=n>input_metadata</span> <span class=o>&lt;-</span> <span class=n>ps_filter</span><span class=o>@</span><span class=n>sam_data</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>input_metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>input_metadata</span><span class=p>)</span> <span class=o>==</span> <span class=n>group</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Compvar&#34;</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status</span> <span class=o>&lt;-</span> <span class=nf>table</span><span class=p>(</span><span class=n>input_metadata</span><span class=o>$</span><span class=n>Compvar</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_number</span> <span class=o>&lt;-</span> <span class=nf>as.numeric</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_name</span> <span class=o>&lt;-</span> <span class=nf>names</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>res_lda</span><span class=o>$</span><span class=n>Block</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=n>dat_status_number[1]</span><span class=p>,</span> <span class=n>dat_status_name[1]</span><span class=p>,</span> <span class=n>sep</span> <span class=o>=</span> <span class=s>&#34;_&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                       <span class=s>&#34;vs&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=nf>paste</span><span class=p>(</span><span class=n>dat_status_number[2]</span><span class=p>,</span> <span class=n>dat_status_name[2]</span><span class=p>,</span> <span class=n>sep</span> <span class=o>=</span> <span class=s>&#34;_&#34;</span><span class=p>))</span>  
</span></span><span class=line><span class=cl>  <span class=n>res_DA</span> <span class=o>&lt;-</span> <span class=n>res_lda</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>TaxaID</span> <span class=o>=</span> <span class=n>feature</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>Enrichment</span> <span class=o>=</span> <span class=n>enrich_group</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>LDA_Score</span> <span class=o>=</span> <span class=n>ef_lda</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>LDA_Score</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>==</span> <span class=n>group_names[1]</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                                     <span class=o>-</span><span class=n>LDA_Score</span><span class=p>,</span> <span class=n>LDA_Score</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>Block</span><span class=p>,</span> <span class=nf>everything</span><span class=p>())</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># final results list</span>
</span></span><span class=line><span class=cl>  <span class=n>res</span> <span class=o>&lt;-</span> <span class=nf>list</span><span class=p>(</span><span class=n>ps</span> <span class=o>=</span> <span class=n>ps_filter</span><span class=p>,</span>
</span></span><span class=line><span class=cl>              <span class=n>test_res</span> <span class=o>=</span> <span class=n>res_DA</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>res</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>get_lefse_pl</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>index</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># dat = res_DA</span>
</span></span><span class=line><span class=cl>  <span class=c1># index = &#34;LDA_Score&#34;</span>
</span></span><span class=line><span class=cl>  <span class=c1># cutoff = 2</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_color = lf_col[c(1, 4)]  </span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>plot_lefse</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>da_res</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>group_names</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>x_index</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>x_index_cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>group_color</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;green&#34;</span><span class=p>,</span> <span class=s>&#34;red&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>    <span class=n>line_size</span> <span class=o>=</span> <span class=m>0.6</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>theme_text_size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>theme_title_size</span> <span class=o>=</span> <span class=m>12</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>theme_legend_size</span> <span class=o>=</span> <span class=m>12</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=c1># group</span>
</span></span><span class=line><span class=cl>    <span class=n>da_res_group_names</span> <span class=o>&lt;-</span> <span class=nf>gsub</span><span class=p>(</span><span class=s>&#34;\\d+_&#34;</span><span class=p>,</span> <span class=s>&#34;&#34;</span><span class=p>,</span> <span class=nf>unlist</span><span class=p>(</span><span class=nf>strsplit</span><span class=p>(</span><span class=n>da_res</span><span class=o>$</span><span class=n>Block[1]</span><span class=p>,</span> <span class=s>&#34; vs &#34;</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=nf>is.null</span><span class=p>(</span><span class=n>group_names</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>group_names</span> <span class=o>&lt;-</span> <span class=n>da_res_group_names</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=nf>all</span><span class=p>(</span><span class=n>group_names</span> <span class=o>==</span> <span class=n>da_res_group_names</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>        <span class=nf>message</span><span class=p>(</span><span class=s>&#34;group names are in wrong order, and reoder them&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>        <span class=n>group_names</span> <span class=o>&lt;-</span> <span class=n>da_res_group_names</span>
</span></span><span class=line><span class=cl>      <span class=p>}</span><span class=n>else</span><span class=p>{</span>
</span></span><span class=line><span class=cl>        <span class=n>group_names</span> <span class=o>&lt;-</span> <span class=n>group_names</span>
</span></span><span class=line><span class=cl>      <span class=p>}</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=n>x_index</span> <span class=o>%in%</span> <span class=nf>colnames</span><span class=p>(</span><span class=n>da_res</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=nf>stop</span><span class=p>(</span><span class=s>&#34;No x_index matched the DA results&#39; column, please check out your inputdata&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>    <span class=nf>colnames</span><span class=p>(</span><span class=n>da_res</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>da_res</span><span class=p>)</span> <span class=o>==</span> <span class=n>x_index</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Xindex&#34;</span>
</span></span><span class=line><span class=cl>    <span class=c1># significant results</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=c1># enrichment by new LDA cutoff</span>
</span></span><span class=line><span class=cl>    <span class=n>da_res</span><span class=nf>[which</span><span class=p>(</span><span class=n>da_res</span><span class=o>$</span><span class=n>Xindex</span> <span class=o>&gt;</span> <span class=n>x_index_cutoff</span><span class=p>),</span> <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[2]</span>
</span></span><span class=line><span class=cl>    <span class=n>da_res</span><span class=nf>[which</span><span class=p>(</span><span class=n>da_res</span><span class=o>$</span><span class=n>Xindex</span> <span class=o>&lt;</span> <span class=o>-</span><span class=n>x_index_cutoff</span><span class=p>),</span> <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[1]</span>
</span></span><span class=line><span class=cl>    <span class=n>da_res</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>da_res</span><span class=o>$</span><span class=n>Xindex</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>x_index_cutoff</span><span class=p>),</span> <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignif&#34;</span>
</span></span><span class=line><span class=cl>    <span class=n>df_status</span> <span class=o>&lt;-</span> <span class=nf>table</span><span class=p>(</span><span class=n>da_res</span><span class=o>$</span><span class=n>EnrichedDir</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>stats</span><span class=o>::</span><span class=nf>setNames</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;Group&#34;</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>    <span class=n>grp1_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=n>grp2_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=n>group_names[2]</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=n>nsf_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=n>legend_label</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>grp1_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                      <span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>nsf_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                      <span class=nf>paste0</span><span class=p>(</span><span class=n>group_names[2]</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>grp2_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=n>da_res_signif</span> <span class=o>&lt;-</span> <span class=n>da_res</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>arrange</span><span class=p>(</span><span class=n>Xindex</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>Xindex</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=n>x_index_cutoff</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=nf>nrow</span><span class=p>(</span><span class=n>da_res_signif</span><span class=p>)</span> <span class=o>==</span> <span class=m>0</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=nf>message</span><span class=p>(</span><span class=s>&#34;There is no significant taxa matched the threshold of LDA_Score&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=nf>is.null</span><span class=p>(</span><span class=n>group_color</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>plot_group_color</span> <span class=o>&lt;-</span> <span class=n>group_color</span>
</span></span><span class=line><span class=cl>      <span class=nf>names</span><span class=p>(</span><span class=n>plot_group_color</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=n>group_names</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span> <span class=n>else</span><span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>plot_group_color</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;green&#34;</span><span class=p>,</span> <span class=s>&#34;red&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>      <span class=nf>names</span><span class=p>(</span><span class=n>plot_group_color</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=n>group_names</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=n>dat_range</span> <span class=o>&lt;-</span> <span class=nf>range</span><span class=p>(</span><span class=n>da_res_signif</span><span class=o>$</span><span class=n>Xindex</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=n>dat_range[1]</span> <span class=o>&gt;</span> <span class=m>0</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>x_range</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span> <span class=nf>ceiling</span><span class=p>(</span><span class=n>dat_range[2]</span><span class=p>))</span>
</span></span><span class=line><span class=cl>      <span class=n>limits</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=m>0</span><span class=p>,</span> <span class=nf>range</span><span class=p>(</span><span class=n>da_res_signif</span><span class=o>$</span><span class=n>Xindex</span><span class=p>)</span><span class=n>[2]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=n>dat_range[2]</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>dat_start</span> <span class=o>&lt;-</span> <span class=nf>round</span><span class=p>(</span><span class=n>dat_range[1]</span> <span class=o>-</span> <span class=m>1</span><span class=p>)</span>
</span></span><span class=line><span class=cl>      <span class=n>x_range</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>dat_start</span><span class=p>,</span> <span class=nf>round</span><span class=p>(</span><span class=n>dat_range[2]</span><span class=p>))</span>
</span></span><span class=line><span class=cl>      <span class=n>limits</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=nf>range</span><span class=p>(</span><span class=n>da_res_signif</span><span class=o>$</span><span class=n>Xindex</span><span class=p>)</span><span class=n>[1]</span><span class=p>,</span> <span class=m>0</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=nf>all</span><span class=p>(</span><span class=n>dat_range[1]</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>,</span> <span class=n>dat_range[2]</span> <span class=o>&gt;</span> <span class=m>0</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>dat_start</span> <span class=o>&lt;-</span> <span class=nf>round</span><span class=p>(</span><span class=n>dat_range[1]</span> <span class=o>-</span> <span class=m>1</span><span class=p>)</span>
</span></span><span class=line><span class=cl>      <span class=n>x_range</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>dat_start</span><span class=p>,</span> <span class=nf>ceiling</span><span class=p>(</span><span class=n>dat_range[2]</span><span class=p>))</span>
</span></span><span class=line><span class=cl>      <span class=n>limits</span> <span class=o>&lt;-</span> <span class=n>x_range</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=n>break_scale</span> <span class=o>&lt;-</span> <span class=nf>sum</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=nf>ceiling</span><span class=p>(</span><span class=nf>range</span><span class=p>(</span><span class=n>da_res_signif</span><span class=o>$</span><span class=n>Xindex</span><span class=p>))))</span> <span class=o>/</span> <span class=m>6</span>
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=n>break_scale</span> <span class=o>&gt;</span> <span class=m>0.5</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>break_scale_final</span> <span class=o>&lt;-</span> <span class=nf>ceiling</span><span class=p>(</span><span class=n>break_scale</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>break_scale_final</span> <span class=o>&lt;-</span> <span class=nf>round</span><span class=p>(</span><span class=n>break_scale</span><span class=p>,</span> <span class=m>1</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>    <span class=n>breaks</span> <span class=o>&lt;-</span> <span class=nf>seq</span><span class=p>(</span><span class=n>x_range[1]</span><span class=p>,</span> <span class=n>x_range[2]</span><span class=p>,</span> <span class=n>break_scale_final</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>da_res_signif</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span> <span class=o>=</span> <span class=nf>reorder</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>Xindex</span><span class=p>),</span> <span class=n>y</span> <span class=o>=</span> <span class=n>Xindex</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>geom_bar</span><span class=p>(</span><span class=n>stat</span> <span class=o>=</span> <span class=s>&#34;identity&#34;</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>fill</span> <span class=o>=</span> <span class=n>Enrichment</span><span class=p>),</span>
</span></span><span class=line><span class=cl>               <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>width</span> <span class=o>=</span> <span class=m>.6</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>geom_hline</span><span class=p>(</span><span class=n>yintercept</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=n>line_size</span> <span class=o>+</span> <span class=m>0.1</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>geom_hline</span><span class=p>(</span><span class=n>yintercept</span> <span class=o>=</span> <span class=n>breaks[breaks</span> <span class=o>!=</span> <span class=m>0</span><span class=n>]</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=n>line_size</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>scale_fill_manual</span><span class=p>(</span><span class=n>values</span> <span class=o>=</span> <span class=n>plot_group_color</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>scale_y_continuous</span><span class=p>(</span><span class=n>breaks</span> <span class=o>=</span> <span class=n>breaks</span><span class=p>,</span> <span class=n>limits</span> <span class=o>=</span> <span class=n>limits</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>ylab</span><span class=p>(</span><span class=n>x_index</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>xlab</span><span class=p>(</span><span class=s>&#34;&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>coord_flip</span><span class=p>()</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>theme_bw</span><span class=p>()</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>      <span class=nf>theme</span><span class=p>(</span><span class=n>axis.ticks.length</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.4</span><span class=p>,</span> <span class=s>&#34;lines&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>axis.ticks</span> <span class=o>=</span> <span class=nf>element_line</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>axis.line</span> <span class=o>=</span> <span class=nf>element_line</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>axis.title.x</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=n>theme_title_size</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>axis.text.x</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=n>theme_text_size</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>axis.text.y</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=n>theme_text_size</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;italic&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>legend.title</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>            <span class=n>legend.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=n>theme_legend_size</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                                     <span class=n>margin</span> <span class=o>=</span> <span class=nf>margin</span><span class=p>(</span><span class=n>r</span> <span class=o>=</span> <span class=m>20</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>            <span class=n>legend.position</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>.76</span><span class=p>,</span> <span class=m>.05</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>legend.direction</span> <span class=o>=</span> <span class=s>&#34;horizontal&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>legend.key.width</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.8</span><span class=p>,</span> <span class=s>&#34;cm&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>            <span class=n>legend.key.height</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.5</span><span class=p>,</span> <span class=s>&#34;cm&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>            <span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>    <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>plot_lefse</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>da_res</span> <span class=o>=</span> <span class=n>dat</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>x_index</span> <span class=o>=</span> <span class=n>index</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>x_index_cutoff</span> <span class=o>=</span> <span class=n>cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>group_color</span> <span class=o>=</span> <span class=n>group_color</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>theme_legend_size</span> <span class=o>=</span> <span class=m>8</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme</span><span class=p>(</span><span class=n>legend.background</span> <span class=o>=</span> <span class=nf>element_rect</span><span class=p>(</span><span class=n>fill</span> <span class=o>=</span> <span class=nf>rgb</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>1</span><span class=p>,</span> <span class=m>1</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>0.001</span><span class=p>),</span> <span class=n>color</span> <span class=o>=</span> <span class=kc>NA</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># plot boxplot</span>
</span></span><span class=line><span class=cl><span class=n>get_boxplot</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>feature</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span> <span class=s>&#34;Gender&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>,</span> <span class=s>&#34;Moderate&#34;</span><span class=p>,</span> <span class=s>&#34;Severe&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=s>&#34;Male&#34;</span><span class=p>,</span> <span class=s>&#34;Female&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pl_title</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pos_cutoff</span> <span class=o>=</span> <span class=m>-0.002</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>                                  
</span></span><span class=line><span class=cl>  <span class=c1># ps = lefse_df$ps</span>
</span></span><span class=line><span class=cl>  <span class=c1># feature = &#34;s__Dorea_longicatena&#34;</span>
</span></span><span class=line><span class=cl>  <span class=c1># group = &#34;LiverFatClass&#34;</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_names = c(&#34;None&#34;, &#34;Mild&#34;)</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_color = lf_col[c(1, 2)]</span>
</span></span><span class=line><span class=cl>  <span class=c1># pl_title = &#34;Dorea_longicatena&#34;</span>
</span></span><span class=line><span class=cl>  <span class=c1># pos_cutoff = -0.002</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># metadata</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_phe</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>ps</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>dat_phe</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>dat_phe</span><span class=p>)</span> <span class=o>==</span> <span class=n>group</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Group_new&#34;</span>
</span></span><span class=line><span class=cl>  <span class=n>phen</span> <span class=o>&lt;-</span> <span class=n>dat_phe</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>Group_new</span> <span class=o>%in%</span> <span class=n>group_names</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>Group_new</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>tibble</span><span class=o>::</span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;TempRowNames&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># features</span>
</span></span><span class=line><span class=cl>  <span class=n>prof</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>otu_table</span><span class=p>(</span><span class=n>ps</span><span class=p>)</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span> <span class=nf>t</span><span class=p>()</span> <span class=o>%&gt;%</span> <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=nf>all_of</span><span class=p>(</span><span class=n>feature</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>tibble</span><span class=o>::</span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;TempRowNames&#34;</span><span class=p>)</span>    
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>plotdata</span> <span class=o>&lt;-</span> <span class=n>phen</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>inner_join</span><span class=p>(</span><span class=n>prof</span><span class=p>,</span> <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;TempRowNames&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>Group_new</span> <span class=o>=</span> <span class=nf>factor</span><span class=p>(</span><span class=n>Group_new</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>group_names</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>plotdata</span><span class=p>)</span><span class=n>[2</span><span class=o>:</span><span class=m>3</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Group&#34;</span><span class=p>,</span> <span class=s>&#34;Index&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>occ_cutoff</span> <span class=o>&lt;-</span> <span class=m>0</span>
</span></span><span class=line><span class=cl>  <span class=n>occ_fun</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span><span class=n>x</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=nf>return</span><span class=p>(</span><span class=nf>round</span><span class=p>(</span><span class=nf>length</span><span class=p>(</span><span class=n>x[x</span> <span class=o>&gt;</span> <span class=n>occ_cutoff]</span><span class=p>)</span><span class=o>/</span><span class=nf>length</span><span class=p>(</span><span class=n>x</span><span class=p>),</span> <span class=m>4</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>plotOcc</span> <span class=o>&lt;-</span> <span class=n>plotdata</span> <span class=o>|&gt;</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>group_by</span><span class=p>(</span><span class=n>Group</span><span class=p>)</span> <span class=o>|&gt;</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>summarise</span><span class=p>(</span><span class=n>occ</span> <span class=o>=</span> <span class=nf>occ_fun</span><span class=p>(</span><span class=n>Index</span><span class=p>))</span> <span class=o>|&gt;</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>occ_lab</span> <span class=o>=</span> <span class=nf>paste0</span><span class=p>(</span><span class=nf>round</span><span class=p>(</span><span class=n>occ</span><span class=p>,</span> <span class=m>3</span><span class=p>)</span> <span class=o>*</span> <span class=m>100</span><span class=p>,</span> <span class=s>&#34;%&#34;</span><span class=p>))</span> <span class=o>|&gt;</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>position</span> <span class=o>=</span> <span class=nf>min</span><span class=p>(</span><span class=n>plotdata</span><span class=o>$</span><span class=n>Index</span><span class=p>)</span> <span class=o>-</span> <span class=nf>min</span><span class=p>(</span><span class=n>plotdata</span><span class=o>$</span><span class=n>Index</span><span class=p>)</span> <span class=o>*</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>position</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=n>position</span> <span class=o>==</span> <span class=m>0</span><span class=p>,</span> <span class=n>pos_cutoff</span><span class=p>,</span> <span class=n>position</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>data</span> <span class=o>=</span> <span class=n>plotdata</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span> <span class=o>=</span> <span class=n>Group</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=n>Index</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=n>Group</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>stat_boxplot</span><span class=p>(</span><span class=n>geom</span> <span class=o>=</span> <span class=s>&#34;errorbar&#34;</span><span class=p>,</span> <span class=n>width</span> <span class=o>=</span> <span class=m>0.15</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_boxplot</span><span class=p>(</span><span class=n>width</span> <span class=o>=</span> <span class=m>.4</span><span class=p>,</span> <span class=n>outlier.shape</span> <span class=o>=</span> <span class=kc>NA</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_point</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>shape</span> <span class=o>=</span> <span class=m>5</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>labs</span><span class=p>(</span><span class=n>x</span> <span class=o>=</span> <span class=s>&#34;&#34;</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=s>&#34;Relative Abundance&#34;</span><span class=p>,</span> <span class=n>title</span> <span class=o>=</span> <span class=n>pl_title</span><span class=p>)</span> <span class=o>+</span> 
</span></span><span class=line><span class=cl>    <span class=nf>scale_y_continuous</span><span class=p>(</span><span class=n>expand</span> <span class=o>=</span> <span class=nf>expansion</span><span class=p>(</span><span class=n>mult</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>0.1</span><span class=p>,</span> <span class=m>0.1</span><span class=p>)))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_point</span><span class=p>(</span><span class=n>data</span> <span class=o>=</span> <span class=n>plotOcc</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span> <span class=o>=</span> <span class=n>Group</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=n>position</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=n>occ</span><span class=p>),</span> 
</span></span><span class=line><span class=cl>               <span class=n>show.legend</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span> <span class=n>shape</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>stroke</span> <span class=o>=</span> <span class=m>1</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_text</span><span class=p>(</span><span class=n>data</span> <span class=o>=</span> <span class=n>plotOcc</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span> <span class=o>=</span> <span class=n>Group</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=n>position</span><span class=p>,</span> <span class=n>label</span> <span class=o>=</span> <span class=n>occ_lab</span><span class=p>),</span>
</span></span><span class=line><span class=cl>              <span class=n>show.legend</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>scale_size_continuous</span><span class=p>(</span><span class=n>range</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>10</span><span class=p>,</span> <span class=m>12</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>coord_flip</span><span class=p>()</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>guides</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;none&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme_classic</span><span class=p>()</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme</span><span class=p>(</span><span class=n>plot.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>13</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>hjust</span> <span class=o>=</span> <span class=m>.5</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>axis.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>12</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>axis.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>))</span>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span></code></pre></div><ul><li>Differential Analysis from LefSe</li></ul><p>Using <em>LiverFatClass</em> as the comparison group, two subgroups, <em>None</em> and <em>Mild</em>, were selected for difference analysis.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>lefse_df</span> <span class=o>&lt;-</span> <span class=nf>get_lefse</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span> <span class=o>=</span> <span class=n>phy</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>taxa_level</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterCol</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterVars</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>prev_cutoff</span> <span class=o>=</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>mean_threshold</span> <span class=o>=</span> <span class=m>0.0001</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>one_threshold</span> <span class=o>=</span> <span class=m>0.001</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>LDA_cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>head</span><span class=p>(</span><span class=n>lefse_df</span><span class=o>$</span><span class=n>test_res[</span><span class=p>,</span> <span class=m>1</span><span class=o>:</span><span class=m>5</span><span class=n>]</span><span class=p>,</span> <span class=m>3</span><span class=p>)</span>
</span></span></code></pre></div><pre><code>##                      TaxaID             Block Enrichment LDA_Score     pvalue
## 1  s__Prevotella_sp_CAG_520 9_None vs 12_Mild       None -4.263390 0.01128208
## 2 s__Prevotella_sp_CAG_5226 9_None vs 12_Mild       None -4.023555 0.04089029
## 3  s__Prevotella_sp_AM42_24 9_None vs 12_Mild       None -3.921591 0.01972625
</code></pre><p><strong>Result</strong>: A total of 16 significantly different species were identified by the lefse approach</p><ul><li>Visualization</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>lefse_pl</span> <span class=o>&lt;-</span> <span class=nf>get_lefse_pl</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span> <span class=o>=</span> <span class=n>lefse_df</span><span class=o>$</span><span class=n>test_res</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>index</span> <span class=o>=</span> <span class=s>&#34;LDA_Score&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>cutoff</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span> <span class=o>=</span> <span class=n>lf_col</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>lefse_pl</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-8-1.png width=672><p><strong>Result</strong>: 13 types of differential bacteria were enriched in the None group and 3 types of differential bacteria were enriched in the Mild group.</p><ul><li>Boxplot showing the relative abundance of the specific bacteria</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=nf>get_boxplot</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span> <span class=o>=</span> <span class=n>lefse_df</span><span class=o>$</span><span class=n>ps</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>feature</span> <span class=o>=</span> <span class=s>&#34;s__Dorea_longicatena&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span> <span class=o>=</span> <span class=n>lf_col</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pl_title</span> <span class=o>=</span> <span class=s>&#34;Dorea_longicatena&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pos_cutoff</span> <span class=o>=</span> <span class=m>-0.002</span><span class=p>)</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-9-1.png width=480><p><strong>Result</strong>: <em>Dorea longicatena</em> had high occurrences (>90%) in both groups, and relative abundance was higher in the None group, but this phenomenon may be due to the rightmost outlier.</p><h2 id=microbiome-multivariable-association-with-linear-models-maaslin2>Microbiome Multivariable Association with Linear Models (Maaslin2)</h2><p>Provide encapsulated function for calculation and plotting.</p><ol><li><p><em>get_Maaslin2</em>: Obtaining the Maaslin2 result;</p></li><li><p><em>get_Maaslin2_pl</em>: Plotting the result.</p></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>get_Maaslin2</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>taxa_level</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=kc>NULL</span><span class=p>,</span> <span class=s>&#34;Phylum&#34;</span><span class=p>,</span> <span class=s>&#34;Class&#34;</span><span class=p>,</span> <span class=s>&#34;Order&#34;</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                 <span class=s>&#34;Family&#34;</span><span class=p>,</span> <span class=s>&#34;Genus&#34;</span><span class=p>,</span> <span class=s>&#34;Species&#34;</span><span class=p>,</span> <span class=s>&#34;Strain&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>filterCol</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterVars</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span> <span class=s>&#34;Gender&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>,</span> <span class=s>&#34;Moderate&#34;</span><span class=p>,</span> <span class=s>&#34;Severe&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=s>&#34;Male&#34;</span><span class=p>,</span> <span class=s>&#34;Female&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>prev_cutoff</span> <span class=o>=</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>mean_threshold</span> <span class=o>=</span> <span class=m>0.0001</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>one_threshold</span> <span class=o>=</span> <span class=m>0.001</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>fix_vars</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Age&#34;</span><span class=p>,</span> <span class=s>&#34;Gender&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>outputname</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># ps = phy</span>
</span></span><span class=line><span class=cl>  <span class=c1># taxa_level = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># filterCol = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># filterVars = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># group = &#34;LiverFatClass&#34;</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_names = c(&#34;None&#34;, &#34;Mild&#34;)</span>
</span></span><span class=line><span class=cl>  <span class=c1># prev_cutoff = 0.1</span>
</span></span><span class=line><span class=cl>  <span class=c1># mean_threshold = 0.0001</span>
</span></span><span class=line><span class=cl>  <span class=c1># one_threshold = 0.001</span>
</span></span><span class=line><span class=cl>  <span class=c1># fix_vars = c(&#34;Age&#34;, &#34;Gender&#34;)</span>
</span></span><span class=line><span class=cl>  <span class=c1># outputname = &#34;LiverFatClass_NM&#34;</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=nf>is.null</span><span class=p>(</span><span class=n>taxa_level</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>ps_taxa</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>aggregate_taxa</span><span class=p>(</span>
</span></span><span class=line><span class=cl>      <span class=n>x</span> <span class=o>=</span> <span class=n>ps</span><span class=p>,</span> <span class=n>level</span> <span class=o>=</span> <span class=n>taxa_level</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>ps_taxa</span> <span class=o>&lt;-</span> <span class=n>ps</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=n>ps_taxa</span><span class=o>@</span><span class=n>sam_data</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=nf>is.null</span><span class=p>(</span><span class=n>filterCol</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln</span> <span class=o>&lt;-</span> <span class=n>metadata</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span> <span class=o>==</span> <span class=n>filterCol</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;FiltCol&#34;</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln</span> <span class=o>&lt;-</span> <span class=n>metadata</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>FiltCol</span> <span class=o>%in%</span> <span class=n>filterVars</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span> <span class=o>==</span> <span class=s>&#34;FiltCol&#34;</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>filterCol</span>   
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group for test </span>
</span></span><span class=line><span class=cl>  <span class=n>dat_cln2</span> <span class=o>&lt;-</span> <span class=n>dat_cln</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln2</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln2</span><span class=p>)</span> <span class=o>==</span> <span class=n>group</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Group_new&#34;</span>
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=n>group_names[1]</span> <span class=o>==</span> <span class=s>&#34;all&#34;</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln3</span> <span class=o>&lt;-</span> <span class=n>dat_cln2</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln3</span> <span class=o>&lt;-</span> <span class=n>dat_cln2</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>Group_new</span> <span class=o>%in%</span> <span class=n>group_names</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>dat_cln3</span><span class=o>$</span><span class=n>Group_new</span> <span class=o>&lt;-</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_cln3</span><span class=o>$</span><span class=n>Group_new</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>group_names</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span> <span class=o>==</span> <span class=s>&#34;Group_new&#34;</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>ps_temp</span> <span class=o>&lt;-</span> <span class=n>ps_taxa</span>
</span></span><span class=line><span class=cl>  <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>ps_temp</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># trim &amp; filter</span>
</span></span><span class=line><span class=cl>  <span class=n>ps_trim</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>trim_prevalence</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>object</span> <span class=o>=</span> <span class=n>ps_temp</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff</span> <span class=o>=</span> <span class=n>prev_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>trim</span> <span class=o>=</span> <span class=s>&#34;feature&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>ps_filter</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>filter_abundance</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>object</span> <span class=o>=</span> <span class=n>ps_trim</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff_mean</span> <span class=o>=</span> <span class=n>mean_threshold</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff_one</span> <span class=o>=</span> <span class=n>one_threshold</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># maaslin2</span>
</span></span><span class=line><span class=cl>  <span class=n>inputdata</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>otu_table</span><span class=p>(</span><span class=n>ps_filter</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>t</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  <span class=n>input_metadata</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>ps_filter</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span> 
</span></span><span class=line><span class=cl>  <span class=c1>## output</span>
</span></span><span class=line><span class=cl>  <span class=n>abs_path</span> <span class=o>&lt;-</span> <span class=nf>getwd</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  <span class=n>outdir</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=n>abs_path</span><span class=p>,</span> <span class=s>&#34;Maaslin2/&#34;</span><span class=p>,</span> <span class=n>sep</span> <span class=o>=</span> <span class=s>&#34;/&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if</span><span class=p>(</span><span class=o>!</span><span class=nf>dir.exists</span><span class=p>(</span><span class=n>outdir</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=nf>dir.create</span><span class=p>(</span><span class=n>outdir</span><span class=p>,</span> <span class=n>recursive</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  <span class=n>output</span> <span class=o>&lt;-</span> <span class=nf>paste0</span><span class=p>(</span><span class=n>outdir</span><span class=p>,</span> <span class=n>outputname</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>if</span><span class=p>(</span><span class=o>!</span><span class=nf>dir.exists</span><span class=p>(</span><span class=n>output</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=nf>unlink</span><span class=p>(</span><span class=n>output</span><span class=p>,</span> <span class=n>recursive</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1>#################################################################</span>
</span></span><span class=line><span class=cl>  <span class=c1># how to set random or fixed effects: https://forum.biobakery.org/t/confounding-factors/154/3</span>
</span></span><span class=line><span class=cl>  <span class=n>fix_factors</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>group</span><span class=p>,</span> <span class=n>fix_vars</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># reference </span>
</span></span><span class=line><span class=cl>  <span class=n>ref_value_group</span> <span class=o>&lt;-</span> <span class=nf>paste0</span><span class=p>(</span><span class=n>group</span><span class=p>,</span> <span class=s>&#34;,&#34;</span> <span class=p>,</span><span class=n>group_names[1]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>ref_value_factor</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  <span class=nf>for </span><span class=p>(</span><span class=n>g</span> <span class=n>in</span> <span class=n>fix_vars</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=nf>is.numeric</span><span class=p>(</span><span class=n>input_metadata[</span><span class=p>,</span> <span class=n>g]</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>      <span class=n>temp_ref</span> <span class=o>&lt;-</span> <span class=nf>paste0</span><span class=p>(</span><span class=n>g</span><span class=p>,</span> <span class=s>&#34;,&#34;</span><span class=p>,</span> <span class=nf>unique</span><span class=p>(</span><span class=n>input_metadata[</span><span class=p>,</span> <span class=n>g]</span><span class=p>)</span><span class=n>[1]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>      <span class=n>ref_value_factor</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>ref_value_factor</span><span class=p>,</span> <span class=n>temp_ref</span><span class=p>)</span>      
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>ref_value_final</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=n>ref_value_group</span><span class=p>,</span> <span class=n>ref_value_factor</span><span class=p>),</span> <span class=n>collapse</span> <span class=o>=</span> <span class=s>&#34;;&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=c1>#################################################################</span>
</span></span><span class=line><span class=cl>  <span class=c1># Masslin2 (v.1.4.0) was run using Logit-transformed relative abundances </span>
</span></span><span class=line><span class=cl>  <span class=c1># that were normalized with total-sum-scaling (TSS) and using the variable of interest as a fixed effect</span>
</span></span><span class=line><span class=cl>  <span class=n>fit</span> <span class=o>&lt;-</span> <span class=n>Maaslin2</span><span class=o>::</span><span class=nf>Maaslin2</span><span class=p>(</span>
</span></span><span class=line><span class=cl>            <span class=n>input_data</span> <span class=o>=</span> <span class=n>inputdata</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>input_metadata</span> <span class=o>=</span> <span class=n>input_metadata</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>output</span> <span class=o>=</span> <span class=n>output</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>min_abundance</span> <span class=o>=</span> <span class=m>0.0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>min_prevalence</span> <span class=o>=</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>min_variance</span> <span class=o>=</span> <span class=m>0.0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>normalization</span> <span class=o>=</span> <span class=s>&#34;TSS&#34;</span><span class=p>,</span> <span class=c1># &#34;NONE&#34;, &#34;TSS&#34;, &#34;CLR&#34;, &#34;CSS&#34;, &#34;TMM&#34;</span>
</span></span><span class=line><span class=cl>            <span class=n>transform</span> <span class=o>=</span> <span class=s>&#34;LOGIT&#34;</span><span class=p>,</span> <span class=c1># &#34;NONE&#34;, &#34;LOG&#34;, &#34;LOGIT&#34;, &#34;AST&#34;</span>
</span></span><span class=line><span class=cl>            <span class=n>analysis_method</span> <span class=o>=</span> <span class=s>&#34;LM&#34;</span><span class=p>,</span> <span class=c1># &#34;LM&#34;, &#34;CPLM&#34;, &#34;NEGBIN&#34;, &#34;ZINB&#34;</span>
</span></span><span class=line><span class=cl>            <span class=n>max_significance</span> <span class=o>=</span> <span class=m>0.3</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=c1># random_effects = rand_var,</span>
</span></span><span class=line><span class=cl>            <span class=n>fixed_effects</span> <span class=o>=</span> <span class=n>fix_factors</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>correction</span> <span class=o>=</span> <span class=s>&#34;BH&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>standardize</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>cores</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>plot_heatmap</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>plot_scatter</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>heatmap_first_n</span> <span class=o>=</span> <span class=m>50</span><span class=p>,</span>
</span></span><span class=line><span class=cl>            <span class=n>reference</span> <span class=o>=</span> <span class=n>ref_value_final</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>res_temp</span> <span class=o>&lt;-</span> <span class=n>fit</span><span class=o>$</span><span class=n>results</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>slice</span><span class=p>(</span><span class=nf>grep</span><span class=p>(</span><span class=n>group_names[2]</span><span class=p>,</span> <span class=n>name</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># Number of Group</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>input_metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>input_metadata</span><span class=p>)</span> <span class=o>==</span> <span class=n>group</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Compvar&#34;</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status</span> <span class=o>&lt;-</span> <span class=nf>table</span><span class=p>(</span><span class=n>input_metadata</span><span class=o>$</span><span class=n>Compvar</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_number</span> <span class=o>&lt;-</span> <span class=nf>as.numeric</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_name</span> <span class=o>&lt;-</span> <span class=nf>names</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>res_temp</span><span class=o>$</span><span class=n>Block</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=n>dat_status_number[1]</span><span class=p>,</span> <span class=n>dat_status_name[1]</span><span class=p>,</span> <span class=n>sep</span> <span class=o>=</span> <span class=s>&#34;_&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                       <span class=s>&#34;vs&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=nf>paste</span><span class=p>(</span><span class=n>dat_status_number[2]</span><span class=p>,</span> <span class=n>dat_status_name[2]</span><span class=p>,</span> <span class=n>sep</span> <span class=o>=</span> <span class=s>&#34;_&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>res_test</span> <span class=o>&lt;-</span> <span class=n>res_temp</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>feature</span><span class=p>,</span> <span class=n>Block</span><span class=p>,</span> <span class=nf>everything</span><span class=p>())</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>inner_join</span><span class=p>(</span><span class=n>ps_filter</span><span class=o>@</span><span class=n>tax_table</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>                        <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>                        <span class=n>tibble</span><span class=o>::</span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;feature&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                      <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;feature&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>TaxaID</span> <span class=o>=</span> <span class=n>feature</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=nf>as.numeric</span><span class=p>(</span><span class=n>coef</span><span class=p>)</span> <span class=o>&gt;</span> <span class=m>0</span><span class=p>,</span> <span class=n>group_names[1]</span><span class=p>,</span> <span class=n>group_names[2]</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>Block</span><span class=p>,</span> <span class=n>metadata</span><span class=p>,</span> <span class=n>value</span><span class=p>,</span> <span class=n>Enrichment</span><span class=p>,</span> <span class=n>coef</span><span class=p>,</span> <span class=nf>everything</span><span class=p>())</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># final results list</span>
</span></span><span class=line><span class=cl>  <span class=n>res</span> <span class=o>&lt;-</span> <span class=nf>list</span><span class=p>(</span><span class=n>ps</span> <span class=o>=</span> <span class=n>ps_filter</span><span class=p>,</span>
</span></span><span class=line><span class=cl>              <span class=n>test_res</span> <span class=o>=</span> <span class=n>res_test</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>res</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># plot</span>
</span></span><span class=line><span class=cl><span class=n>get_Maaslin2_pl</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>dat</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>lg2fc_cutoff</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>pval_cutoff</span> <span class=o>=</span> <span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>qval_cutoff</span> <span class=o>=</span> <span class=m>0.8</span><span class=p>,</span> <span class=c1>#0.05,</span>
</span></span><span class=line><span class=cl>    <span class=n>group_color</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>plot_title</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># dat = Maaslin2_df$test_res</span>
</span></span><span class=line><span class=cl>  <span class=c1># lg2fc_cutoff = 0</span>
</span></span><span class=line><span class=cl>  <span class=c1># pval_cutoff = 0.05</span>
</span></span><span class=line><span class=cl>  <span class=c1># qval_cutoff = 0.8</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_color = lf_col[c(1, 2)]</span>
</span></span><span class=line><span class=cl>  <span class=c1># plot_title = &#34;None vs Mild&#34;  </span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group_names</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>&lt;-</span> <span class=nf>gsub</span><span class=p>(</span><span class=s>&#34;\\d+_&#34;</span><span class=p>,</span> <span class=s>&#34;&#34;</span><span class=p>,</span> <span class=nf>unlist</span><span class=p>(</span><span class=nf>strsplit</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>Block[1]</span><span class=p>,</span> <span class=s>&#34; vs &#34;</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># convert coef into log2foldchange</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=o>$</span><span class=n>fc</span> <span class=o>&lt;-</span> <span class=nf>exp</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>coef</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span> <span class=o>&lt;-</span> <span class=nf>log2</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>fc</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># enrichment by beta and Pvalue AdjustedPvalue</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=nf>[which</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span> <span class=o>&gt;</span> <span class=n>lg2fc_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[1]</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=nf>[which</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span> <span class=o>&lt;</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[2]</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>lg2fc_cutoff</span> <span class=o>|</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>pval</span> <span class=o>&gt;=</span> <span class=n>pval_cutoff</span> <span class=o>|</span>
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>qval</span> <span class=o>&gt;=</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignif&#34;</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=o>$</span><span class=n>EnrichedDir</span> <span class=o>&lt;-</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>EnrichedDir</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                            <span class=n>levels</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=n>group_names[2]</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># dat status </span>
</span></span><span class=line><span class=cl>  <span class=n>df_status</span> <span class=o>&lt;-</span> <span class=nf>table</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>EnrichedDir</span><span class=p>)</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>stats</span><span class=o>::</span><span class=nf>setNames</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;Group&#34;</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>grp1_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>grp2_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=n>group_names[2]</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>nsf_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>legend_label</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>grp1_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>nsf_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=nf>paste0</span><span class=p>(</span><span class=n>group_names[2]</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>grp2_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># significant features</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_signif</span> <span class=o>&lt;-</span> <span class=n>dat</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>arrange</span><span class=p>(</span><span class=n>lg2fc</span><span class=p>,</span> <span class=n>pval</span><span class=p>,</span> <span class=n>qval</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span><span class=p>)</span> <span class=o>%&gt;%</span>    
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>lg2fc</span><span class=p>)</span> <span class=o>&gt;</span> <span class=n>lg2fc_cutoff</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># color </span>
</span></span><span class=line><span class=cl>  <span class=n>plot_group_color</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>group_color[1]</span><span class=p>,</span> <span class=s>&#34;grey&#34;</span><span class=p>,</span> <span class=n>group_color[2]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>names</span><span class=p>(</span><span class=n>plot_group_color</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=n>group_names[2]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># labels</span>
</span></span><span class=line><span class=cl>  <span class=n>xlabel</span> <span class=o>&lt;-</span> <span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;Log2FC[Maaslin2] (&#34;</span><span class=p>,</span> <span class=nf>paste</span><span class=p>(</span><span class=n>group_names</span><span class=p>,</span> <span class=n>collapse</span> <span class=o>=</span> <span class=s>&#34;/&#34;</span><span class=p>),</span> <span class=s>&#34;)&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>ylable</span> <span class=o>&lt;-</span> <span class=nf>expression</span><span class=p>(</span><span class=o>-</span><span class=n>log[10</span><span class=nf>]</span><span class=p>(</span><span class=s>&#34;Pvalue (AdjustedPvalue &lt; 0.8)&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>dat</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span> <span class=o>=</span> <span class=n>lg2fc</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>pval</span><span class=p>),</span> <span class=n>color</span> <span class=o>=</span> <span class=n>EnrichedDir</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_point</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>stroke</span> <span class=o>=</span> <span class=m>1</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>scale_color_manual</span><span class=p>(</span><span class=n>name</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>values</span> <span class=o>=</span> <span class=n>plot_group_color</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>labels</span> <span class=o>=</span> <span class=n>legend_label</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>xlab</span><span class=p>(</span><span class=n>xlabel</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>ylab</span><span class=p>(</span><span class=n>ylable</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_vline</span><span class=p>(</span><span class=n>xintercept</span> <span class=o>=</span> <span class=n>lg2fc_cutoff</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.6</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_vline</span><span class=p>(</span><span class=n>xintercept</span> <span class=o>=</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.6</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>x</span> <span class=o>=</span> <span class=n>lg2fc_cutoff</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>label</span> <span class=o>=</span> <span class=n>lg2fc_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>size</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;red&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span> <span class=n>x</span> <span class=o>=</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>label</span> <span class=o>=</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>size</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;red&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>guides</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=nf>guide_legend</span><span class=p>(</span><span class=n>override.aes</span> <span class=o>=</span> <span class=nf>list</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>2</span><span class=p>)))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>scale_y_continuous</span><span class=p>(</span><span class=n>trans</span> <span class=o>=</span> <span class=s>&#34;log1p&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_hline</span><span class=p>(</span><span class=n>yintercept</span> <span class=o>=</span> <span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>pval_cutoff</span><span class=p>),</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.6</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>x</span> <span class=o>=</span> <span class=nf>min</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span><span class=p>),</span>
</span></span><span class=line><span class=cl>             <span class=n>y</span> <span class=o>=</span> <span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>pval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>             <span class=n>label</span> <span class=o>=</span> <span class=n>pval_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>size</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;red&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=n>ggrepel</span><span class=o>::</span><span class=nf>geom_text_repel</span><span class=p>(</span><span class=n>data</span> <span class=o>=</span> <span class=n>dat_signif</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=nf>aes</span><span class=p>(</span><span class=n>label</span> <span class=o>=</span> <span class=n>TaxaID</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                             <span class=n>size</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=n>max.overlaps</span> <span class=o>=</span> <span class=nf>getOption</span><span class=p>(</span><span class=s>&#34;ggrepel.max.overlaps&#34;</span><span class=p>,</span> <span class=n>default</span> <span class=o>=</span> <span class=m>80</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                             <span class=n>segment.linetype</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=n>segment.curvature</span> <span class=o>=</span> <span class=m>-1e-20</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=n>box.padding</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.35</span><span class=p>,</span> <span class=s>&#34;lines&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                             <span class=n>point.padding</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.3</span><span class=p>,</span> <span class=s>&#34;lines&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                             <span class=n>arrow</span> <span class=o>=</span> <span class=nf>arrow</span><span class=p>(</span><span class=n>length</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.005</span><span class=p>,</span> <span class=s>&#34;npc&#34;</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                             <span class=n>bg.r</span> <span class=o>=</span> <span class=m>0.15</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme_bw</span><span class=p>()</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme</span><span class=p>(</span><span class=n>axis.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>axis.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.position</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>.15</span><span class=p>,</span> <span class=m>.1</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.key.height</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.6</span><span class=p>,</span> <span class=s>&#34;cm&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>strip.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.background</span> <span class=o>=</span> <span class=nf>element_rect</span><span class=p>(</span><span class=n>fill</span> <span class=o>=</span> <span class=nf>rgb</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>1</span><span class=p>,</span> <span class=m>1</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>0.001</span><span class=p>),</span> <span class=n>color</span> <span class=o>=</span> <span class=kc>NA</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>ggtitle</span><span class=p>(</span><span class=n>plot_title</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme</span><span class=p>(</span><span class=n>plot.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>12</span><span class=p>,</span> <span class=n>hjust</span> <span class=o>=</span> <span class=m>.5</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span></code></pre></div><ul><li>Differential Analysis from Maaslin2</li></ul><p>Using <em>LiverFatClass</em> as the comparison group, two subgroups, <em>None</em> and <em>Mild</em>, were selected for difference analysis.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>Maaslin2_df</span> <span class=o>&lt;-</span> <span class=nf>get_Maaslin2</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span> <span class=o>=</span> <span class=n>phy</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>taxa_level</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterCol</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterVars</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>prev_cutoff</span> <span class=o>=</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>mean_threshold</span> <span class=o>=</span> <span class=m>0.0001</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>one_threshold</span> <span class=o>=</span> <span class=m>0.001</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>fix_vars</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Age&#34;</span><span class=p>,</span> <span class=s>&#34;Gender&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>outputname</span> <span class=o>=</span> <span class=s>&#34;LiverFatClass_NM&#34;</span><span class=p>)</span>
</span></span></code></pre></div><pre><code>## [1] &quot;Creating output folder&quot;
## 2024-02-07 10:11:52.457939 INFO::Writing function arguments to log file
## 2024-02-07 10:11:52.473346 INFO::Verifying options selected are valid
## 2024-02-07 10:11:52.512139 INFO::Determining format of input files
## 2024-02-07 10:11:52.512753 INFO::Input format is data samples as rows and metadata samples as rows
## 2024-02-07 10:11:52.515622 INFO::Formula for fixed effects: expr ~  LiverFatClass + Age + Gender
## 2024-02-07 10:11:52.517097 INFO::Filter data based on min abundance and min prevalence
## 2024-02-07 10:11:52.51776 INFO::Total samples in data: 21
## 2024-02-07 10:11:52.519618 INFO::Min samples required with min abundance for a feature not to be filtered: 2.100000
## 2024-02-07 10:11:52.522779 INFO::Total filtered features: 0
## 2024-02-07 10:11:52.523802 INFO::Filtered feature names from abundance and prevalence filtering:
## 2024-02-07 10:11:52.528166 INFO::Total filtered features with variance filtering: 0
## 2024-02-07 10:11:52.528827 INFO::Filtered feature names from variance filtering:
## 2024-02-07 10:11:52.529255 INFO::Running selected normalization method: TSS
## 2024-02-07 10:11:52.532315 INFO::Bypass z-score application to metadata
## 2024-02-07 10:11:52.53299 INFO::Running selected transform method: LOGIT
## 2024-02-07 10:11:52.536326 INFO::Running selected analysis method: LM
## 2024-02-07 10:11:52.53839 INFO::Creating cluster of 5 R processes
## 2024-02-07 10:11:54.189838 INFO::Counting total values for each feature
## 2024-02-07 10:11:54.20929 INFO::Writing residuals to file /Users/zouhua/Documents/github/MyPersonGitHub/MyOwnWebsite/content/post/Omics_Analysis/Metagenomics/2024-02-06-Differetial/Maaslin2/LiverFatClass_NM/residuals.rds
## 2024-02-07 10:11:54.211013 INFO::Writing fitted values to file /Users/zouhua/Documents/github/MyPersonGitHub/MyOwnWebsite/content/post/Omics_Analysis/Metagenomics/2024-02-06-Differetial/Maaslin2/LiverFatClass_NM/fitted.rds
## 2024-02-07 10:11:54.212362 INFO::Writing all results to file (ordered by increasing q-values): /Users/zouhua/Documents/github/MyPersonGitHub/MyOwnWebsite/content/post/Omics_Analysis/Metagenomics/2024-02-06-Differetial/Maaslin2/LiverFatClass_NM/all_results.tsv
## 2024-02-07 10:11:54.219198 INFO::Writing the significant results (those which are less than or equal to the threshold of 0.300000 ) to file (ordered by increasing q-values): /Users/zouhua/Documents/github/MyPersonGitHub/MyOwnWebsite/content/post/Omics_Analysis/Metagenomics/2024-02-06-Differetial/Maaslin2/LiverFatClass_NM/significant_results.tsv
</code></pre><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=nf>head</span><span class=p>(</span><span class=n>Maaslin2_df</span><span class=o>$</span><span class=n>test_res[</span><span class=p>,</span> <span class=m>1</span><span class=o>:</span><span class=m>3</span><span class=n>]</span><span class=p>,</span> <span class=m>5</span><span class=p>)</span>
</span></span></code></pre></div><pre><code>##                            TaxaID             Block      metadata
## 1     s__Actinomyces_graevenitzii 9_None vs 12_Mild LiverFatClass
## 2       s__Alistipes_indistinctus 9_None vs 12_Mild LiverFatClass
## 3     s__Bacteroides_intestinalis 9_None vs 12_Mild LiverFatClass
## 4 s__Bifidobacterium_adolescentis 9_None vs 12_Mild LiverFatClass
## 5        s__Bilophila_wadsworthia 9_None vs 12_Mild LiverFatClass
</code></pre><p><strong>Result</strong>: The adjusted pvalue, also known as qvalue, were all greater than 0.05, and no differential species were screened; to visualize the results, a qvalue threshold of 0.8 was used in the following figure</p><ul><li>Visualization</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>Maaslin2_pl</span> <span class=o>&lt;-</span> <span class=nf>get_Maaslin2_pl</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span> <span class=o>=</span> <span class=n>Maaslin2_df</span><span class=o>$</span><span class=n>test_res</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>lg2fc_cutoff</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pval_cutoff</span> <span class=o>=</span> <span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>qval_cutoff</span> <span class=o>=</span> <span class=m>0.8</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span> <span class=o>=</span> <span class=n>lf_col</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>plot_title</span> <span class=o>=</span> <span class=s>&#34;None vs Mild&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>Maaslin2_pl</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-12-1.png width=672><p><strong>Result</strong>: 9 types of differential bacteria were enriched in the None group and 2 types of differential bacteria were enriched in the Mild group.</p><p>It is also recommended to use MicrobiomeAnalysis built-in volcano plot <code>plot_volcano</code> function, which provides more parameters for visualizing volcano plots.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>plot_volcano</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>da_res</span> <span class=o>=</span> <span class=n>Maaslin2_df</span><span class=o>$</span><span class=n>test_res</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=n>lf_grp</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>x_index</span> <span class=o>=</span> <span class=s>&#34;coef&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>x_index_cutoff</span> <span class=o>=</span> <span class=m>0.5</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>y_index</span> <span class=o>=</span> <span class=s>&#34;pval&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>y_index_cutoff</span> <span class=o>=</span> <span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>topN</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_colors</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=n>lf_col[1]</span><span class=p>,</span> <span class=s>&#34;grey&#34;</span><span class=p>,</span> <span class=n>lf_col[2]</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>add_enrich_arrow</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>)</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-13-1.png width=672><ul><li>Boxplot showing the relative abundance of the specific bacteria</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=nf>get_boxplot</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span> <span class=o>=</span> <span class=n>Maaslin2_df</span><span class=o>$</span><span class=n>ps</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>feature</span> <span class=o>=</span> <span class=s>&#34;s__Ruminococcus_torques&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span> <span class=o>=</span> <span class=n>lf_col</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pl_title</span> <span class=o>=</span> <span class=s>&#34;Ruminococcus_torques&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pos_cutoff</span> <span class=o>=</span> <span class=m>-0.002</span><span class=p>)</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-14-1.png width=480><p><strong>Result</strong>: <em>Ruminococcus_torques</em> was present at high rates in all None groups and its relative abundance was higher in the None group.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=nf>get_boxplot</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span> <span class=o>=</span> <span class=n>Maaslin2_df</span><span class=o>$</span><span class=n>ps</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>feature</span> <span class=o>=</span> <span class=s>&#34;s__Streptococcus_oralis&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span> <span class=o>=</span> <span class=n>lf_col</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pl_title</span> <span class=o>=</span> <span class=s>&#34;Streptococcus_oralis&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pos_cutoff</span> <span class=o>=</span> <span class=m>-0.002</span><span class=p>)</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-15-1.png width=480><p><strong>Result</strong>: <em>Streptococcus_oralis</em> was present at a high rate in both Mild groups and its relative abundance was higher in the Mild group.</p><h2 id=analysis-of-composition-of-microbiomes-with-bias-correction-ancom-bc>Analysis of Composition of Microbiomes with Bias Correction (ANCOM-BC)</h2><p>Provide encapsulated function for calculation and plotting.</p><ol><li><p><em>get_ANCOMBC</em>: Obtaining the ANCOMBC result;</p></li><li><p><em>get_ANCOMBC_pl</em>: Plotting the result.</p></li></ol><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>get_ANCOMBC</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>taxa_level</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=kc>NULL</span><span class=p>,</span> <span class=s>&#34;Phylum&#34;</span><span class=p>,</span> <span class=s>&#34;Class&#34;</span><span class=p>,</span> <span class=s>&#34;Order&#34;</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>                 <span class=s>&#34;Family&#34;</span><span class=p>,</span> <span class=s>&#34;Genus&#34;</span><span class=p>,</span> <span class=s>&#34;Species&#34;</span><span class=p>,</span> <span class=s>&#34;Strain&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>filterCol</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterVars</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span> <span class=s>&#34;Gender&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>,</span> <span class=s>&#34;Moderate&#34;</span><span class=p>,</span> <span class=s>&#34;Severe&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=s>&#34;Male&#34;</span><span class=p>,</span> <span class=s>&#34;Female&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>prev_cutoff</span> <span class=o>=</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>mean_threshold</span> <span class=o>=</span> <span class=m>100</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>one_threshold</span> <span class=o>=</span> <span class=m>1000</span><span class=p>,</span>  
</span></span><span class=line><span class=cl>  <span class=n>fix_vars</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Age&#34;</span><span class=p>,</span> <span class=s>&#34;Gender&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># ps = phy_count</span>
</span></span><span class=line><span class=cl>  <span class=c1># taxa_level = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># filterCol = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># filterVars = NULL</span>
</span></span><span class=line><span class=cl>  <span class=c1># group = &#34;LiverFatClass&#34;</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_names = c(&#34;None&#34;, &#34;Mild&#34;)</span>
</span></span><span class=line><span class=cl>  <span class=c1># prev_cutoff = 0.1</span>
</span></span><span class=line><span class=cl>  <span class=c1># mean_threshold = 100</span>
</span></span><span class=line><span class=cl>  <span class=c1># one_threshold = 1000  </span>
</span></span><span class=line><span class=cl>  <span class=c1># fix_vars = c(&#34;Age&#34;, &#34;Gender&#34;)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=o>!</span><span class=nf>is.null</span><span class=p>(</span><span class=n>taxa_level</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>ps_taxa</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>aggregate_taxa</span><span class=p>(</span>
</span></span><span class=line><span class=cl>      <span class=n>x</span> <span class=o>=</span> <span class=n>ps</span><span class=p>,</span> <span class=n>level</span> <span class=o>=</span> <span class=n>taxa_level</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>ps_taxa</span> <span class=o>&lt;-</span> <span class=n>ps</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>metadata</span> <span class=o>&lt;-</span> <span class=n>ps_taxa</span><span class=o>@</span><span class=n>sam_data</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=nf>is.null</span><span class=p>(</span><span class=n>filterCol</span><span class=p>))</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln</span> <span class=o>&lt;-</span> <span class=n>metadata</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span> <span class=o>==</span> <span class=n>filterCol</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;FiltCol&#34;</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln</span> <span class=o>&lt;-</span> <span class=n>metadata</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>FiltCol</span> <span class=o>%in%</span> <span class=n>filterVars</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>metadata</span><span class=p>)</span> <span class=o>==</span> <span class=s>&#34;FiltCol&#34;</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>filterCol</span>   
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group for test </span>
</span></span><span class=line><span class=cl>  <span class=n>dat_cln2</span> <span class=o>&lt;-</span> <span class=n>dat_cln</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln2</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln2</span><span class=p>)</span> <span class=o>==</span> <span class=n>group</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Group_new&#34;</span>
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=n>group_names[1]</span> <span class=o>==</span> <span class=s>&#34;all&#34;</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln3</span> <span class=o>&lt;-</span> <span class=n>dat_cln2</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>dat_cln3</span> <span class=o>&lt;-</span> <span class=n>dat_cln2</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>Group_new</span> <span class=o>%in%</span> <span class=n>group_names</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>dat_cln3</span><span class=o>$</span><span class=n>Group_new</span> <span class=o>&lt;-</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat_cln3</span><span class=o>$</span><span class=n>Group_new</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>group_names</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span> <span class=o>==</span> <span class=s>&#34;Group_new&#34;</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>ps_temp</span> <span class=o>&lt;-</span> <span class=n>ps_taxa</span>
</span></span><span class=line><span class=cl>  <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>ps_temp</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>dat_cln3</span><span class=p>)</span>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># trim &amp; filter</span>
</span></span><span class=line><span class=cl>  <span class=n>ps_trim</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>trim_prevalence</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>object</span> <span class=o>=</span> <span class=n>ps_temp</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff</span> <span class=o>=</span> <span class=n>prev_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>trim</span> <span class=o>=</span> <span class=s>&#34;feature&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>ps_filter</span> <span class=o>&lt;-</span> <span class=n>MicrobiomeAnalysis</span><span class=o>::</span><span class=nf>filter_abundance</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>object</span> <span class=o>=</span> <span class=n>ps_trim</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff_mean</span> <span class=o>=</span> <span class=n>mean_threshold</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>cutoff_one</span> <span class=o>=</span> <span class=n>one_threshold</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>ps_filter</span> <span class=o>&lt;-</span> <span class=n>ps_trim</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1>#################################################################</span>
</span></span><span class=line><span class=cl>  <span class=c1># https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC.html</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>  <span class=n>fix_factors</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>group</span><span class=p>,</span> <span class=n>fix_vars</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>out</span> <span class=o>&lt;-</span> <span class=n>ANCOMBC</span><span class=o>::</span><span class=nf>ancombc</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>phyloseq</span> <span class=o>=</span> <span class=n>ps_filter</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>formula</span> <span class=o>=</span> <span class=nf>paste</span><span class=p>(</span><span class=n>fix_factors</span><span class=p>,</span> <span class=n>collapse</span> <span class=o>=</span> <span class=s>&#34; + &#34;</span><span class=p>),</span> 
</span></span><span class=line><span class=cl>    <span class=n>p_adj_method</span> <span class=o>=</span> <span class=s>&#34;BH&#34;</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>    <span class=n>prv_cut</span> <span class=o>=</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>lib_cut</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>group</span> <span class=o>=</span> <span class=n>group</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>struc_zero</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>neg_lb</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>tol</span> <span class=o>=</span> <span class=m>1e-05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>max_iter</span> <span class=o>=</span> <span class=m>100</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>conserve</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>alpha</span> <span class=o>=</span> <span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>global</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>n_cl</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>verbose</span> <span class=o>=</span> <span class=kc>FALSE</span><span class=p>)</span>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># Result from the ANCOM-BC log-linear model to determine taxa that are differentially abundant according to the covariate of interest. </span>
</span></span><span class=line><span class=cl>  <span class=c1># It contains: 1) log fold changes; 2) standard errors; </span>
</span></span><span class=line><span class=cl>  <span class=c1># 3) test statistics; 4) p-values; 5) adjusted p-values; </span>
</span></span><span class=line><span class=cl>  <span class=c1># 6) indicators whether the taxon is differentially abundant (TRUE) or not (FALSE)  </span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>res_list</span> <span class=o>&lt;-</span> <span class=n>out</span><span class=o>$</span><span class=n>res</span>
</span></span><span class=line><span class=cl>  <span class=n>res_temp</span> <span class=o>&lt;-</span> <span class=nf>cbind</span><span class=p>(</span><span class=n>res_list</span><span class=o>$</span><span class=n>lfc</span><span class=o>$</span><span class=n>taxon</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>res_list</span><span class=o>$</span><span class=n>lfc[3]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>res_list</span><span class=o>$</span><span class=n>se[3]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>res_list</span><span class=o>$</span><span class=n>W[3]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>res_list</span><span class=o>$</span><span class=n>p_val[3]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>res_list</span><span class=o>$</span><span class=n>q_val[3]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                    <span class=n>res_list</span><span class=o>$</span><span class=n>diff_abn[3]</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>setNames</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;feature&#34;</span><span class=p>,</span> <span class=s>&#34;lfc&#34;</span><span class=p>,</span> <span class=s>&#34;se&#34;</span><span class=p>,</span> <span class=s>&#34;W&#34;</span><span class=p>,</span> <span class=s>&#34;p_val&#34;</span><span class=p>,</span> <span class=s>&#34;q_val&#34;</span><span class=p>,</span> <span class=s>&#34;diff_abu&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># Number of Group</span>
</span></span><span class=line><span class=cl>  <span class=n>input_metadata</span> <span class=o>&lt;-</span> <span class=n>phyloseq</span><span class=o>::</span><span class=nf>sample_data</span><span class=p>(</span><span class=n>ps_filter</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span>   
</span></span><span class=line><span class=cl>  <span class=nf>colnames</span><span class=p>(</span><span class=n>input_metadata</span><span class=p>)</span><span class=nf>[which</span><span class=p>(</span><span class=nf>colnames</span><span class=p>(</span><span class=n>input_metadata</span><span class=p>)</span> <span class=o>==</span> <span class=n>group</span><span class=p>)</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Compvar&#34;</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status</span> <span class=o>&lt;-</span> <span class=nf>table</span><span class=p>(</span><span class=n>input_metadata</span><span class=o>$</span><span class=n>Compvar</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_number</span> <span class=o>&lt;-</span> <span class=nf>as.numeric</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_status_name</span> <span class=o>&lt;-</span> <span class=nf>names</span><span class=p>(</span><span class=n>dat_status</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>res_temp</span><span class=o>$</span><span class=n>Block</span> <span class=o>&lt;-</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=n>dat_status_number[1]</span><span class=p>,</span> <span class=n>dat_status_name[1]</span><span class=p>,</span> <span class=n>sep</span> <span class=o>=</span> <span class=s>&#34;_&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                       <span class=s>&#34;vs&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=nf>paste</span><span class=p>(</span><span class=n>dat_status_number[2]</span><span class=p>,</span> <span class=n>dat_status_name[2]</span><span class=p>,</span> <span class=n>sep</span> <span class=o>=</span> <span class=s>&#34;_&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>res_test</span> <span class=o>&lt;-</span> <span class=n>res_temp</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>feature</span><span class=p>,</span> <span class=n>Block</span><span class=p>,</span> <span class=nf>everything</span><span class=p>())</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>inner_join</span><span class=p>(</span><span class=n>ps_filter</span><span class=o>@</span><span class=n>tax_table</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>                        <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>                        <span class=n>tibble</span><span class=o>::</span><span class=nf>rownames_to_column</span><span class=p>(</span><span class=s>&#34;feature&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                      <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;feature&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>TaxaID</span> <span class=o>=</span> <span class=n>feature</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>Enrichment</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=nf>as.numeric</span><span class=p>(</span><span class=n>lfc</span><span class=p>)</span> <span class=o>&gt;</span> <span class=m>0</span><span class=p>,</span> <span class=n>group_names[2]</span><span class=p>,</span> <span class=n>group_names[1]</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>Block</span><span class=p>,</span> <span class=n>lfc</span><span class=p>,</span> <span class=n>Enrichment</span><span class=p>,</span> <span class=nf>everything</span><span class=p>())</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># final results list</span>
</span></span><span class=line><span class=cl>  <span class=n>res</span> <span class=o>&lt;-</span> <span class=nf>list</span><span class=p>(</span><span class=n>ps</span> <span class=o>=</span> <span class=n>ps_filter</span><span class=p>,</span>
</span></span><span class=line><span class=cl>              <span class=n>test_res</span> <span class=o>=</span> <span class=n>res_test</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>res</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># plot</span>
</span></span><span class=line><span class=cl><span class=n>get_ANCOMBC_pl</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>dat</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>lg2fc_cutoff</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>pval_cutoff</span> <span class=o>=</span> <span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>qval_cutoff</span> <span class=o>=</span> <span class=m>0.8</span><span class=p>,</span> <span class=c1>#0.05,</span>
</span></span><span class=line><span class=cl>    <span class=n>group_color</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>plot_title</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group_names</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>&lt;-</span> <span class=nf>gsub</span><span class=p>(</span><span class=s>&#34;\\d+_&#34;</span><span class=p>,</span> <span class=s>&#34;&#34;</span><span class=p>,</span> <span class=nf>unlist</span><span class=p>(</span><span class=nf>strsplit</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>Block[1]</span><span class=p>,</span> <span class=s>&#34; vs &#34;</span><span class=p>)))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span> <span class=o>&lt;-</span> <span class=n>dat</span><span class=o>$</span><span class=n>lfc</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=o>$</span><span class=n>pval</span> <span class=o>&lt;-</span> <span class=n>dat</span><span class=o>$</span><span class=n>p_val</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=o>$</span><span class=n>qval</span> <span class=o>&lt;-</span> <span class=n>dat</span><span class=o>$</span><span class=n>q_val</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># enrichment by beta and Pvalue AdjustedPvalue</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=nf>[which</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span> <span class=o>&gt;</span> <span class=n>lg2fc_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[2]</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=nf>[which</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span> <span class=o>&lt;</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[1]</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>lg2fc_cutoff</span> <span class=o>|</span> 
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>pval</span> <span class=o>&gt;=</span> <span class=n>pval_cutoff</span> <span class=o>|</span>
</span></span><span class=line><span class=cl>              <span class=n>dat</span><span class=o>$</span><span class=n>qval</span> <span class=o>&gt;=</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;EnrichedDir&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignif&#34;</span>   
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>dat</span><span class=o>$</span><span class=n>EnrichedDir</span> <span class=o>&lt;-</span> <span class=nf>factor</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>EnrichedDir</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                            <span class=n>levels</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=n>group_names[2]</span><span class=p>,</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=n>group_names[1]</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># dat status </span>
</span></span><span class=line><span class=cl>  <span class=n>df_status</span> <span class=o>&lt;-</span> <span class=nf>table</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>EnrichedDir</span><span class=p>)</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>    <span class=nf>data.frame</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>stats</span><span class=o>::</span><span class=nf>setNames</span><span class=p>(</span><span class=nf>c</span><span class=p>(</span><span class=s>&#34;Group&#34;</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>grp1_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=n>group_names[2]</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>grp2_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>nsf_number</span> <span class=o>&lt;-</span> <span class=nf>with</span><span class=p>(</span><span class=n>df_status</span><span class=p>,</span> <span class=n>df_status[Group</span> <span class=o>%in%</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34;Number&#34;</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>legend_label</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=nf>paste0</span><span class=p>(</span><span class=n>group_names[2]</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>grp1_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>nsf_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                    <span class=nf>paste0</span><span class=p>(</span><span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>grp2_number</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># significant features</span>
</span></span><span class=line><span class=cl>  <span class=n>dat_signif</span> <span class=o>&lt;-</span> <span class=n>dat</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>arrange</span><span class=p>(</span><span class=n>lg2fc</span><span class=p>,</span> <span class=n>pval</span><span class=p>,</span> <span class=n>qval</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span><span class=p>)</span> <span class=o>%&gt;%</span>    
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=n>qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>filter</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>lg2fc</span><span class=p>)</span> <span class=o>&gt;</span> <span class=n>lg2fc_cutoff</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># color </span>
</span></span><span class=line><span class=cl>  <span class=n>plot_group_color</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>group_color[1]</span><span class=p>,</span> <span class=s>&#34;grey&#34;</span><span class=p>,</span> <span class=n>group_color[2]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=nf>names</span><span class=p>(</span><span class=n>plot_group_color</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>c</span><span class=p>(</span><span class=n>group_names[1]</span><span class=p>,</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=n>group_names[2]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># labels</span>
</span></span><span class=line><span class=cl>  <span class=n>xlabel</span> <span class=o>&lt;-</span> <span class=nf>paste0</span><span class=p>(</span><span class=s>&#34;Log2FC[ANCOMBC] (&#34;</span><span class=p>,</span> <span class=nf>paste</span><span class=p>(</span><span class=nf>rev</span><span class=p>(</span><span class=n>group_names</span><span class=p>),</span> <span class=n>collapse</span> <span class=o>=</span> <span class=s>&#34;/&#34;</span><span class=p>),</span> <span class=s>&#34;)&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  <span class=n>ylable</span> <span class=o>&lt;-</span> <span class=nf>expression</span><span class=p>(</span><span class=o>-</span><span class=n>log[10</span><span class=nf>]</span><span class=p>(</span><span class=s>&#34;Pvalue (AdjustedPvalue &lt; 0.25)&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>dat</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span> <span class=o>=</span> <span class=n>lg2fc</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>pval</span><span class=p>),</span> <span class=n>color</span> <span class=o>=</span> <span class=n>EnrichedDir</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_point</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span> <span class=n>stroke</span> <span class=o>=</span> <span class=m>1</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>scale_color_manual</span><span class=p>(</span><span class=n>name</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>values</span> <span class=o>=</span> <span class=n>plot_group_color</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                       <span class=n>labels</span> <span class=o>=</span> <span class=n>legend_label</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>xlab</span><span class=p>(</span><span class=n>xlabel</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>ylab</span><span class=p>(</span><span class=n>ylable</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_vline</span><span class=p>(</span><span class=n>xintercept</span> <span class=o>=</span> <span class=n>lg2fc_cutoff</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.6</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_vline</span><span class=p>(</span><span class=n>xintercept</span> <span class=o>=</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.6</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>x</span> <span class=o>=</span> <span class=n>lg2fc_cutoff</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>label</span> <span class=o>=</span> <span class=n>lg2fc_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>size</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;red&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span> <span class=n>x</span> <span class=o>=</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>label</span> <span class=o>=</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>size</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;red&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>guides</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=nf>guide_legend</span><span class=p>(</span><span class=n>override.aes</span> <span class=o>=</span> <span class=nf>list</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>2</span><span class=p>)))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>scale_y_continuous</span><span class=p>(</span><span class=n>trans</span> <span class=o>=</span> <span class=s>&#34;log1p&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_hline</span><span class=p>(</span><span class=n>yintercept</span> <span class=o>=</span> <span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>pval_cutoff</span><span class=p>),</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>.8</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=m>2</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>0.6</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>annotate</span><span class=p>(</span><span class=s>&#34;text&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>x</span> <span class=o>=</span> <span class=nf>min</span><span class=p>(</span><span class=n>dat</span><span class=o>$</span><span class=n>lg2fc</span><span class=p>),</span>
</span></span><span class=line><span class=cl>             <span class=n>y</span> <span class=o>=</span> <span class=o>-</span><span class=nf>log10</span><span class=p>(</span><span class=n>pval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>             <span class=n>label</span> <span class=o>=</span> <span class=n>pval_cutoff</span><span class=p>,</span>
</span></span><span class=line><span class=cl>             <span class=n>size</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;red&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=n>ggrepel</span><span class=o>::</span><span class=nf>geom_text_repel</span><span class=p>(</span><span class=n>data</span> <span class=o>=</span> <span class=n>dat_signif</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=nf>aes</span><span class=p>(</span><span class=n>label</span> <span class=o>=</span> <span class=n>TaxaID</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                             <span class=n>size</span> <span class=o>=</span> <span class=m>5</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=n>max.overlaps</span> <span class=o>=</span> <span class=nf>getOption</span><span class=p>(</span><span class=s>&#34;ggrepel.max.overlaps&#34;</span><span class=p>,</span> <span class=n>default</span> <span class=o>=</span> <span class=m>80</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                             <span class=n>segment.linetype</span> <span class=o>=</span> <span class=m>1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=n>segment.curvature</span> <span class=o>=</span> <span class=m>-1e-20</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                             <span class=n>box.padding</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.35</span><span class=p>,</span> <span class=s>&#34;lines&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                             <span class=n>point.padding</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.3</span><span class=p>,</span> <span class=s>&#34;lines&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                             <span class=n>arrow</span> <span class=o>=</span> <span class=nf>arrow</span><span class=p>(</span><span class=n>length</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.005</span><span class=p>,</span> <span class=s>&#34;npc&#34;</span><span class=p>)),</span>
</span></span><span class=line><span class=cl>                             <span class=n>bg.r</span> <span class=o>=</span> <span class=m>0.15</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme_bw</span><span class=p>()</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme</span><span class=p>(</span><span class=n>axis.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>10</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>axis.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.position</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>.15</span><span class=p>,</span> <span class=m>.1</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.key.height</span> <span class=o>=</span> <span class=nf>unit</span><span class=p>(</span><span class=m>0.6</span><span class=p>,</span> <span class=s>&#34;cm&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>9</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;black&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>strip.text</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>size</span> <span class=o>=</span> <span class=m>8</span><span class=p>,</span> <span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.background</span> <span class=o>=</span> <span class=nf>element_rect</span><span class=p>(</span><span class=n>fill</span> <span class=o>=</span> <span class=nf>rgb</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>1</span><span class=p>,</span> <span class=m>1</span><span class=p>,</span> <span class=n>alpha</span> <span class=o>=</span> <span class=m>0.001</span><span class=p>),</span> <span class=n>color</span> <span class=o>=</span> <span class=kc>NA</span><span class=p>))</span> <span class=o>+</span> <span class=c1># legend background transparency</span>
</span></span><span class=line><span class=cl>    <span class=nf>ggtitle</span><span class=p>(</span><span class=n>plot_title</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme</span><span class=p>(</span><span class=n>plot.title</span> <span class=o>=</span> <span class=nf>element_text</span><span class=p>(</span><span class=n>face</span> <span class=o>=</span> <span class=s>&#34;bold&#34;</span><span class=p>,</span> <span class=n>size</span> <span class=o>=</span> <span class=m>12</span><span class=p>,</span> <span class=n>hjust</span> <span class=o>=</span> <span class=m>.5</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span></code></pre></div><ul><li>Differential Analysis from ANCOMBC</li></ul><p>Using <em>LiverFatClass</em> as the comparison group, two subgroups, <em>None</em> and <em>Mild</em>, were selected for difference analysis.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>ANCOMBC_df</span> <span class=o>&lt;-</span> <span class=nf>get_ANCOMBC</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span> <span class=o>=</span> <span class=n>phy_count</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>taxa_level</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterCol</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>filterVars</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>prev_cutoff</span> <span class=o>=</span> <span class=m>0.1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>mean_threshold</span> <span class=o>=</span> <span class=m>100</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>one_threshold</span> <span class=o>=</span> <span class=m>1000</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>fix_vars</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;Age&#34;</span><span class=p>,</span> <span class=s>&#34;Gender&#34;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=nf>head</span><span class=p>(</span><span class=n>ANCOMBC_df</span><span class=o>$</span><span class=n>test_res[</span><span class=p>,</span> <span class=m>1</span><span class=o>:</span><span class=m>3</span><span class=n>]</span><span class=p>,</span> <span class=m>5</span><span class=p>)</span>
</span></span></code></pre></div><pre><code>##                            TaxaID             Block         lfc
## 1     s__Actinomyces_graevenitzii 9_None vs 12_Mild -0.69410607
## 2    s__Actinomyces_odontolyticus 9_None vs 12_Mild  0.07148274
## 3         s__Actinomyces_sp_ICM47 9_None vs 12_Mild  1.80267320
## 4 s__Bifidobacterium_adolescentis 9_None vs 12_Mild -1.83673365
## 5    s__Bifidobacterium_angulatum 9_None vs 12_Mild  2.10805876
</code></pre><p><strong>Result</strong>: The corrected pval, also known as qval, were all greater than 0.05, and no differential species were screened; a qvalue threshold of 0.25 was used in the lower panel in order to visualize the results.</p><ul><li>Visualization</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>ANCOMBC_pl</span> <span class=o>&lt;-</span> <span class=nf>get_ANCOMBC_pl</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>dat</span> <span class=o>=</span> <span class=n>ANCOMBC_df</span><span class=o>$</span><span class=n>test_res</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>lg2fc_cutoff</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pval_cutoff</span> <span class=o>=</span> <span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>qval_cutoff</span> <span class=o>=</span> <span class=m>0.25</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span> <span class=o>=</span> <span class=n>lf_col</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>plot_title</span> <span class=o>=</span> <span class=s>&#34;None vs Mild&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>ANCOMBC_pl</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-18-1.png width=672><p><strong>Result</strong>: 14 types of differential bacteria were enriched in the None group and 7 types of differential bacteria were enriched in the Mild group.</p><ul><li>Boxplot showing the relative abundance of the specific bacteria</li></ul><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=nf>get_boxplot</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>ps</span> <span class=o>=</span> <span class=n>ANCOMBC_df</span><span class=o>$</span><span class=n>ps</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>feature</span> <span class=o>=</span> <span class=s>&#34;s__Clostridium_leptum&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group</span> <span class=o>=</span> <span class=s>&#34;LiverFatClass&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=s>&#34;None&#34;</span><span class=p>,</span> <span class=s>&#34;Mild&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>  <span class=n>group_color</span> <span class=o>=</span> <span class=n>lf_col</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pl_title</span> <span class=o>=</span> <span class=s>&#34;Clostridium_leptum&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pos_cutoff</span> <span class=o>=</span> <span class=m>-200</span><span class=p>)</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-19-1.png width=480><p><strong>Result</strong>: <em>Clostridium_leptum</em> had high occurrences in both Mild groups, and relative abundance was higher in the Mild group.</p><h2 id=consolidataion-of-the-above-results>Consolidataion of the above results</h2><p>The results of lefse were used as a benchmark to combine the results of the three analytical methods mentioned above.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>get_mergeRes</span> <span class=o>&lt;-</span> <span class=nf>function</span><span class=p>(</span>
</span></span><span class=line><span class=cl>    <span class=n>DA1</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>DA2</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>DA3</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>lg2fc_cutoff</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>pval_cutoff</span> <span class=o>=</span> <span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>qval_cutoff</span> <span class=o>=</span> <span class=m>0.3</span><span class=p>,</span>
</span></span><span class=line><span class=cl>    <span class=n>group_colors</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># DA1 = lefse_df$test_res</span>
</span></span><span class=line><span class=cl>  <span class=c1># DA2 = Maaslin2_df$test_res</span>
</span></span><span class=line><span class=cl>  <span class=c1># DA3 = ANCOMBC_df$test_res</span>
</span></span><span class=line><span class=cl>  <span class=c1># lg2fc_cutoff = 0</span>
</span></span><span class=line><span class=cl>  <span class=c1># pval_cutoff = 0.05</span>
</span></span><span class=line><span class=cl>  <span class=c1># qval_cutoff = 0.8</span>
</span></span><span class=line><span class=cl>  <span class=c1># group_colors = lf_col[c(1, 2)]</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=c1># group_names</span>
</span></span><span class=line><span class=cl>  <span class=n>group_names</span> <span class=o>&lt;-</span> <span class=nf>gsub</span><span class=p>(</span><span class=s>&#34;\\d+_&#34;</span><span class=p>,</span> <span class=s>&#34;&#34;</span><span class=p>,</span> <span class=nf>unlist</span><span class=p>(</span><span class=nf>strsplit</span><span class=p>(</span><span class=n>DA1</span><span class=o>$</span><span class=n>Block[1]</span><span class=p>,</span> <span class=s>&#34; vs &#34;</span><span class=p>)))</span>  
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>DA_lefse_cln</span> <span class=o>&lt;-</span> <span class=n>DA1</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>Block</span><span class=p>,</span> <span class=n>LDA_Score</span><span class=p>,</span> <span class=n>Enrichment</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>Kingdom</span><span class=p>,</span> <span class=n>Phylum</span><span class=p>,</span> <span class=n>Genus</span><span class=p>,</span> <span class=n>Species</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>Lefse_Enrichment</span> <span class=o>=</span> <span class=n>Enrichment</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>DA_maaslin2_cln</span> <span class=o>&lt;-</span> <span class=n>DA2</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>coef</span><span class=p>,</span> <span class=n>Enrichment</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>pval</span><span class=p>,</span> <span class=n>qval</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>lg2fc</span> <span class=o>=</span> <span class=nf>log2</span><span class=p>(</span><span class=nf>exp</span><span class=p>(</span><span class=n>coef</span><span class=p>)))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>Maaslin2_Enrichment</span> <span class=o>=</span> <span class=n>Enrichment</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>Maaslin2_pval</span> <span class=o>=</span> <span class=n>pval</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>Maaslin2_qval</span> <span class=o>=</span> <span class=n>qval</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>Maaslin2_lg2fc</span> <span class=o>=</span> <span class=n>lg2fc</span><span class=p>)</span> 
</span></span><span class=line><span class=cl>  <span class=n>DA_maaslin2_cln</span><span class=nf>[which</span><span class=p>(</span><span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_lg2fc</span> <span class=o>&gt;</span> <span class=n>lg2fc_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;Maaslin2_Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[2]</span>
</span></span><span class=line><span class=cl>  <span class=n>DA_maaslin2_cln</span><span class=nf>[which</span><span class=p>(</span><span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_lg2fc</span> <span class=o>&lt;</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;Maaslin2_Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[1]</span>
</span></span><span class=line><span class=cl>  <span class=n>DA_maaslin2_cln</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_lg2fc</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>lg2fc_cutoff</span> <span class=o>|</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_pval</span> <span class=o>&gt;=</span> <span class=n>pval_cutoff</span> <span class=o>|</span>
</span></span><span class=line><span class=cl>            <span class=n>DA_maaslin2_cln</span><span class=o>$</span><span class=n>Maaslin2_qval</span> <span class=o>&gt;=</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;Maaslin2_Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignif&#34;</span>   
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>DA_ANCOMBC_cln</span> <span class=o>&lt;-</span> <span class=n>DA3</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>lfc</span><span class=p>,</span> <span class=n>Enrichment</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>p_val</span><span class=p>,</span> <span class=n>q_val</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>rename</span><span class=p>(</span><span class=n>ANCOMBC_Enrichment</span> <span class=o>=</span> <span class=n>Enrichment</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>ANCOMBC_pval</span> <span class=o>=</span> <span class=n>p_val</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>ANCOMBC_qval</span> <span class=o>=</span> <span class=n>q_val</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>ANCOMBC_lg2fc</span> <span class=o>=</span> <span class=n>lfc</span><span class=p>)</span> 
</span></span><span class=line><span class=cl>  <span class=n>DA_ANCOMBC_cln</span><span class=nf>[which</span><span class=p>(</span><span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_lg2fc</span> <span class=o>&gt;</span> <span class=n>lg2fc_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;ANCOMBC_Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[2]</span>
</span></span><span class=line><span class=cl>  <span class=n>DA_ANCOMBC_cln</span><span class=nf>[which</span><span class=p>(</span><span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_lg2fc</span> <span class=o>&lt;</span> <span class=o>-</span><span class=n>lg2fc_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_pval</span> <span class=o>&lt;</span> <span class=n>pval_cutoff</span> <span class=o>&amp;</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_qval</span> <span class=o>&lt;</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;ANCOMBC_Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=n>group_names[1]</span>
</span></span><span class=line><span class=cl>  <span class=n>DA_ANCOMBC_cln</span><span class=nf>[which</span><span class=p>(</span><span class=nf>abs</span><span class=p>(</span><span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_lg2fc</span><span class=p>)</span> <span class=o>&lt;=</span> <span class=n>lg2fc_cutoff</span> <span class=o>|</span> 
</span></span><span class=line><span class=cl>            <span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_pval</span> <span class=o>&gt;=</span> <span class=n>pval_cutoff</span> <span class=o>|</span>
</span></span><span class=line><span class=cl>            <span class=n>DA_ANCOMBC_cln</span><span class=o>$</span><span class=n>ANCOMBC_qval</span> <span class=o>&gt;=</span> <span class=n>qval_cutoff</span><span class=p>),</span>
</span></span><span class=line><span class=cl>      <span class=s>&#34;ANCOMBC_Enrichment&#34;</span><span class=n>]</span> <span class=o>&lt;-</span> <span class=s>&#34;Nonsignif&#34;</span>   
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>res</span> <span class=o>&lt;-</span> <span class=n>DA_lefse_cln</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>left_join</span><span class=p>(</span><span class=n>DA_maaslin2_cln</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                      <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;TaxaID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>left_join</span><span class=p>(</span><span class=n>DA_ANCOMBC_cln</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                      <span class=n>by</span> <span class=o>=</span> <span class=s>&#34;TaxaID&#34;</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=n>Block</span><span class=p>,</span> <span class=n>LDA_Score</span><span class=p>,</span> <span class=n>Lefse_Enrichment</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>Maaslin2_Enrichment</span><span class=p>,</span> <span class=n>coef</span><span class=p>,</span> <span class=n>Maaslin2_lg2fc</span><span class=p>,</span> <span class=n>Maaslin2_pval</span><span class=p>,</span> <span class=n>Maaslin2_qval</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>ANCOMBC_Enrichment</span><span class=p>,</span> <span class=n>ANCOMBC_lg2fc</span><span class=p>,</span> <span class=n>ANCOMBC_pval</span><span class=p>,</span> <span class=n>ANCOMBC_qval</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                  <span class=n>Kingdom</span><span class=p>,</span> <span class=n>Phylum</span><span class=p>,</span> <span class=n>Genus</span><span class=p>,</span> <span class=n>Species</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>group_by</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>Signif</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span>
</span></span><span class=line><span class=cl>      <span class=nf>all</span><span class=p>(</span><span class=n>Maaslin2_Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>          <span class=n>ANCOMBC_Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>),</span> <span class=s>&#34;+&#34;</span><span class=p>,</span> 
</span></span><span class=line><span class=cl>      <span class=nf>ifelse</span><span class=p>(</span><span class=n>Maaslin2_Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34;*&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>        <span class=nf>ifelse</span><span class=p>(</span><span class=n>ANCOMBC_Enrichment</span> <span class=o>!=</span> <span class=s>&#34;Nonsignif&#34;</span><span class=p>,</span> <span class=s>&#34;#&#34;</span><span class=p>,</span> <span class=s>&#34;&#34;</span><span class=p>))))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>    
</span></span><span class=line><span class=cl>    <span class=c1># dplyr::mutate(Signif = ifelse(Maaslin2_Enrichment != &#34;Nonsignif&#34;, &#34;*&#34;,</span>
</span></span><span class=line><span class=cl>    <span class=c1>#                               ifelse(ANCOMBC_Enrichment != &#34;Nonsignif&#34;, &#34;#&#34;, </span>
</span></span><span class=line><span class=cl>    <span class=c1>#                                      ifelse(all(Maaslin2_Enrichment != &#34;Nonsignif&#34;,</span>
</span></span><span class=line><span class=cl>    <span class=c1>#                                                 ANCOMBC_Enrichment != &#34;Nonsignif&#34;), </span>
</span></span><span class=line><span class=cl>    <span class=c1>#                                             &#34;+&#34;, &#34;&#34;)))) %&gt;%</span>
</span></span><span class=line><span class=cl>    <span class=n>dplyr</span><span class=o>::</span><span class=nf>ungroup</span><span class=p>()</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=nf>length</span><span class=p>(</span><span class=nf>grep</span><span class=p>(</span><span class=s>&#34;^t__&#34;</span><span class=p>,</span> <span class=n>res</span><span class=o>$</span><span class=n>TaxaID</span><span class=p>))</span> <span class=o>&gt;</span> <span class=m>1</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>plotdata</span> <span class=o>&lt;-</span> <span class=n>res</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>group_by</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>TaxaID</span> <span class=o>=</span> <span class=nf>gsub</span><span class=p>(</span><span class=s>&#34;t__&#34;</span><span class=p>,</span> <span class=s>&#34;&#34;</span><span class=p>,</span> <span class=n>TaxaID</span><span class=p>))</span> <span class=o>%&gt;%</span> 
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>ungroup</span><span class=p>()</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>FeatureID</span> <span class=o>=</span> <span class=nf>paste0</span><span class=p>(</span><span class=n>TaxaID</span><span class=p>,</span> <span class=s>&#34; (&#34;</span><span class=p>,</span> <span class=n>Species</span><span class=p>,</span> <span class=s>&#34;)&#34;</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>LDA_Score</span><span class=p>,</span> <span class=n>Lefse_Enrichment</span><span class=p>,</span> <span class=n>Signif</span><span class=p>,</span> <span class=n>Phylum</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>Lefse_Enrichment</span> <span class=o>=</span> <span class=nf>factor</span><span class=p>(</span><span class=n>Lefse_Enrichment</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>group_names</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span>
</span></span><span class=line><span class=cl>        <span class=n>FeatureID</span> <span class=o>=</span> <span class=nf>factor</span><span class=p>(</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>FeatureID</span><span class=nf>[order</span><span class=p>(</span><span class=n>LDA_Score</span><span class=p>,</span> <span class=n>decreasing</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>)</span><span class=n>]</span><span class=p>),</span>
</span></span><span class=line><span class=cl>        <span class=n>label_y</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=n>LDA_Score</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>,</span> <span class=m>0.2</span><span class=p>,</span> <span class=m>-0.2</span><span class=p>),</span>
</span></span><span class=line><span class=cl>        <span class=n>label_hjust</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=n>LDA_Score</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>,</span> <span class=m>1</span><span class=p>),</span>
</span></span><span class=line><span class=cl>        <span class=n>label_hjust2</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=n>LDA_Score</span> <span class=o>&gt;</span> <span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>,</span> <span class=m>1</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>plotdata</span> <span class=o>&lt;-</span> <span class=n>res</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>FeatureID</span> <span class=o>=</span> <span class=n>TaxaID</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>select</span><span class=p>(</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>LDA_Score</span><span class=p>,</span> <span class=n>Lefse_Enrichment</span><span class=p>,</span> <span class=n>Signif</span><span class=p>,</span> <span class=n>Phylum</span><span class=p>)</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>Lefse_Enrichment</span> <span class=o>=</span> <span class=nf>factor</span><span class=p>(</span><span class=n>Lefse_Enrichment</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>group_names</span><span class=p>))</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span>
</span></span><span class=line><span class=cl>        <span class=n>FeatureID</span> <span class=o>=</span> <span class=nf>factor</span><span class=p>(</span><span class=n>FeatureID</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=n>FeatureID</span><span class=nf>[order</span><span class=p>(</span><span class=n>LDA_Score</span><span class=p>,</span> <span class=n>decreasing</span> <span class=o>=</span> <span class=kc>TRUE</span><span class=p>)</span><span class=n>]</span><span class=p>),</span>
</span></span><span class=line><span class=cl>        <span class=n>label_y</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=n>LDA_Score</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>,</span> <span class=m>0.2</span><span class=p>,</span> <span class=m>-0.2</span><span class=p>),</span>
</span></span><span class=line><span class=cl>        <span class=n>label_hjust</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=n>LDA_Score</span> <span class=o>&lt;</span> <span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>,</span> <span class=m>1</span><span class=p>),</span>
</span></span><span class=line><span class=cl>        <span class=n>label_hjust2</span> <span class=o>=</span> <span class=nf>ifelse</span><span class=p>(</span><span class=n>LDA_Score</span> <span class=o>&gt;</span> <span class=m>0</span><span class=p>,</span> <span class=m>0</span><span class=p>,</span> <span class=m>1</span><span class=p>))</span> 
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>if </span><span class=p>(</span><span class=nf>length</span><span class=p>(</span><span class=nf>grep</span><span class=p>(</span><span class=s>&#34;unclassified&#34;</span><span class=p>,</span> <span class=n>plotdata</span><span class=o>$</span><span class=n>Phylum</span><span class=p>))</span> <span class=o>&gt;</span> <span class=m>0</span><span class=p>)</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>plotdata</span> <span class=o>&lt;-</span> <span class=n>plotdata[</span><span class=o>-</span><span class=nf>grep</span><span class=p>(</span><span class=s>&#34;unclassified&#34;</span><span class=p>,</span> <span class=n>plotdata</span><span class=o>$</span><span class=n>Phylum</span><span class=p>),</span> <span class=p>,</span> <span class=n>]</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>Phylum</span> <span class=o>=</span> <span class=nf>gsub</span><span class=p>(</span><span class=s>&#34;p__&#34;</span><span class=p>,</span> <span class=s>&#34;&#34;</span><span class=p>,</span> <span class=n>Phylum</span><span class=p>))</span>    
</span></span><span class=line><span class=cl>  <span class=p>}</span> <span class=n>else</span> <span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>plotdata</span> <span class=o>&lt;-</span> <span class=n>plotdata</span> <span class=o>%&gt;%</span>
</span></span><span class=line><span class=cl>      <span class=n>dplyr</span><span class=o>::</span><span class=nf>mutate</span><span class=p>(</span><span class=n>Phylum</span> <span class=o>=</span> <span class=nf>gsub</span><span class=p>(</span><span class=s>&#34;p__&#34;</span><span class=p>,</span> <span class=s>&#34;&#34;</span><span class=p>,</span> <span class=n>Phylum</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=p>}</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>plotdata</span><span class=o>$</span><span class=n>Phylum</span> <span class=o>&lt;-</span> <span class=nf>factor</span><span class=p>(</span><span class=n>plotdata</span><span class=o>$</span><span class=n>Phylum</span><span class=p>,</span> <span class=n>levels</span> <span class=o>=</span> <span class=nf>unique</span><span class=p>(</span><span class=n>plotdata</span><span class=o>$</span><span class=n>Phylum</span><span class=p>))</span>
</span></span><span class=line><span class=cl>  <span class=n>phylum_colors</span> <span class=o>&lt;-</span> <span class=n>RColorBrewer</span><span class=o>::</span><span class=nf>brewer.pal</span><span class=p>(</span><span class=n>n</span> <span class=o>=</span> <span class=nf>length</span><span class=p>(</span><span class=nf>levels</span><span class=p>(</span><span class=n>plotdata</span><span class=o>$</span><span class=n>Phylum</span><span class=p>)),</span> 
</span></span><span class=line><span class=cl>                                            <span class=n>name</span> <span class=o>=</span> <span class=s>&#34;Dark2&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=n>pl</span> <span class=o>&lt;-</span> <span class=nf>ggplot</span><span class=p>(</span><span class=n>plotdata</span><span class=p>,</span> <span class=nf>aes</span><span class=p>(</span><span class=n>x</span> <span class=o>=</span> <span class=n>FeatureID</span><span class=p>,</span> <span class=n>y</span> <span class=o>=</span> <span class=n>LDA_Score</span><span class=p>,</span> <span class=n>fill</span> <span class=o>=</span> <span class=n>Lefse_Enrichment</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_bar</span><span class=p>(</span><span class=n>stat</span> <span class=o>=</span> <span class=s>&#34;identity&#34;</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=s>&#34;white&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_text</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>y</span> <span class=o>=</span> <span class=n>label_y</span><span class=p>,</span> <span class=n>label</span> <span class=o>=</span> <span class=n>FeatureID</span><span class=p>,</span> <span class=n>hjust</span> <span class=o>=</span> <span class=n>label_hjust</span><span class=p>,</span> <span class=n>color</span> <span class=o>=</span> <span class=n>Phylum</span><span class=p>),</span> <span class=n>fontface</span> <span class=o>=</span> <span class=s>&#34;italic&#34;</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>geom_text</span><span class=p>(</span><span class=nf>aes</span><span class=p>(</span><span class=n>y</span> <span class=o>=</span> <span class=n>LDA_Score</span><span class=p>,</span> <span class=n>label</span> <span class=o>=</span> <span class=n>Signif</span><span class=p>,</span> <span class=n>hjust</span> <span class=o>=</span> <span class=n>label_hjust2</span><span class=p>),</span> <span class=n>size</span> <span class=o>=</span> <span class=m>6</span><span class=p>)</span> <span class=o>+</span>    
</span></span><span class=line><span class=cl>    <span class=nf>coord_flip</span><span class=p>()</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>scale_y_continuous</span><span class=p>(</span><span class=nf>expression</span><span class=p>(</span><span class=n>log[10</span><span class=nf>]</span><span class=p>(</span><span class=nf>italic</span><span class=p>(</span><span class=s>&#34;LDA score&#34;</span><span class=p>))),</span>
</span></span><span class=line><span class=cl>                       <span class=n>breaks</span> <span class=o>=</span> <span class=m>-4</span><span class=o>:</span><span class=m>4</span><span class=p>,</span> <span class=n>limits</span> <span class=o>=</span> <span class=nf>c</span><span class=p>(</span><span class=m>-5</span><span class=p>,</span> <span class=m>5</span><span class=p>))</span> <span class=o>+</span>  
</span></span><span class=line><span class=cl>    <span class=nf>scale_fill_manual</span><span class=p>(</span><span class=n>name</span> <span class=o>=</span> <span class=kc>NULL</span><span class=p>,</span> <span class=n>values</span> <span class=o>=</span> <span class=n>group_colors</span><span class=p>)</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>scale_color_manual</span><span class=p>(</span><span class=n>name</span> <span class=o>=</span> <span class=s>&#34;Phylum&#34;</span><span class=p>,</span> <span class=n>values</span> <span class=o>=</span> <span class=n>phylum_colors</span><span class=p>)</span> <span class=o>+</span>    
</span></span><span class=line><span class=cl>    <span class=nf>guides</span><span class=p>(</span><span class=n>fill</span> <span class=o>=</span> <span class=nf>guide_legend</span><span class=p>(</span><span class=n>order</span> <span class=o>=</span> <span class=m>1</span><span class=p>),</span>
</span></span><span class=line><span class=cl>           <span class=n>color</span> <span class=o>=</span> <span class=nf>guide_legend</span><span class=p>(</span><span class=n>order</span> <span class=o>=</span> <span class=m>2</span><span class=p>))</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme_minimal</span><span class=p>()</span> <span class=o>+</span>
</span></span><span class=line><span class=cl>    <span class=nf>theme</span><span class=p>(</span><span class=n>axis.text.y</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>          <span class=n>axis.ticks.y</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>          <span class=n>axis.title.y</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.position</span> <span class=o>=</span> <span class=s>&#34;right&#34;</span><span class=p>,</span>
</span></span><span class=line><span class=cl>          <span class=n>legend.justification</span> <span class=o>=</span> <span class=m>0.5</span><span class=p>,</span>
</span></span><span class=line><span class=cl>          <span class=n>panel.grid.major.y</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>          <span class=n>panel.grid.minor.y</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>(),</span>
</span></span><span class=line><span class=cl>          <span class=n>panel.grid.major.x</span> <span class=o>=</span> <span class=nf>element_line</span><span class=p>(</span><span class=n>color</span> <span class=o>=</span> <span class=s>&#34;grey80&#34;</span><span class=p>,</span> <span class=n>linetype</span> <span class=o>=</span> <span class=s>&#34;dashed&#34;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>          <span class=n>panel.grid.minor.x</span> <span class=o>=</span> <span class=nf>element_blank</span><span class=p>())</span>
</span></span><span class=line><span class=cl>  
</span></span><span class=line><span class=cl>  <span class=nf>return</span><span class=p>(</span><span class=n>pl</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>Merge_DA_pl</span> <span class=o>&lt;-</span> <span class=nf>get_mergeRes</span><span class=p>(</span>
</span></span><span class=line><span class=cl>  <span class=n>DA1</span> <span class=o>=</span> <span class=n>lefse_df</span><span class=o>$</span><span class=n>test_res</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>DA2</span> <span class=o>=</span> <span class=n>Maaslin2_df</span><span class=o>$</span><span class=n>test_res</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>DA3</span> <span class=o>=</span> <span class=n>ANCOMBC_df</span><span class=o>$</span><span class=n>test_res</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>lg2fc_cutoff</span> <span class=o>=</span> <span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>pval_cutoff</span> <span class=o>=</span> <span class=m>0.05</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>qval_cutoff</span> <span class=o>=</span> <span class=m>0.8</span><span class=p>,</span>
</span></span><span class=line><span class=cl>  <span class=n>group_colors</span> <span class=o>=</span> <span class=n>lf_col</span><span class=nf>[c</span><span class=p>(</span><span class=m>1</span><span class=p>,</span> <span class=m>2</span><span class=p>)</span><span class=n>]</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>Merge_DA_pl</span>
</span></span></code></pre></div><img src=https://zouhua.top/post/omics_analysis/metagenomics/2024-02-06-differentialanalysis/index_files/figure-html/unnamed-chunk-20-1.png width=960><p><strong>Results</strong></p><ul><li><p>Differential abundance of metagenomic species measured by linear discriminant analysis of effect size (LefSe) according to the LiverFatClass group.</p></li><li><p>LDA; Linear discriminant analysis.</p></li><li><p><strong>+</strong> Multivariate analysis (ANCOM-BC & Maaslin2) with a false discovery rate (FDR) adjusted p-value &lt; 0.8.</p></li><li><p>* Multivariate analysis (Maaslin2) with a false discovery rate (FDR) adjusted p-value &lt; 0.8.</p></li><li><p><strong>#</strong> Multivariate analysis (ANCOM-BC) with a false discovery rate (FDR) adjusted p-value &lt; 0.8.</p></li></ul><h2 id=conclusion>Conclusion</h2><ul><li><p>There were relatively high overlap of differential microbiota identified by the three different approaches when using LefSe as a benchmark.</p></li><li><p>Of the three difference methods, the Maaslin2 method was relatively the most stringent and screened for fewer difference species, whereas LefSe and ANCOM-BC screened for more species. LefSe did not take into account multiple test correction, and ANCOM-BC used zero inflation as well as correcting for between-sample variation methods, which resulted in more difference species being identified. It is important to note that, as can be seen from the boxplots of individual species later on, strict filtering of species occurrence is required, which can prevent low occurrence species from being identified (by choosing the appropriate filtering parameters according to the purpose of the study).</p></li></ul><h2 id=session-info>Session info</h2><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=n>devtools</span><span class=o>::</span><span class=nf>session_info</span><span class=p>()</span>
</span></span></code></pre></div><pre><code>##  Session info 
##  setting  value
##  version  R version 4.3.1 (2023-06-16)
##  os       macOS Monterey 12.2.1
##  system   x86_64, darwin20
##  ui       X11
##  language (EN)
##  collate  en_US.UTF-8
##  ctype    en_US.UTF-8
##  tz       Asia/Shanghai
##  date     2024-02-07
##  pandoc   3.1.3 @ /Users/zouhua/opt/anaconda3/bin/ (via rmarkdown)
## 
##  Packages 
##  package                  * version    date (UTC) lib source
##  abind                      1.4-5      2016-07-21 [1] CRAN (R 4.3.0)
##  ade4                       1.7-22     2023-02-06 [1] CRAN (R 4.3.0)
##  ANCOMBC                  * 2.4.0      2023-10-24 [1] Bioconductor
##  ape                        5.7-1      2023-03-13 [1] CRAN (R 4.3.0)
##  backports                  1.4.1      2021-12-13 [1] CRAN (R 4.3.0)
##  base64enc                  0.1-3      2015-07-28 [1] CRAN (R 4.3.0)
##  beachmat                   2.18.0     2023-10-24 [1] Bioconductor
##  beeswarm                   0.4.0      2021-06-01 [1] CRAN (R 4.3.0)
##  biglm                      0.9-2.1    2020-11-27 [1] CRAN (R 4.3.0)
##  Biobase                    2.62.0     2023-10-24 [1] Bioconductor
##  BiocGenerics               0.48.1     2023-11-01 [1] Bioconductor
##  BiocNeighbors              1.20.2     2024-01-07 [1] Bioconductor 3.18 (R 4.3.2)
##  BiocParallel               1.36.0     2023-10-24 [1] Bioconductor
##  BiocSingular               1.18.0     2023-10-24 [1] Bioconductor
##  biomformat                 1.30.0     2023-10-24 [1] Bioconductor
##  Biostrings                 2.70.2     2024-01-28 [1] Bioconductor 3.18 (R 4.3.2)
##  bit                        4.0.5      2022-11-15 [1] CRAN (R 4.3.0)
##  bit64                      4.0.5      2020-08-30 [1] CRAN (R 4.3.0)
##  bitops                     1.0-7      2021-04-24 [1] CRAN (R 4.3.0)
##  blob                       1.2.4      2023-03-17 [1] CRAN (R 4.3.0)
##  blogdown                   1.19       2024-02-01 [1] CRAN (R 4.3.2)
##  bluster                    1.12.0     2023-10-24 [1] Bioconductor
##  bookdown                   0.37       2023-12-01 [1] CRAN (R 4.3.0)
##  boot                       1.3-28.1   2022-11-22 [1] CRAN (R 4.3.1)
##  broom                      1.0.5      2023-06-09 [1] CRAN (R 4.3.0)
##  bslib                      0.6.1      2023-11-28 [1] CRAN (R 4.3.0)
##  cachem                     1.0.8      2023-05-01 [1] CRAN (R 4.3.0)
##  car                        3.1-2      2023-03-30 [1] CRAN (R 4.3.0)
##  carData                    3.0-5      2022-01-06 [1] CRAN (R 4.3.0)
##  caTools                    1.18.2     2021-03-28 [1] CRAN (R 4.3.0)
##  cellranger                 1.1.0      2016-07-27 [1] CRAN (R 4.3.0)
##  checkmate                  2.3.1      2023-12-04 [1] CRAN (R 4.3.0)
##  class                      7.3-22     2023-05-03 [1] CRAN (R 4.3.1)
##  cli                        3.6.2      2023-12-11 [1] CRAN (R 4.3.0)
##  cluster                    2.1.4      2022-08-22 [1] CRAN (R 4.3.1)
##  codetools                  0.2-19     2023-02-01 [1] CRAN (R 4.3.1)
##  colorspace                 2.1-0      2023-01-23 [1] CRAN (R 4.3.0)
##  cowplot                    1.1.3      2024-01-22 [1] CRAN (R 4.3.2)
##  crayon                     1.5.2      2022-09-29 [1] CRAN (R 4.3.0)
##  curl                       5.2.0      2023-12-08 [1] CRAN (R 4.3.0)
##  CVXR                       1.0-12     2024-02-02 [1] CRAN (R 4.3.2)
##  data.table                 1.15.0     2024-01-30 [1] CRAN (R 4.3.2)
##  DBI                        1.2.1      2024-01-12 [1] CRAN (R 4.3.0)
##  DECIPHER                   2.30.0     2023-10-24 [1] Bioconductor
##  decontam                   1.22.0     2023-10-24 [1] Bioconductor
##  DelayedArray               0.28.0     2023-10-24 [1] Bioconductor
##  DelayedMatrixStats         1.24.0     2023-10-24 [1] Bioconductor
##  DEoptimR                   1.1-3      2023-10-07 [1] CRAN (R 4.3.0)
##  DescTools                  0.99.54    2024-02-03 [1] CRAN (R 4.3.2)
##  DESeq2                     1.42.0     2023-10-24 [1] Bioconductor
##  devtools                   2.4.5      2022-10-11 [1] CRAN (R 4.3.0)
##  digest                     0.6.34     2024-01-11 [1] CRAN (R 4.3.0)
##  DirichletMultinomial       1.44.0     2023-10-24 [1] Bioconductor
##  doParallel                 1.0.17     2022-02-07 [1] CRAN (R 4.3.0)
##  doRNG                    * 1.8.6      2023-01-16 [1] CRAN (R 4.3.0)
##  dplyr                    * 1.1.4      2023-11-17 [1] CRAN (R 4.3.0)
##  e1071                      1.7-14     2023-12-06 [1] CRAN (R 4.3.0)
##  ellipsis                   0.3.2      2021-04-29 [1] CRAN (R 4.3.0)
##  energy                     1.7-11     2022-12-22 [1] CRAN (R 4.3.0)
##  evaluate                   0.23       2023-11-01 [1] CRAN (R 4.3.0)
##  Exact                      3.2        2022-09-25 [1] CRAN (R 4.3.0)
##  expm                       0.999-9    2024-01-11 [1] CRAN (R 4.3.0)
##  fansi                      1.0.6      2023-12-08 [1] CRAN (R 4.3.0)
##  farver                     2.1.1      2022-07-06 [1] CRAN (R 4.3.0)
##  fastmap                    1.1.1      2023-02-24 [1] CRAN (R 4.3.0)
##  forcats                  * 1.0.0      2023-01-29 [1] CRAN (R 4.3.0)
##  foreach                  * 1.5.2      2022-02-02 [1] CRAN (R 4.3.0)
##  foreign                    0.8-84     2022-12-06 [1] CRAN (R 4.3.1)
##  Formula                    1.2-5      2023-02-24 [1] CRAN (R 4.3.0)
##  fs                         1.6.3      2023-07-20 [1] CRAN (R 4.3.0)
##  generics                   0.1.3      2022-07-05 [1] CRAN (R 4.3.0)
##  GenomeInfoDb               1.38.5     2023-12-28 [1] Bioconductor 3.18 (R 4.3.2)
##  GenomeInfoDbData           1.2.11     2024-01-24 [1] Bioconductor
##  GenomicRanges              1.54.1     2023-10-29 [1] Bioconductor
##  getopt                     1.20.4     2023-10-01 [1] CRAN (R 4.3.0)
##  ggbeeswarm                 0.7.2      2023-04-29 [1] CRAN (R 4.3.0)
##  ggplot2                  * 3.4.4      2023-10-12 [1] CRAN (R 4.3.0)
##  ggpubr                   * 0.6.0      2023-02-10 [1] CRAN (R 4.3.0)
##  ggrepel                    0.9.5      2024-01-10 [1] CRAN (R 4.3.0)
##  ggsignif                   0.6.4      2022-10-13 [1] CRAN (R 4.3.0)
##  gld                        2.6.6      2022-10-23 [1] CRAN (R 4.3.0)
##  glmnet                     4.1-8      2023-08-22 [1] CRAN (R 4.3.0)
##  glue                       1.7.0      2024-01-09 [1] CRAN (R 4.3.0)
##  gmp                        0.7-4      2024-01-15 [1] CRAN (R 4.3.0)
##  gplots                     3.1.3.1    2024-02-02 [1] CRAN (R 4.3.2)
##  gridExtra                  2.3        2017-09-09 [1] CRAN (R 4.3.0)
##  gsl                        2.1-8      2023-01-24 [1] CRAN (R 4.3.0)
##  gtable                     0.3.4      2023-08-21 [1] CRAN (R 4.3.0)
##  gtools                     3.9.5      2023-11-20 [1] CRAN (R 4.3.0)
##  hash                       2.2.6.3    2023-08-19 [1] CRAN (R 4.3.0)
##  highr                      0.10       2022-12-22 [1] CRAN (R 4.3.0)
##  Hmisc                      5.1-1      2023-09-12 [1] CRAN (R 4.3.0)
##  hms                        1.1.3      2023-03-21 [1] CRAN (R 4.3.0)
##  htmlTable                  2.4.2      2023-10-29 [1] CRAN (R 4.3.0)
##  htmltools                  0.5.7      2023-11-03 [1] CRAN (R 4.3.0)
##  htmlwidgets                1.6.4      2023-12-06 [1] CRAN (R 4.3.0)
##  httpuv                     1.6.14     2024-01-26 [1] CRAN (R 4.3.2)
##  httr                       1.4.7      2023-08-15 [1] CRAN (R 4.3.0)
##  igraph                     2.0.1.1    2024-01-30 [1] CRAN (R 4.3.2)
##  IRanges                    2.36.0     2023-10-24 [1] Bioconductor
##  irlba                      2.3.5.1    2022-10-03 [1] CRAN (R 4.3.0)
##  iterators                  1.0.14     2022-02-05 [1] CRAN (R 4.3.0)
##  jquerylib                  0.1.4      2021-04-26 [1] CRAN (R 4.3.0)
##  jsonlite                   1.8.8      2023-12-04 [1] CRAN (R 4.3.0)
##  KernSmooth                 2.23-21    2023-05-03 [1] CRAN (R 4.3.1)
##  knitr                      1.45       2023-10-30 [1] CRAN (R 4.3.0)
##  labeling                   0.4.3      2023-08-29 [1] CRAN (R 4.3.0)
##  later                      1.3.2      2023-12-06 [1] CRAN (R 4.3.0)
##  lattice                    0.21-8     2023-04-05 [1] CRAN (R 4.3.1)
##  lazyeval                   0.2.2      2019-03-15 [1] CRAN (R 4.3.0)
##  lifecycle                  1.0.4      2023-11-07 [1] CRAN (R 4.3.0)
##  limma                      3.58.1     2023-10-31 [1] Bioconductor
##  lme4                       1.1-35.1   2023-11-05 [1] CRAN (R 4.3.0)
##  lmerTest                   3.1-3      2020-10-23 [1] CRAN (R 4.3.0)
##  lmom                       3.0        2023-08-29 [1] CRAN (R 4.3.0)
##  locfit                     1.5-9.8    2023-06-11 [1] CRAN (R 4.3.0)
##  logging                    0.10-108   2019-07-14 [1] CRAN (R 4.3.0)
##  lpsymphony                 1.30.0     2023-10-24 [1] Bioconductor (R 4.3.1)
##  lubridate                * 1.9.3      2023-09-27 [1] CRAN (R 4.3.0)
##  Maaslin2                 * 1.7.3      2024-01-24 [1] Bioconductor
##  magrittr                   2.0.3      2022-03-30 [1] CRAN (R 4.3.0)
##  MASS                       7.3-60     2023-05-04 [1] CRAN (R 4.3.1)
##  Matrix                     1.6-5      2024-01-11 [1] CRAN (R 4.3.0)
##  MatrixGenerics             1.14.0     2023-10-24 [1] Bioconductor
##  matrixStats                1.2.0      2023-12-11 [1] CRAN (R 4.3.0)
##  memoise                    2.0.1      2021-11-26 [1] CRAN (R 4.3.0)
##  metagenomeSeq              1.43.0     2023-04-25 [1] Bioconductor
##  mgcv                       1.8-42     2023-03-02 [1] CRAN (R 4.3.1)
##  mia                        1.10.0     2023-10-24 [1] Bioconductor
##  MicrobiomeAnalysis       * 1.0.3      2024-02-06 [1] Github (HuaZou/MicrobiomeAnalysis@fd2a6a2)
##  mime                       0.12       2021-09-28 [1] CRAN (R 4.3.0)
##  miniUI                     0.1.1.1    2018-05-18 [1] CRAN (R 4.3.0)
##  minqa                      1.2.6      2023-09-11 [1] CRAN (R 4.3.0)
##  multcomp                   1.4-25     2023-06-20 [1] CRAN (R 4.3.0)
##  MultiAssayExperiment       1.28.0     2023-10-24 [1] Bioconductor
##  multtest                   2.58.0     2023-10-24 [1] Bioconductor
##  munsell                    0.5.0      2018-06-12 [1] CRAN (R 4.3.0)
##  mvtnorm                    1.2-4      2023-11-27 [1] CRAN (R 4.3.0)
##  nlme                       3.1-162    2023-01-31 [1] CRAN (R 4.3.1)
##  nloptr                     2.0.3      2022-05-26 [1] CRAN (R 4.3.0)
##  nnet                       7.3-19     2023-05-03 [1] CRAN (R 4.3.1)
##  numDeriv                   2016.8-1.1 2019-06-06 [1] CRAN (R 4.3.0)
##  optparse                   1.7.4      2024-01-16 [1] CRAN (R 4.3.0)
##  pbapply                    1.7-2      2023-06-27 [1] CRAN (R 4.3.0)
##  pcaPP                      2.0-4      2023-12-07 [1] CRAN (R 4.3.0)
##  permute                    0.9-7      2022-01-27 [1] CRAN (R 4.3.0)
##  phyloseq                 * 1.46.0     2023-10-24 [1] Bioconductor
##  pillar                     1.9.0      2023-03-22 [1] CRAN (R 4.3.0)
##  pkgbuild                   1.4.3      2023-12-10 [1] CRAN (R 4.3.0)
##  pkgconfig                  2.0.3      2019-09-22 [1] CRAN (R 4.3.0)
##  pkgload                    1.3.4      2024-01-16 [1] CRAN (R 4.3.0)
##  plyr                       1.8.9      2023-10-02 [1] CRAN (R 4.3.0)
##  profvis                    0.3.8      2023-05-02 [1] CRAN (R 4.3.0)
##  promises                   1.2.1      2023-08-10 [1] CRAN (R 4.3.0)
##  proxy                      0.4-27     2022-06-09 [1] CRAN (R 4.3.0)
##  purrr                    * 1.0.2      2023-08-10 [1] CRAN (R 4.3.0)
##  R6                         2.5.1      2021-08-19 [1] CRAN (R 4.3.0)
##  rbibutils                  2.2.16     2023-10-25 [1] CRAN (R 4.3.0)
##  RColorBrewer               1.1-3      2022-04-03 [1] CRAN (R 4.3.0)
##  Rcpp                       1.0.12     2024-01-09 [1] CRAN (R 4.3.0)
##  RCurl                      1.98-1.14  2024-01-09 [1] CRAN (R 4.3.0)
##  Rdpack                     2.6        2023-11-08 [1] CRAN (R 4.3.0)
##  readr                    * 2.1.5      2024-01-10 [1] CRAN (R 4.3.0)
##  readxl                     1.4.3      2023-07-06 [1] CRAN (R 4.3.0)
##  remotes                    2.4.2.1    2023-07-18 [1] CRAN (R 4.3.0)
##  reshape2                   1.4.4      2020-04-09 [1] CRAN (R 4.3.0)
##  rhdf5                      2.46.1     2023-11-29 [1] Bioconductor
##  rhdf5filters               1.14.1     2023-11-06 [1] Bioconductor
##  Rhdf5lib                   1.24.1     2023-12-12 [1] Bioconductor 3.18 (R 4.3.2)
##  rlang                      1.1.3      2024-01-10 [1] CRAN (R 4.3.0)
##  rmarkdown                  2.25       2023-09-18 [1] CRAN (R 4.3.0)
##  Rmpfr                      0.9-5      2024-01-21 [1] CRAN (R 4.3.0)
##  rngtools                 * 1.5.2      2021-09-20 [1] CRAN (R 4.3.0)
##  robustbase                 0.99-1     2023-11-29 [1] CRAN (R 4.3.0)
##  rootSolve                  1.8.2.4    2023-09-21 [1] CRAN (R 4.3.0)
##  rpart                      4.1.19     2022-10-21 [1] CRAN (R 4.3.1)
##  RSQLite                    2.3.5      2024-01-21 [1] CRAN (R 4.3.0)
##  rstatix                    0.7.2      2023-02-01 [1] CRAN (R 4.3.0)
##  rstudioapi                 0.15.0     2023-07-07 [1] CRAN (R 4.3.0)
##  rsvd                       1.0.5      2021-04-16 [1] CRAN (R 4.3.0)
##  S4Arrays                   1.2.0      2023-10-24 [1] Bioconductor
##  S4Vectors                  0.40.2     2023-11-23 [1] Bioconductor
##  sandwich                   3.1-0      2023-12-11 [1] CRAN (R 4.3.0)
##  sass                       0.4.8      2023-12-06 [1] CRAN (R 4.3.0)
##  ScaledMatrix               1.10.0     2023-10-24 [1] Bioconductor
##  scales                     1.3.0      2023-11-28 [1] CRAN (R 4.3.0)
##  scater                     1.30.1     2023-12-06 [1] Bioconductor
##  scuttle                    1.12.0     2023-10-24 [1] Bioconductor
##  sessioninfo                1.2.2      2021-12-06 [1] CRAN (R 4.3.0)
##  shape                      1.4.6      2021-05-19 [1] CRAN (R 4.3.0)
##  shiny                      1.8.0      2023-11-17 [1] CRAN (R 4.3.0)
##  SingleCellExperiment       1.24.0     2023-10-24 [1] Bioconductor
##  SparseArray                1.2.3      2023-12-25 [1] Bioconductor 3.18 (R 4.3.2)
##  sparseMatrixStats          1.14.0     2023-10-24 [1] Bioconductor
##  statmod                    1.5.0      2023-01-06 [1] CRAN (R 4.3.0)
##  stringi                    1.8.3      2023-12-11 [1] CRAN (R 4.3.0)
##  stringr                  * 1.5.1      2023-11-14 [1] CRAN (R 4.3.0)
##  SummarizedExperiment       1.32.0     2023-10-24 [1] Bioconductor
##  survival                   3.5-5      2023-03-12 [1] CRAN (R 4.3.1)
##  TH.data                    1.1-2      2023-04-17 [1] CRAN (R 4.3.0)
##  tibble                   * 3.2.1      2023-03-20 [1] CRAN (R 4.3.0)
##  tidyr                    * 1.3.1      2024-01-24 [1] CRAN (R 4.3.2)
##  tidyselect                 1.2.0      2022-10-10 [1] CRAN (R 4.3.0)
##  tidytree                   0.4.6      2023-12-12 [1] CRAN (R 4.3.0)
##  tidyverse                * 2.0.0      2023-02-22 [1] CRAN (R 4.3.0)
##  timechange                 0.3.0      2024-01-18 [1] CRAN (R 4.3.0)
##  treeio                     1.26.0     2023-10-24 [1] Bioconductor
##  TreeSummarizedExperiment   2.10.0     2023-10-24 [1] Bioconductor
##  tzdb                       0.4.0      2023-05-12 [1] CRAN (R 4.3.0)
##  urlchecker                 1.0.1      2021-11-30 [1] CRAN (R 4.3.0)
##  usethis                    2.2.2      2023-07-06 [1] CRAN (R 4.3.0)
##  utf8                       1.2.4      2023-10-22 [1] CRAN (R 4.3.0)
##  vctrs                      0.6.5      2023-12-01 [1] CRAN (R 4.3.0)
##  vegan                      2.6-4      2022-10-11 [1] CRAN (R 4.3.0)
##  vipor                      0.4.7      2023-12-18 [1] CRAN (R 4.3.0)
##  viridis                    0.6.5      2024-01-29 [1] CRAN (R 4.3.2)
##  viridisLite                0.4.2      2023-05-02 [1] CRAN (R 4.3.0)
##  withr                      3.0.0      2024-01-16 [1] CRAN (R 4.3.0)
##  Wrench                     1.20.0     2023-10-24 [1] Bioconductor
##  xfun                       0.41       2023-11-01 [1] CRAN (R 4.3.0)
##  xtable                     1.8-4      2019-04-21 [1] CRAN (R 4.3.0)
##  XVector                    0.42.0     2023-10-24 [1] Bioconductor
##  yaml                       2.3.8      2023-12-11 [1] CRAN (R 4.3.0)
##  yulab.utils                0.1.4      2024-01-28 [1] CRAN (R 4.3.2)
##  zlibbioc                   1.48.0     2023-10-24 [1] Bioconductor
##  zoo                        1.8-12     2023-04-13 [1] CRAN (R 4.3.0)
## 
##  [1] /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library
## 
## 
</code></pre><h2 id=reference>Reference</h2><div id=refs class="references csl-bib-body hanging-indent"><div id=ref-derosa2022intestinal class=csl-entry><p>Derosa, Lisa, Bertrand Routy, Andrew Maltez Thomas, Valerio Iebba, Gerard Zalcman, Sylvie Friard, Julien Mazieres, et al. 2022. Intestinal Akkermansia Muciniphila Predicts Clinical Response to PD-1 Blockade in Patients with Advanced Non-Small-Cell Lung Cancer. <em>Nature Medicine</em> 28 (2): 31524.</p></div></div></div><div class=article-tags><a class="badge badge-light" href=/tag/differential-analysis/>Differential Analysis</a></div><div class=share-box><ul class=share><li><a href="https://twitter.com/intent/tweet?url=https%3A%2F%2Fzouhua.top%2Fpost%2Fomics_analysis%2Fmetagenomics%2F2024-02-06-differentialanalysis%2F&text=Multiple+approaches+for+identification+of+microbial+biomarkers" target=_blank rel=noopener class=share-btn-twitter aria-label=twitter><i class="fab fa-twitter"></i></a></li><li><a href="https://www.facebook.com/sharer.php?u=https%3A%2F%2Fzouhua.top%2Fpost%2Fomics_analysis%2Fmetagenomics%2F2024-02-06-differentialanalysis%2F&t=Multiple+approaches+for+identification+of+microbial+biomarkers" target=_blank rel=noopener class=share-btn-facebook aria-label=facebook><i class="fab fa-facebook"></i></a></li><li><a href="mailto:?subject=Multiple%20approaches%20for%20identification%20of%20microbial%20biomarkers&body=https%3A%2F%2Fzouhua.top%2Fpost%2Fomics_analysis%2Fmetagenomics%2F2024-02-06-differentialanalysis%2F" target=_blank rel=noopener class=share-btn-email aria-label=envelope><i class="fas fa-envelope"></i></a></li><li><a href="https://www.linkedin.com/shareArticle?url=https%3A%2F%2Fzouhua.top%2Fpost%2Fomics_analysis%2Fmetagenomics%2F2024-02-06-differentialanalysis%2F&title=Multiple+approaches+for+identification+of+microbial+biomarkers" target=_blank rel=noopener class=share-btn-linkedin aria-label=linkedin-in><i class="fab fa-linkedin-in"></i></a></li><li><a href="whatsapp://send?text=Multiple+approaches+for+identification+of+microbial+biomarkers%20https%3A%2F%2Fzouhua.top%2Fpost%2Fomics_analysis%2Fmetagenomics%2F2024-02-06-differentialanalysis%2F" target=_blank rel=noopener class=share-btn-whatsapp aria-label=whatsapp><i class="fab fa-whatsapp"></i></a></li><li><a href="https://service.weibo.com/share/share.php?url=https%3A%2F%2Fzouhua.top%2Fpost%2Fomics_analysis%2Fmetagenomics%2F2024-02-06-differentialanalysis%2F&title=Multiple+approaches+for+identification+of+microbial+biomarkers" target=_blank rel=noopener class=share-btn-weibo aria-label=weibo><i class="fab fa-weibo"></i></a></li></ul></div><div class="media author-card content-widget-hr"><a href=https://zouhua.top/><img class="avatar mr-3 avatar-circle" src=/authors/admin/avatar_hu2ca0f82a4ba69bdab5bcba8a2e118070_85979_270x270_fill_q75_lanczos_center.jpg alt="Hua Zou"></a><div class=media-body><h5 class=card-title><a href=https://zouhua.top/>Hua Zou</a></h5><h6 class=card-subtitle>Senior Bioinformatic Analyst</h6><p class=card-text>My research interests include host-microbiota intersection, machine learning and multi-omics data integration.</p><ul class=network-icon aria-hidden=true><li><a href=/#contact><i class="fas fa-envelope"></i></a></li><li><a href=https://twitter.com/HuaZou01 target=_blank rel=noopener><i class="fab fa-twitter"></i></a></li><li><a href=https://github.com/HuaZou target=_blank rel=noopener><i class="fab fa-github"></i></a></li></ul></div></div></div></article></div><div class=page-footer><div class=container><footer class=site-footer><p class=powered-by> 2024 Hua Zou &#183;
Rendered by <a href=https://cran.r-project.org/ target=_blank rel=noopener><i class="fab fa-r-project"></i> </a><a href=https://github.com/rstudio/blogdown target=_blank rel=noopener>blogdown</a> ,
<a href=https://sourcethemes.com/academic/ target=_blank rel=noopener>Academic theme</a> and
<a href=https://gohugo.io target=_blank rel=noopener>Hugo</a>.
<span class=float-right aria-hidden=true><a href=# id=back_to_top><span class=button_icon><i class="fas fa-chevron-up fa-2x"></i></span></a></span></p></footer><script type=text/javascript src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script></div></div><script src=/js/vendor-bundle.min.d26509351aa0ff874abbee824e982e9b.js></script>
<script id=page-data type=application/json>{"use_headroom":true}</script><script src=/js/wowchemy-headroom.db4755770454eb63685f8de785c0a172.js type=module></script>
<script src=/en/js/wowchemy.min.e973d49cb2d568aa6f8eaf3638337473.js></script><div id=modal class="modal fade" role=dialog><div class=modal-dialog><div class=modal-content><div class=modal-header><h5 class=modal-title>Cite</h5><button type=button class=close data-dismiss=modal aria-label=Close>
<span aria-hidden=true>&#215;</span></button></div><div class=modal-body><pre><code></code></pre></div><div class=modal-footer><a class="btn btn-outline-primary my-1 js-copy-cite" href=# target=_blank><i class="fas fa-copy"></i> Copy</a>
<a class="btn btn-outline-primary my-1 js-download-cite" href=# target=_blank><i class="fas fa-download"></i> Download</a><div id=modal-error></div></div></div></div></div><script src=/js/wowchemy-publication.68f8d7090562ca65fc6d3cb3f8f2d2cb.js type=module></script></body></html>